---
title: "satmut_utils analysis"
author: "Ian Hoskins"
date: "11/24/2022"
output:
  html_document:
    df_print: paged
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, cache=TRUE)
```

```{r Setup}

library(data.table)
library(readxl)
library(ggplot2)
library(cowplot)
library(ggsci)
library(viridis)
library(pheatmap)
library(RColorBrewer)
library(factoextra)
library(ggrepel)
library(bnstruct)
library(viridis)
library(lawstat)
library(splines)
library(kernlab)
library(VennDiagram)

# Use BiocManager::install to install Bioconductor packages
library(limma)
library(edgeR)
library(Biostrings)
library(ggseqlogo)

base_dir<- "/Users/ianhoskins/data_staging/satmut_utils"
source("/Users/ianhoskins/satmut_utils/src/prototype/summarization_utils.r")
source("/Users/ianhoskins/satmut_utils/src/prototype/modeling_utils.r")

```

## sim on in silico reads with 183 variants at 0.1% frequency

```{r sim in silico reads- optimized alignment parameters}

# Set file paths
sim_insilico_reads_dir<- paste(base_dir, "sim_insilico_reads", sep="/")

sim_insilico_reads_vcf_summary<- paste(
  sim_insilico_reads_dir, "08Nov2021.12:53PM.read_gen.dna.r.var.cand.vcf.summary.txt", sep="/")

sim_insilico_reads_truth_vcf<- paste(
  sim_insilico_reads_dir, "08Nov2021.12:53PM.read_gen.dna.r.truth.vcf", sep="/")

# Read in the truth VCF and the calls
sim_insilico_reads_dt<- fread(sim_insilico_reads_vcf_summary, sep="\t", header=TRUE)

sim_insilico_reads_truth_dt<- fread(sim_insilico_reads_truth_vcf, skip=10)

# Determine sensitivity
sim_insilico_reads_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

sim_insilico_reads_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

sim_insilico_reads_sensitivity<- sim_insilico_reads_dt[
  ,sum(unique(VAR_ID)%in%sim_insilico_reads_truth_dt[,VAR_ID])/nrow(sim_insilico_reads_truth_dt)]

# Assess frequency accuracy
sim_insilico_reads_dt[,CAF_diff:=CAF-.001]

sim_insilico_reads_acc_dt<- sim_insilico_reads_dt[
  ,.(Mean_CAF_diff=mean(CAF_diff), SD_CAF_diff=sd(CAF_diff))]

# Inaccuracies are due to lack of numerical precision only

```

```{r sim in silico reads- default alignment parameters}

sim_insilico_reads_default_aln_dir<- paste(base_dir, "sim_insilico_reads_default_aln", sep="/")

sim_insilico_reads_default_aln_vcf_summary<- paste(
  sim_insilico_reads_default_aln_dir, "08Nov2021.12:53PM.read_gen.dna.r.var.cand.vcf.summary.txt", sep="/")

sim_insilico_reads_truth_vcf<- paste(
  sim_insilico_reads_dir, "08Nov2021.12:53PM.read_gen.dna.r.truth.vcf", sep="/")

# Read in the truth VCF and the calls
sim_insilico_reads_default_aln_dt<- fread(sim_insilico_reads_default_aln_vcf_summary, sep="\t", header=TRUE)

sim_insilico_reads_truth_dt<- fread(sim_insilico_reads_truth_vcf, skip=10)

# Determine sensitivity
sim_insilico_reads_default_aln_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

sim_insilico_reads_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

sim_insilico_reads_default_aln_sensitivity<- sim_insilico_reads_default_aln_dt[
  ,sum(unique(VAR_ID)%in%sim_insilico_reads_truth_dt[,VAR_ID])/nrow(sim_insilico_reads_truth_dt)]

```

## Saturation mutagenesis variant caller benchmarking: sim on CBS tile 6 

```{r CBS_pAG415GAL_tile6_sim}

count_thresh<- 100
software_levels<- c("DiMSum", "Enrich2", "dms_tools2", "satmut_utils")

npg_cols<- pal_npg("nrc")(10)
dimsum_col<- npg_cols[1]
dmstools2_col<- npg_cols[2]
enrich2_col<- npg_cols[3]
satmut_utils_col<- npg_cols[4]

compare_dir<- paste(base_dir, "variant_call_compare", "combined", sep="/")

CBS_pAG415GAL_tile6_sim_truth_vcf<- "CBS_pAG415GAL_tile6_sim.truth.vcf"

satmut_utils_counts<- "CBS_pAG415GAL_tile6_sim.var.cand.vcf.summary.txt"

# dms_tools2 reports a codon REF-ALT matrix, so we can merge variants
dmstools2_counts<- "CBS-pAG415GAL-tile6-sim-dmstools2-converted_codoncounts.csv"

# Enrich2 reports transcript changes, which must be preprocessed before merging
# additionally, separate files for synonymous and nonsynonymous changes
enrich2_main_counts<- "CBS_pAG415GAL_tile6_sim_basic_lib.h5.main_counts.txt"
enrich2_synon_counts<- "CBS_pAG415GAL_tile6_sim_basic_lib.h5.synon_counts.txt"

# DiMSum reports full variant sequences and only amino acid changes for 
# single and double codon changes separately
# must compare at the level of protein (loses variant/codon-specific information)
dimsum_single_counts<- "fitness_singles.txt"
dimsum_double_counts<- "fitness_doubles.txt"

CBS_pAG415GAL_tile6_sim_truth_dt<- fread(paste(compare_dir, CBS_pAG415GAL_tile6_sim_truth_vcf, sep="/"))
compare_satmut_utils_dt<- fread(paste(compare_dir, satmut_utils_counts, sep="/"))
compare_dmstools2_dt<- fread(paste(compare_dir, dmstools2_counts, sep="/"))
compare_enrich_main_dt<- fread(paste(compare_dir, enrich2_main_counts, sep="/"))
compare_enrich_synon_dt<- fread(paste(compare_dir, enrich2_synon_counts, sep="/"))
compare_dimsum_single_dt<- fread(paste(compare_dir, dimsum_single_counts, sep="/"))
compare_dimsum_double_dt<- fread(paste(compare_dir, dimsum_double_counts, sep="/"))

# Filter the truth VCF for only those variants that are expected
CBS_pAG415GAL_tile6_sim_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

CBS_pAG415GAL_tile6_sim_truth_dt[,CAO:=as.integer(sapply(strsplit(
  sapply(strsplit(INFO, ";"), "[", 1), "=", fixed=T), "[", 2))]

CBS_pAG415GAL_tile6_sim_truth_dt[,CAF:=as.numeric(sapply(strsplit(
  sapply(strsplit(INFO, ";"), "[", 2), "=", fixed=T), "[", 2))]

CBS_pAG415GAL_tile6_sim_truth_dt[,log10_CAF:=log10(CAF)]

# CBS_pAG415GAL_tile6_sim_truth_filt_dt<- CBS_pAG415GAL_tile6_sim_truth_dt[CAO>0]
# All variants should be called

# satmut_utils calls
# First preprocess the satmut_utils data and collapse MNPs
compare_satmut_utils_postprocess_dt<- postprocess_vcf_summary_dt(
  compare_satmut_utils_dt, key_features="VAR_ID")

compare_satmut_utils_postprocess_dt[,CAO:=as.integer(CAO)]

# Plot variant frequency accuracy for satmut_utils
compare_satmut_utils_merge_dt<- merge(CBS_pAG415GAL_tile6_sim_truth_dt[,.(VAR_ID, CAO, CAF, log10_CAF)], 
                                      compare_satmut_utils_postprocess_dt, by="VAR_ID", all=TRUE)

compare_satmut_utils_merge_dt[,log10_CAO_diff:=CAO.y-CAO.x]
compare_satmut_utils_merge_dt[,log10_CAF_diff:=log10_CAF.y-log10_CAF.x]
compare_satmut_utils_merge_dt[,TYPE:=factor(TYPE, levels=c("SNP", "di_nt_MNP", "tri_nt_MNP"))]

# UPDATE from bioRxiv manuscript: include Truth CAO == 0 for Suppl. Figure 2A
compare_satmut_utils_merge_dt[is.na(CAO.x), Truth:="F"]
compare_satmut_utils_merge_dt[!is.na(CAO.x), Truth:="T"]
compare_satmut_utils_merge_dt[,Truth:=factor(Truth, levels=c("F", "T"))]
compare_satmut_utils_merge_dt[is.na(CAO.x), CAO.x:=0]

# Plot counts for truth and satmut_utils
compare_satmut_utils_freq_acc_gg<- ggplot(
  data=compare_satmut_utils_merge_dt, 
  aes(x=log10(CAO.x+1), y=log10(CAO.y+1), col=Truth))

compare_satmut_utils_freq_acc_gg+
  facet_wrap(~TYPE)+
  scale_color_simpsons()+
  geom_point(alpha=0.7)+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(0,4.5), ylim=c(0,4.5))+
  labs(x="Expected count", 
       y="Observed count")+
  theme_cowplot()


# dms_tools2 calls
# Melt the dms_tools2 codon change matrix
# Name the count CAO to facilitate merge with satmut_utils data
compare_dmstools2_melt_dt<- melt(
  data=compare_dmstools2_dt, id.vars=c("site", "wildtype"), 
  variable.name="ALT", value.name="CAO")

# Adjust the site (AA position)
# POS 734 start matches p.G185
# POS 716 (start of primer) matches p.D179
tile6_aa_offset<- 178
compare_dmstools2_melt_dt[,AA_POS:=site+tile6_aa_offset]
compare_dmstools2_melt_dt[,dmstools_key:=paste(wildtype,AA_POS,ALT,sep=":")]

# FIXED bug from manuscript: ensure count filter is applied to dms_tools2 results for
# comparison
compare_dmstools2_melt_filt_dt<- compare_dmstools2_melt_dt[CAO>1 & wildtype!=ALT]

# Merge with the satmut_utils dataset
compare_satmut_utils_merge_dt[,dmstools_key:=paste(REF_CODON,AA_POS,ALT_CODON,sep=":")]
compare_satmut_utils_merge_dt[,Truth_CAO:=CAO.x]
compare_satmut_utils_merge_dt[,CAO:=CAO.y]

compare_satmut_utils_merge_dmstools_dt<- merge(
  compare_satmut_utils_merge_dt[
    !grepl("False", MATCHES_MUT_SIG), 
    .(VAR_ID, Truth_CAO, CAO, AA_CHANGE, MATCHES_MUT_SIG, TYPE, REF_CODON, ALT_CODON, dmstools_key)], 
  compare_dmstools2_melt_filt_dt[,.(CAO, dmstools_key)], 
  by="dmstools_key", all=TRUE)

compare_satmut_utils_merge_dmstools_dt[,Truth:="T"]
compare_satmut_utils_merge_dmstools_dt[is.na(Truth_CAO) | Truth_CAO==0, Truth:="F"]
compare_satmut_utils_merge_dmstools_dt[,Truth:=factor(Truth, levels=c("F", "T"))]

# Re-compute the TYPE for each variant so we can annotate the variants that 
# dms_tools2 called that were not in the truth VCF or satmut_utils callset
compare_satmut_utils_merge_dmstools_dt[,dmstools_REF_CODON:=sapply(strsplit(dmstools_key, ":"), "[", 1)]
compare_satmut_utils_merge_dmstools_dt[,dmstools_ALT_CODON:=sapply(strsplit(dmstools_key, ":"), "[", 3)]

compare_satmut_utils_merge_dmstools_dt[,HD:=hamming.distance(
  unlist(strsplit(dmstools_REF_CODON, "")), unlist(strsplit(dmstools_ALT_CODON, ""))), by=seq_len(nrow(compare_satmut_utils_merge_dmstools_dt))]

compare_satmut_utils_merge_dmstools_dt[HD==1,TYPE:="SNP"]
compare_satmut_utils_merge_dmstools_dt[HD==2,TYPE:="di_nt_MNP"]
compare_satmut_utils_merge_dmstools_dt[HD==3,TYPE:="tri_nt_MNP"]
compare_satmut_utils_merge_dmstools_dt[,TYPE:=factor(TYPE, levels=c("SNP", "di_nt_MNP", "tri_nt_MNP"))]
compare_satmut_utils_merge_dmstools_dt[is.na(CAO.x), CAO.x:=0]
compare_satmut_utils_merge_dmstools_dt[is.na(CAO.y), CAO.y:=0]

compare_satmut_utils_vs_dmstools_gg<- ggplot(
  data=compare_satmut_utils_merge_dmstools_dt, 
  aes(x=log10(CAO.x+1), y=log10(CAO.y+1), col=Truth))

compare_satmut_utils_vs_dmstools_gg+
  facet_wrap(~TYPE)+
  scale_color_simpsons()+
  geom_point(alpha=0.7)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(0,4.5), ylim=c(0,4.5))+
  labs(x="satmut_utils log10(counts + 1)", 
       y="dms_tools2 log10(counts + 1)")+
  theme_cowplot()

# So satmut_utils and dms_tools2 report very similar counts

# For dms_tools2 performance evalution:
# We must use the satmut_utils-merged dataset because no way to merge dms_tools2 data with
# truth datatable directly (given different annotations).
# Thus, filter out any variants that satmut_utils called that were not called in dms_tools2
# prior to determining the performance of dms_tools2; also apply count filter here
compare_satmut_utils_merge_dmstools_filt_dt<- compare_satmut_utils_merge_dmstools_dt[!is.na(CAO.y) & CAO.y > 1]

# compare_satmut_utils_merge_dmstools_filt_dt[,sum(Truth_CAO>0, na.rm=T)] is 281
# If it were not after filtering, we would have to determine a way to update the number 
# of Truth variants in contigency table; instead satmut_utils calls all truth variants so
# we always have annotations for them in the merged dt, which propagate to this dms_tools2 dt


# Enrich2 calls
# Discard the counts for WT and synonymous variants; not a parseable format
# Use satmut_utils callset to bootstrap up a VAR_ID from the available data
# When calling Enrich2, POS offset should be the offset from the first nt of the first codon, 
# not the POS offset from the transcript start
# Tile 6 primer starts at 715; CDS starts at 181; offset is 715-181=534
# for Enrich2 configuration

enrich_cds_offset<- 181
#compare_enrich_dt<- rbind(compare_enrich_main_dt[!grepl("_wt", V1)],compare_enrich_synon_dt[!grepl("_wt|_sy", V1)])

compare_enrich_dt<- compare_enrich_main_dt[!grepl("_wt", V1)]

# It looks like main contains synonymous variants and the synonymous table
# contains missense variants? Variants in synon table are reported in main table
# but with lower counts? Perhaps this table means synonymous variants
# phased with the missense variant?

#rm(list=c("compare_enrich_main_dt", "compare_enrich_synon_dt"))

compare_enrich_main_filt_dt<- compare_enrich_main_dt[!grepl("_wt", V1)]
compare_enrich_main_filt_dt[,first:=sapply(strsplit(V1, " "), "[", 1)]
compare_enrich_main_filt_dt[,second:=sapply(strsplit(V1, " "), "[", 3)]
compare_enrich_main_filt_dt[,third:=sapply(strsplit(V1, " "), "[", 5)]

compare_enrich_main_filt_dt[,first_pos:=as.integer(gsub("[[:alpha:]]|[[:punct:]]", "", first))+enrich_cds_offset]
compare_enrich_main_filt_dt[,second_pos:=as.integer(gsub("[[:alpha:]]|[[:punct:]]", "", second))+enrich_cds_offset]
compare_enrich_main_filt_dt[,third_pos:=as.integer(gsub("[[:alpha:]]|[[:punct:]]", "", third))+enrich_cds_offset]

#compare_enrich_main_filt_dt[,first_nts:=gsub("[[:digit:]]|[[:punct:]]|[c]", "", first)]

compare_enrich_main_filt_dt[,first_ref:=substr(first,6,6)]
compare_enrich_main_filt_dt[,second_ref:=substr(second,6,6)]
compare_enrich_main_filt_dt[,third_ref:=substr(third,6,6)]

compare_enrich_main_filt_dt[,first_alt:=substr(first,8,8)]
compare_enrich_main_filt_dt[,second_alt:=substr(second,8,8)]
compare_enrich_main_filt_dt[,third_alt:=substr(third,8,8)]

compare_enrich_main_filt_dt[,first_second_diff:=second_pos-first_pos]
compare_enrich_main_filt_dt[,first_third_diff:=third_pos-first_pos]

# Annotate the SNPs first
compare_enrich_main_filt_dt[is.na(second_alt), `:=`(
  POS=as.character(first_pos), REF=first_ref, ALT=first_alt)]

# Capture di-nt MNPs where mismatches are adjacent
compare_enrich_main_filt_dt[first_second_diff==1 & is.na(third_pos), `:=`(
  POS=as.character(first_pos), REF=paste0(first_ref, second_ref),
  ALT=paste0(first_alt, second_alt))]

# Capture di-nt MNPs where mismatches are 1 nt away
# Use the satmut_utils calls to determine the intervening REF base
compare_enrich_main_filt_dt[first_second_diff==2 & is.na(third_pos), middle_pos:=first_pos+1]

pos_match<- compare_enrich_main_filt_dt[first_second_diff==2 & is.na(third_pos), 
                                        match(middle_pos, compare_satmut_utils_postprocess_dt[,POS])]

compare_enrich_main_filt_dt[
  first_second_diff==2 & is.na(third_pos), 
  middle_ref:=compare_satmut_utils_postprocess_dt[pos_match, as.character(REF_NT)]]

# There are still some middle REF nts not identified
# compare_enrich_main_filt_dt[first_second_diff==2 & is.na(third_pos) & is.na(middle_ref)]

# Match the remaining missing reference nts for the middle position
# using the Enrich2 dataset
pos_match<- compare_enrich_main_filt_dt[
  first_second_diff==2 & is.na(third_pos) & is.na(middle_ref), 
  match(middle_pos, compare_enrich_main_filt_dt[,first_pos])]

compare_enrich_main_filt_dt[
  first_second_diff==2 & is.na(third_pos) & is.na(middle_ref), 
  middle_ref:=compare_enrich_main_filt_dt[pos_match, first_ref]]

# Now ready to annotate for di-nts with intervening REF nt
compare_enrich_main_filt_dt[first_second_diff==2 & is.na(third_pos), `:=`(
  POS=as.character(first_pos), REF=paste0(first_ref, middle_ref, second_ref),
  ALT=paste0(first_alt, middle_ref, second_alt))]
                            
# Capture tri-nt MNPs
compare_enrich_main_filt_dt[first_third_diff==2, `:=`(
  POS=as.character(first_pos), REF=paste0(first_ref, second_ref, third_ref),
  ALT=paste0(first_alt, second_alt, third_alt))]

# Finally capture haplotypes (variants with mismatch positions > span of 3)

# 2-mismatch haplotypes
compare_enrich_main_filt_dt[is.na(POS) & is.na(third_pos), `:=`(
  POS=paste(first_pos, second_pos, sep=","), REF=paste(first_ref, second_ref, sep=","),
  ALT=paste(first_alt, second_alt, sep=","))]

# 3-mismatch haplotypes
compare_enrich_main_filt_dt[is.na(POS) & !is.na(third_pos), `:=`(
  POS=paste(first_pos, second_pos, third_pos, sep=","), REF=paste(first_ref, second_ref, third_ref, sep=","),
  ALT=paste(first_alt, second_alt, third_alt, sep=","))]

# Finally create the VAR_ID and merge with the satmut_utils, dms_tools2 merged dataset
compare_enrich_main_filt_dt[,CAO.z:=count]
compare_enrich_main_filt_dt[,`#CHROM`:="CBS_pAG415GAL"]
compare_enrich_main_filt_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

# Remove changes with N
compare_enrich_main_filt_noN_dt<- compare_enrich_main_filt_dt[!grepl("N", ALT)]

# Determine if there any duplicated VAR_IDs at this point
compare_enrich_main_filt_noN_varid_counts_dt<- compare_enrich_main_filt_noN_dt[,.N,by=.(VAR_ID)]
#compare_enrich_main_filt_noN_varid_counts_dt[,unique(N)]

# Merge with truth dt
compare_enrich_main_filt_noN_merge_dt<- merge(
  CBS_pAG415GAL_tile6_sim_truth_dt[,VAR_ID, CAO],
  compare_enrich_main_filt_noN_dt[,.(VAR_ID, CAO.z)],
  by="VAR_ID", all=TRUE)


# Merge the Enrich2 counts with the satmut_utils-dms_tools2 dataset
# CAO.x is satmut_utils
# CAO.y is dms_tools2
# CAO.z is Enrich2
compare_satmut_utils_merge_dmstools_enrich_dt<- merge(
  compare_satmut_utils_merge_dmstools_dt,
  compare_enrich_main_filt_noN_dt[,.(VAR_ID, CAO.z)],
  by="VAR_ID", all=TRUE)

# Filter to those variants that were called in Enrich2

# FIXED bug from manuscript: ensure to plot variants called in satmut_utils which are not in Enrich2
# only one variant was accrued
compare_satmut_utils_merge_dmstools_enrich_filt_dt<- compare_satmut_utils_merge_dmstools_enrich_dt[
  !is.na(CAO.z) | CAO.x > 0]

# compare_satmut_utils_merge_dmstools_enrich_filt_dt[,sum(!(is.na(Truth_CAO) | Truth_CAO==0))] confirms all truth variants still present

# Recompute the variant TYPE as now we have Enrich2 variants not in the other sets
compare_satmut_utils_merge_dmstools_enrich_filt_dt[,enrich_REF:=sapply(strsplit(VAR_ID, ":"), "[", 3)]
compare_satmut_utils_merge_dmstools_enrich_filt_dt[,enrich_ALT:=sapply(strsplit(VAR_ID, ":"), "[", 4)]

compare_satmut_utils_merge_dmstools_enrich_filt_dt[
  ,TYPE:=get_var_type(enrich_REF, enrich_ALT), by=seq_len(nrow(
    compare_satmut_utils_merge_dmstools_enrich_filt_dt))]

compare_satmut_utils_merge_dmstools_enrich_filt_dt[,Truth:="T"]

# UPDATE from manuscript: include Truth CAO == 0 for Suppl. Figure 2C
compare_satmut_utils_merge_dmstools_enrich_filt_dt[is.na(Truth_CAO) | Truth_CAO==0, Truth:="F"]
compare_satmut_utils_merge_dmstools_enrich_filt_dt[is.na(CAO.x), CAO.x:=0]
compare_satmut_utils_merge_dmstools_enrich_filt_dt[is.na(CAO.y), CAO.y:=0]
compare_satmut_utils_merge_dmstools_enrich_filt_dt[is.na(CAO.z), CAO.z:=0]

# Compare variant frequencies between satmut_utils and Enrich2
compare_satmut_utils_vs_dmstools_enrich_gg<- ggplot(
  data=compare_satmut_utils_merge_dmstools_enrich_filt_dt[
    TYPE%in%c("SNP", "di_nt_MNP", "tri_nt_MNP")], 
  aes(x=log10(CAO.x+1), y=log10(CAO.z+1), col=Truth))

compare_satmut_utils_vs_dmstools_enrich_gg+
  facet_wrap(~TYPE)+
  scale_color_simpsons()+
  geom_point(alpha=0.7)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(0,4.5), ylim=c(0,4.5))+
  labs(x="satmut_utils log10(counts + 1)", 
       y="Enrich2 log10(counts + 1)")+
  theme_cowplot()


# Compare dms_tools, Enrich2, and satmut_utils in terms of variant frequency accuracy
# UPDATE from manuscript: include Truth CAO == 0 due to changes for Supp. Fig. 2
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_dt<- 
  compare_satmut_utils_merge_dmstools_enrich_filt_dt[!(is.na(Truth_CAO) | Truth_CAO==0),]

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_dt[,satmut_utils:=CAO.x-Truth_CAO]
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_dt[,dms_tools2:=CAO.y-Truth_CAO]
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_dt[,Enrich2:=CAO.z-Truth_CAO]

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt<- melt(
  data=compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_dt, 
  id.vars=c("VAR_ID", "TYPE", "Truth_CAO"),
  measure.vars=c("satmut_utils", "dms_tools2", "Enrich2"), 
  variable.name="Software", value.name="CAO_diff")

# Some of the diffs can be quite large so we want to log10 transform
# However to do this we must first take the absolute value and add a pseudocount
# To facilitate visualization, after this transformation multiple the value by -1
# for those variants with diff < 0
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[
  ,`:=`(CAO_diff_abs=abs(CAO_diff), log10_CAO_diff_abs=log10(abs(CAO_diff)+1))]

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[
  CAO_diff>=0, log10_CAO_diff_abs_mirror:=log10_CAO_diff_abs]

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[
  CAO_diff<0, log10_CAO_diff_abs_mirror:=-1*log10_CAO_diff_abs]

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[
  ,Accuracy:=CAO_diff/Truth_CAO]

# Determine accuracy of variant frequencies for each variant type
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[
  ,Software:=factor(Software, levels=software_levels)]

# Reconstruct the observed count from the truth and difference column
# to enable log(obs/exp) plot
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[
  ,Observed:=Truth_CAO+CAO_diff]

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[
  ,log_obs_exp:=log(Observed/Truth_CAO)]

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_snp_gg<- ggplot(
  data=compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt,
  aes(x=TYPE, y=Accuracy, col=Software))

# geom_point(alpha=0.7, position=position_jitterdodge(jitter.width=0.35))+
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_snp_gg+
  coord_cartesian(ylim=c(0, 15))+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(position="dodge")+
  labs(x="", y="(Observed - expected) / expected")+
  scale_color_manual(values=npg_cols[c(3,2,4)])+
  theme_cowplot()

# Plot log(obs/exp) instead
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_snp_gg<- ggplot(
  data=compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt,
  aes(x=TYPE, y=log_obs_exp, col=Software))

# geom_point(alpha=0.7, position=position_jitterdodge(jitter.width=0.35))+
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_snp_gg+
  coord_cartesian(ylim=c(0, 3))+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(position="dodge")+
  labs(x="", y="log(Observed/expected)")+
  scale_color_manual(values=npg_cols[c(3,2,4)])+
  theme_cowplot()


# MNPs plot separate from SNPs because scale varies greatly
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_mnp_gg<- ggplot(
  data=compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[TYPE!="SNP"],
  aes(x=TYPE, y=Accuracy, col=Software))

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_mnp_gg+
  facet_wrap(~TYPE, scales="free")+
  coord_cartesian(ylim=c(-.2, .02))+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(position="dodge")+
  labs(x="", y="(Observed - expected) / expected")+
  scale_color_manual(values=npg_cols[c(3,2,4)])+
  theme_cowplot()

# Plot log(obs/exp)
compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_mnp_gg<- ggplot(
  data=compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_melt_dt[TYPE!="SNP"],
  aes(x=TYPE, y=log_obs_exp, col=Software))

compare_satmut_utils_merge_dmstools_enrich_filt_count_acc_mnp_gg+
  facet_wrap(~TYPE, scales="free")+
  coord_cartesian(ylim=c(-.2, .02))+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(position="dodge")+
  labs(x="", y="log(Observed/expected)")+
  scale_color_manual(values=npg_cols[c(3,2,4)])+
  theme_cowplot()


# Concatenate changes for DiMSum doubles and fix the protein position
compare_dimsum_single_dt[,AA_POS:=Pos+tile6_aa_offset]
compare_dimsum_single_dt[,AA_CHANGE:=paste0("p.", WT_AA, AA_POS, Mut)]

compare_dimsum_double_dt[,AA_POS1:=Pos1+tile6_aa_offset]
compare_dimsum_double_dt[,AA_POS2:=Pos2+tile6_aa_offset]

compare_dimsum_double_dt[,AA_CHANGE:=paste0(
  "p.", WT_AA1, AA_POS1, Mut1, ",", "p.", WT_AA2, AA_POS2, Mut2)]

compare_dimsum_dt<- rbind(compare_dimsum_single_dt[,.(AA_CHANGE, mean_count)], 
                          compare_dimsum_double_dt[,.(AA_CHANGE, mean_count)])

# Sanity check to make sure there are no duplicated AA changes
compare_dimsum_aa_change_counts_dt<- compare_dimsum_dt[,.N,by=.(AA_CHANGE)]
# compare_dimsum_aa_change_counts_dt[,max(N)]

# Annotate the satmut_utils dataset with truth labels
compare_satmut_utils_merge_dt[,Truth:="T"]

# UPDATE from manuscript: include Truth CAO == 0 due to change for Suppl. Figure 2A
compare_satmut_utils_merge_dt[is.na(Truth_CAO) | Truth_CAO==0, Truth:="F"]
compare_satmut_utils_merge_dt[,Truth:=factor(Truth, levels=c("F", "T"))]

# UPDATE labeling to keep consistent and reassign di-codon variants with True,True mutagenesis signature as FP
# will increase precision very slightly
compare_satmut_utils_merge_dt[,satmut_utils_called:="F"]
compare_satmut_utils_merge_dt[MATCHES_MUT_SIG=="True", satmut_utils_called:="T"]
compare_satmut_utils_merge_dt[,satmut_utils_called:=factor(satmut_utils_called, levels=c("F", "T"))]

# Merge with the satmut_utils dataset to provide protein annotations
compare_satmut_utils_merge_summ_dt<- compare_satmut_utils_merge_dt[
  ,.(CAO=mean(CAO, na.rm=TRUE), Truth_CAO=mean(Truth_CAO, na.rm=TRUE)), 
  by=.(AA_CHANGE, MATCHES_MUT_SIG)]

compare_satmut_utils_merge_summ_dt[,Truth:="T"]
compare_satmut_utils_merge_summ_dt[Truth_CAO==0, Truth:="F"]
compare_satmut_utils_merge_summ_dt[,Truth:=factor(Truth, levels=c("F", "T"))]
    
compare_dimsum_aa_change_merge_dt<- merge(
  compare_satmut_utils_merge_summ_dt, compare_dimsum_dt[mean_count>1,],
  by="AA_CHANGE", all=TRUE)

# Remove satmut_utils calls not in DiMSum calls for comparison of DiMSum merged dataset
# to truth
compare_dimsum_aa_change_merge_filt_dt<- compare_dimsum_aa_change_merge_dt[
  !(is.na(mean_count) & Truth=="F")]


## Compare performance

# satmut_utils-nt default
# Determine satmut_utils-nt performance on tile 6 with -m 2 -q 30 -s NNK, NO MUT_SIG filter
compare_satmut_utils_postprocess_cont_table<- create_cont_table(
  compare_satmut_utils_merge_dt[,Truth],
  compare_satmut_utils_merge_dt[,satmut_utils_called])

compare_satmut_utils_sensitivity<- get_sensitivity(compare_satmut_utils_postprocess_cont_table)
compare_satmut_utils_precision<- get_precision(compare_satmut_utils_postprocess_cont_table)

# satmut_utils-nt stringent default
# FIXED labeling and do not apply MATCHES_MUT_SIG filter for satmut_utils
compare_satmut_utils_merge_dt[,satmut_utils_called:="F"]
compare_satmut_utils_merge_dt[CAO.y >= count_thresh, satmut_utils_called:="T"]
compare_satmut_utils_merge_dt[,satmut_utils_called:=factor(satmut_utils_called, levels=c("F", "T"))]

compare_satmut_utils_postprocess_stringent_cont_table<- create_cont_table(
  compare_satmut_utils_merge_dt[,Truth],
  compare_satmut_utils_merge_dt[,satmut_utils_called])

compare_satmut_utils_stringent_sensitivity<- get_sensitivity(
  compare_satmut_utils_postprocess_stringent_cont_table)

compare_satmut_utils_stringent_precision<- get_precision(
  compare_satmut_utils_postprocess_stringent_cont_table)

# Determine how TPs and FPs distribute by counts
compare_satmut_utils_count_tp_fp_gg<- ggplot(
  data=compare_satmut_utils_merge_dt, 
  aes(x=log10(CAO.y+1), col=Truth))

compare_satmut_utils_count_tp_fp_gg+
  geom_density()+
  scale_color_simpsons()+
  labs(x="Observed log10(counts + 1)")+
  theme_cowplot()


# Get AA level for comparison to DiMSum with MUTSIG filter
# satmut_utils-aa default
# UPDATE labeling and do not apply MATCHES_MUT_SIG filter for satmut_utils
compare_satmut_utils_merge_summ_dt[,satmut_utils_called:="T"]
#compare_satmut_utils_merge_summ_dt[MATCHES_MUT_SIG=="True", satmut_utils_called:="T"]
compare_satmut_utils_merge_summ_dt[,satmut_utils_called:=factor(satmut_utils_called, levels=c("F", "T"))]

compare_satmut_utils_merge_filt_summ_cont_table<- create_cont_table(
  compare_satmut_utils_merge_summ_dt[,Truth],
  compare_satmut_utils_merge_summ_dt[,satmut_utils_called])

compare_satmut_utils_aa_sensitivity<- get_sensitivity(compare_satmut_utils_merge_filt_summ_cont_table)
compare_satmut_utils_aa_precision<- get_precision(compare_satmut_utils_merge_filt_summ_cont_table)

# satmut_utils-aa stringent
# UPDATE from manuscript: do not apply MATCHES_MUT_SIG filter for satmut_utils
compare_satmut_utils_merge_summ_dt[,satmut_utils_called:="F"]
compare_satmut_utils_merge_summ_dt[CAO >= count_thresh, satmut_utils_called:="T"]
compare_satmut_utils_merge_summ_dt[,satmut_utils_called:=factor(satmut_utils_called, levels=c("F", "T"))]

compare_satmut_utils_merge_summ_stringent_cont_table<- create_cont_table(
  compare_satmut_utils_merge_summ_dt[,Truth],
  compare_satmut_utils_merge_summ_dt[,satmut_utils_called])

compare_satmut_utils_aa_stringent_sensitivity<- get_sensitivity(
  compare_satmut_utils_merge_summ_stringent_cont_table)

compare_satmut_utils_aa_stringent_precision<- get_precision(
  compare_satmut_utils_merge_summ_stringent_cont_table)


# dms_tools2
# dms_tools2 default
# compare_satmut_utils_merge_dmstools_filt_dt[,min(CAO.y)] is 2
compare_satmut_utils_merge_dmstools_filt_dt[,dmstools2_called:=factor(
  rep("T", nrow(compare_satmut_utils_merge_dmstools_filt_dt)), levels=c("F", "T"))]

compare_satmut_utils_merge_dmstools_filt_cont_table<- create_cont_table(
  compare_satmut_utils_merge_dmstools_filt_dt[,Truth],
  compare_satmut_utils_merge_dmstools_filt_dt[,dmstools2_called])

compare_dms_tools2_sensitivity<- get_sensitivity(compare_satmut_utils_merge_dmstools_filt_cont_table)
compare_dms_tools2_precision<- get_precision(compare_satmut_utils_merge_dmstools_filt_cont_table)

# dms_tools2 stringent
compare_satmut_utils_merge_dmstools_filt_dt[CAO.y >= count_thresh, dmstools2_called:="T"]
compare_satmut_utils_merge_dmstools_filt_dt[CAO.y < count_thresh, dmstools2_called:="F"]
compare_satmut_utils_merge_dmstools_filt_dt[,dmstools2_called:=factor(dmstools2_called, levels=c("F", "T"))]

compare_satmut_utils_merge_dmstools_stringent_filt_cont_table<- create_cont_table(
  compare_satmut_utils_merge_dmstools_filt_dt[,Truth],
  compare_satmut_utils_merge_dmstools_filt_dt[,dmstools2_called])

compare_dms_tools2_stringent_sensitivity<- get_sensitivity(
  compare_satmut_utils_merge_dmstools_stringent_filt_cont_table)

compare_dms_tools2_stringent_precision<- get_precision(
  compare_satmut_utils_merge_dmstools_stringent_filt_cont_table)


# Enrich2
# Enrich2 default
compare_enrich_main_filt_noN_merge_dt[,Truth:="T"]
compare_enrich_main_filt_noN_merge_dt[is.na(CAO), Truth:="F"]
compare_enrich_main_filt_noN_merge_dt[,Truth:=factor(Truth, levels=c("F", "T"))]
compare_enrich_main_filt_noN_merge_dt[,enrich2_called:=factor(
  rep("T", nrow(compare_enrich_main_filt_noN_merge_dt)), levels=c("F", "T"))]

compare_enrich_main_filt_noN_merge_cont_table<- create_cont_table(
  compare_enrich_main_filt_noN_merge_dt[,Truth],
  compare_enrich_main_filt_noN_merge_dt[,enrich2_called])

compare_enrich2_sensitivity<- get_sensitivity(compare_enrich_main_filt_noN_merge_cont_table)
compare_enrich2_precision<- get_precision(compare_enrich_main_filt_noN_merge_cont_table)

# Enrich2 stringent
compare_enrich_main_filt_noN_merge_dt[CAO.z >= count_thresh, enrich2_called:="T"]
compare_enrich_main_filt_noN_merge_dt[CAO.z < count_thresh, enrich2_called:="F"]
compare_enrich_main_filt_noN_merge_dt[,enrich2_called:=factor(enrich2_called, levels=c("F", "T"))]

compare_enrich_main_filt_noN_merge_stringent_cont_table<- create_cont_table(
  compare_enrich_main_filt_noN_merge_dt[,Truth],
  compare_enrich_main_filt_noN_merge_dt[,enrich2_called])

compare_enrich2_stringent_sensitivity<- get_sensitivity(
  compare_enrich_main_filt_noN_merge_stringent_cont_table)

compare_enrich2_stringent_precision<- get_precision(
  compare_enrich_main_filt_noN_merge_stringent_cont_table)


# DiMSum
# DiMSum default
compare_dimsum_aa_change_merge_filt_dt[,Truth:="T"]

# UPDATE from manuscript: include Truth CAO == 0 due to change for Suppl. Figure 2
compare_dimsum_aa_change_merge_filt_dt[is.na(Truth_CAO) | Truth_CAO==0, Truth:="F"]
compare_dimsum_aa_change_merge_filt_dt[,Truth:=factor(Truth, levels=c("F", "T"))]
compare_dimsum_aa_change_merge_filt_dt[,dimsum_called:=factor(rep(
  "T", nrow(compare_dimsum_aa_change_merge_filt_dt)), levels=c("F", "T"))]

# compare_dimsum_aa_change_merge_filt_dt[Truth=="T",length(unique(AA_CHANGE))]
# Confirms we have all truth AA changes in the dataset

compare_dimsum_aa_change_merge_cont_table<- create_cont_table(
  compare_dimsum_aa_change_merge_filt_dt[,Truth],
  compare_dimsum_aa_change_merge_filt_dt[,dimsum_called])

compare_dimsum_aa_sensitivity<- get_sensitivity(compare_dimsum_aa_change_merge_cont_table)
compare_dimsum_aa_precision<- get_precision(compare_dimsum_aa_change_merge_cont_table)

# DiMSum stringent
compare_dimsum_aa_change_merge_filt_dt[mean_count >= count_thresh, dimsum_called:="T"]
compare_dimsum_aa_change_merge_filt_dt[mean_count < count_thresh, dimsum_called:="F"]
compare_dimsum_aa_change_merge_filt_dt[,dimsum_called:=factor(dimsum_called, levels=c("F", "T"))]

compare_dimsum_aa_change_merge_stringent_cont_table<- create_cont_table(
  compare_dimsum_aa_change_merge_filt_dt[,Truth],
  compare_dimsum_aa_change_merge_filt_dt[,dimsum_called])

compare_dimsum_aa_stringent_sensitivity<- get_sensitivity(
  compare_dimsum_aa_change_merge_stringent_cont_table)

compare_dimsum_aa_stringent_precision<- get_precision(
  compare_dimsum_aa_change_merge_stringent_cont_table)

# Create a master table of all results
compare_all_default_sensitivities<- c(
  "satmut_utils"=compare_satmut_utils_sensitivity, 
  "satmut_utils_aa"=compare_satmut_utils_aa_sensitivity, 
  "dms_tools2"=compare_dms_tools2_sensitivity,
  "enrich2"=compare_enrich2_sensitivity,
  "dimsum_aa"=compare_dimsum_aa_sensitivity)

compare_all_default_precisions<- c(
  "satmut_utils"=compare_satmut_utils_precision, 
  "satmut_utils_aa"=compare_satmut_utils_aa_precision, 
  "dms_tools2"=compare_dms_tools2_precision,
  "enrich2"=compare_enrich2_precision,
  "dimsum_aa"=compare_dimsum_aa_precision)

compare_all_stringent_sensitivities<- c(
  "satmut_utils"=compare_satmut_utils_stringent_sensitivity, 
  "satmut_utils_aa"=compare_satmut_utils_aa_stringent_sensitivity, 
  "dms_tools2"=compare_dms_tools2_stringent_sensitivity,
  "enrich2"=compare_enrich2_stringent_sensitivity,
  "dimsum_aa"=compare_dimsum_aa_stringent_sensitivity)

compare_all_stringent_precisions<- c(
  "satmut_utils"=compare_satmut_utils_stringent_precision, 
  "satmut_utils_aa"=compare_satmut_utils_aa_stringent_precision, 
  "dms_tools2"=compare_dms_tools2_stringent_precision,
  "enrich2"=compare_enrich2_stringent_precision,
  "dimsum_aa"=compare_dimsum_aa_stringent_precision)

compare_all_perf_dt<- data.table(
  Recall=c(compare_all_default_sensitivities, compare_all_stringent_sensitivities),
  Precision=c(compare_all_default_precisions, compare_all_stringent_precisions))

compare_all_perf_dt[,Fscore:=get_fscore(Precision, Recall), 
                       by=seq_len(nrow(compare_all_perf_dt))]

compare_all_perf_dt[,Software:=factor(
  rep(c("satmut_utils-nt", "satmut_utils-aa", "dms_tools2-nt", "Enrich2-nt", "DiMSum-aa"), 2),
  levels=c("DiMSum-aa", "dms_tools2-nt", "Enrich2-nt", "satmut_utils-nt", "satmut_utils-aa"))]

compare_all_perf_dt[,Measure:=factor(
  rep(c("Nucleotide", "Amino_acid", "Nucleotide", "Nucleotide", "Amino_acid"), 2),
  levels=c("Nucleotide", "Amino_acid"))]

compare_all_perf_dt[,Filter:=factor(
  rep(c("Default", "Stringent"), each=5),
  levels=c("Default", "Stringent"))]

compare_all_perf_gg<- ggplot(
  data=compare_all_perf_dt, 
  aes(x=Recall, y=Precision, col=Software, shape=Filter))

compare_all_perf_gg+
  geom_point(size=3)+
  scale_x_continuous(breaks=seq(0.9,1,0.02))+
  scale_y_continuous(breaks=seq(0,1,0.1))+
  coord_cartesian(xlim=c(0.9,1), ylim=c(0,1))+
  scale_color_npg()+
  theme_cowplot()
  
```

## Error modeling with sim on real reads

Datasets contain thousands of CBS variants at variable frequencies. Frequencies are estimated from calls in a true positive library.

### Roth lab- CBS WT, yeast DNA-sequencing, NextSeq 500

```{r CBS complete CDS simulation- Roth WT nonselected alignments from Song and Weile et al 2020}

roth_ec_dir<- paste(base_dir, "error_correction", "roth_cbs", sep="/")
roth_ec_truth<- paste(roth_ec_dir, "CBS_pAG415GAL_train.truth.vcf", sep="/")
roth_ec_train<- paste(roth_ec_dir, "CBS_pAG415GAL_train.var.cand.vcf.summary.txt", sep="/")

roth_ec_truth_dt<- fread(roth_ec_truth, skip=18, sep="\t")
roth_ec_train_dt<- fread(roth_ec_train, sep="\t")

roth_ec_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]
roth_ec_train_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

# Look at frequency distributions for true positives and false negatives
roth_ec_truth_dt[,CAO:=as.integer(sapply(strsplit(
  sapply(strsplit(INFO, ";"), "[", 1), "=", fixed=T), "[", 2))]

roth_ec_truth_dt[,CAF:=as.numeric(sapply(strsplit(
  sapply(strsplit(INFO, ";"), "[", 2), "=", fixed=T), "[", 2))]

roth_ec_truth_dt[,log10_CAF:=log10(CAF)]

roth_ec_truth_filt_dt<- roth_ec_truth_dt[CAO>0]

# Determine if any calls were missed
roth_ec_truth_filt_dt[VAR_ID%in%roth_ec_train_dt[,unique(VAR_ID)], Called:=TRUE]
roth_ec_truth_filt_dt[is.na(Called), Called:=FALSE]

roth_ec_train_sensitivity<- roth_ec_train_dt[
  ,sum(roth_ec_truth_filt_dt[,VAR_ID]%in%unique(VAR_ID))/nrow(roth_ec_truth_filt_dt)]

# Plot frequency accuracy for each variant type
roth_ec_train_postproc_dt<- postprocess_vcf_summary_dt(
  in_dt=roth_ec_train_dt, mapping_dt=NULL, tile_positions_dt=NULL, 
  key_features=c("VAR_ID"))

roth_ec_train_merge_dt<- merge(roth_ec_truth_filt_dt[,.(VAR_ID, CAO, CAF, log10_CAF)], 
                               roth_ec_train_postproc_dt, by=c("VAR_ID"))

roth_ec_train_merge_dt[,log10_CAF_diff:=log10_CAF.y-log10_CAF.x]

# roth_ec_train_merge_dt[TYPE!="SNP" & !is.infinite(log10_CAF_diff), summary(log10_CAF_diff)]
roth_ec_train_merge_dt[,TYPE:=factor(TYPE, levels=c("SNP", "di_nt_MNP", "tri_nt_MNP"))]

roth_ec_freq_acc_gg<- ggplot(
  data=roth_ec_train_merge_dt, 
  aes(x=log10_CAF.x, y=log10_CAF.y))

roth_ec_freq_acc_gg+
  facet_wrap(~TYPE)+
  geom_point(alpha=0.7)+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7,-2), ylim=c(-7,-2))+
  labs(x="Expected log10 frequency", 
       y="Observed log10 frequency")+
  theme_cowplot()

# Determine if the MNP variants off the diagonal are due to partial overlap
# with primers, which is expected to cause deviation for MNPs with 1 or more 
# bases intersecting a primer region. The min depth is taken for the denominator
# of the frequency calculation.- Yes

# Look at counts
roth_ec_freq_acc_gg<- ggplot(
  data=roth_ec_train_merge_dt, 
  aes(x=log10(CAO.x + 1), y=log10(CAO.y + 1)))

roth_ec_freq_acc_gg+
  facet_wrap(~TYPE)+
  geom_point(alpha=0.7)+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_smooth(method="lm")+
  labs(x="Expected log10 (count + 1)", 
       y="Observed log10 (count + 1)")+
  theme_cowplot()

```

```{r Figure 2A: error signatures across platforms}

# CBS WT nonselect, expressed in yeast, Nextseq 500, Roth lab
roth_ec_dir<- paste(base_dir, "error_correction", "roth_cbs", sep="/")
roth_wt<- paste(roth_ec_dir, "wt_nonselect_r.var.cand.vcf.summary.txt", sep="/")
roth_wt_dt<- fread(roth_wt, sep="\t", header=TRUE)

# CBS WT cDNA from HEK293T, various platforms, Cenik lab
cenik_ec_dir<- paste(base_dir, "error_correction", "cenik_cbs", sep="/")
cenik_cbs10_wt<- paste(cenik_ec_dir, "CBS1-10_S10_L001_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs21_wt<- paste(cenik_ec_dir, "CBS1_21_S10_L001_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs31_wt<- paste(cenik_ec_dir, "CBS1_31_S10_L006_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs34_wt<- paste(cenik_ec_dir, "CBS1_34_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs49_wt<- paste(cenik_ec_dir, "CBS1_49_R.var.cand.vcf.summary.txt", sep="/")

cenik_cbs10_wt_dt<- fread(cenik_cbs10_wt, sep="\t", header=TRUE)
cenik_cbs21_wt_dt<- fread(cenik_cbs21_wt, sep="\t", header=TRUE)
cenik_cbs31_wt_dt<- fread(cenik_cbs31_wt, sep="\t", header=TRUE)
cenik_cbs34_wt_dt<- fread(cenik_cbs34_wt, sep="\t", header=TRUE)
cenik_cbs49_wt_dt<- fread(cenik_cbs49_wt, sep="\t", header=TRUE)

# CBS WT amplified from plasmid, HiSeqX, Cenik lab
cbs_02dec2021_dir<- paste(base_dir, "CBS_02DEC2021", sep="/")
cenik_cbs75_wt<- paste(cbs_02dec2021_dir, "CBS1_75_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs82_wt<- paste(cbs_02dec2021_dir, "CBS1_82_R.var.cand.vcf.summary.txt", sep="/")

cenik_cbs75_wt_dt<- fread(cenik_cbs75_wt, sep="\t", header=TRUE)
cenik_cbs82_wt_dt<- fread(cenik_cbs82_wt, sep="\t", header=TRUE)

# Combine all datasets and annotate library and seq platform
cenik_roth_ec_wt_dt<- rbind(
  roth_wt_dt, cenik_cbs10_wt_dt, cenik_cbs21_wt_dt, 
  cenik_cbs31_wt_dt, cenik_cbs34_wt_dt, cenik_cbs49_wt_dt, 
  cenik_cbs75_wt_dt, cenik_cbs82_wt_dt)

cenik_roth_ec_wt_dt[,Platform:=factor(
  rep(c("Nextseq 500", "MiSeq rep1", "MiSeq rep2", "HiSeq 4000", "NovaSeq 6000", "HiSeq 2500", "HiSeq X lib1", "HiSeq X lib2"), 
      c(nrow(roth_wt_dt), nrow(cenik_cbs10_wt_dt), nrow(cenik_cbs21_wt_dt), 
        nrow(cenik_cbs31_wt_dt), nrow(cenik_cbs34_wt_dt), nrow(cenik_cbs49_wt_dt), 
        nrow(cenik_cbs75_wt_dt), nrow(cenik_cbs82_wt_dt))),
  levels=c(c("Nextseq 500", "NovaSeq 6000", "HiSeq 2500", "HiSeq 4000",  
             "HiSeq X lib1", "HiSeq X lib2", "MiSeq rep1", "MiSeq rep2")))]

cenik_roth_ec_wt_dt[Platform=="Nextseq 500", Library:="A"]
cenik_roth_ec_wt_dt[Platform=="MiSeq rep1", Library:="B"]
cenik_roth_ec_wt_dt[Platform%in%c("MiSeq rep2", "HiSeq 4000", "NovaSeq 6000", "HiSeq 2500"), Library:="C"]
cenik_roth_ec_wt_dt[Platform=="HiSeq X lib1", Library:="D"]
cenik_roth_ec_wt_dt[Platform=="HiSeq X lib2", Library:="E"]

cenik_roth_ec_wt_dt[,Library:=factor(Library, levels=c("A", "B", "C", "D", "E"))]

cenik_roth_ec_wt_dt[,TYPE:=get_var_type(REF, ALT), 
                    by=seq_len(nrow(cenik_roth_ec_wt_dt))]

cenik_roth_ec_wt_dt[,SUBST:=as.factor(paste(REF_NT, ALT_NT, sep=":"))]

cenik_roth_ec_wt_dt[SUBST%in%c("A:C", "T:G"), SUBST:="A.C_T.G"]
cenik_roth_ec_wt_dt[SUBST%in%c("A:G", "T:C"), SUBST:="A.G_T.C"]
cenik_roth_ec_wt_dt[SUBST%in%c("A:T", "T:A"), SUBST:="A.T_T.A"]
cenik_roth_ec_wt_dt[SUBST%in%c("C:A", "G:T"), SUBST:="C.A_G.T"]
cenik_roth_ec_wt_dt[SUBST%in%c("C:G", "G:C"), SUBST:="C.G_G.C"]
cenik_roth_ec_wt_dt[SUBST%in%c("C:T", "G:A"), SUBST:="C.T_G.A"]
cenik_roth_ec_wt_dt[,SUBST:=as.factor(SUBST)]
  
cenik_roth_ec_wt_subst_counts_dt<- cenik_roth_ec_wt_dt[,.N,by=.(Library, Platform, SUBST)]
cenik_roth_ec_wt_subst_counts_dt[,`:=`(prop=N/sum(N), log_prop=log(N/sum(N))), by=.(Library, Platform)]

cenik_roth_ec_wt_subst_counts_cast_dt<- dcast(
  data=cenik_roth_ec_wt_subst_counts_dt, 
  Library+Platform~SUBST, value.var="log_prop")

cenik_roth_ec_wt_subst_counts_cast_pca_in<- as.matrix(
  cenik_roth_ec_wt_subst_counts_cast_dt[,-c("Library", "Platform")])

rownames(cenik_roth_ec_wt_subst_counts_cast_pca_in)<-
  cenik_roth_ec_wt_subst_counts_cast_dt[,Platform]

cenik_roth_ec_wt_subst_counts_cast_prcomp<- prcomp(
  cenik_roth_ec_wt_subst_counts_cast_pca_in, scale=TRUE)

fviz_eig(cenik_roth_ec_wt_subst_counts_cast_prcomp)

cenik_roth_ec_wt_subst_counts_cast_prcomp_gg<- 
  fviz_pca_ind(cenik_roth_ec_wt_subst_counts_cast_prcomp,
               col.ind=cenik_roth_ec_wt_subst_counts_cast_dt[,Library],
               palette="aaas", repel=TRUE)

cenik_roth_ec_wt_subst_counts_cast_prcomp_gg+
  theme_cowplot()+
  coord_cartesian(xlim=c(-2,6), ylim=c(-2,6))

# While the results make some sense, N<<p so PCA or other
# factorization methods may not be reliable


# All data no summarization
cenik_roth_ec_wt_subst_counts_subst_gg<- ggplot(
  data=cenik_roth_ec_wt_subst_counts_dt, 
  aes(x=Platform, col=SUBST, shape=Library, y=prop))

cenik_roth_ec_wt_subst_counts_subst_gg+
  geom_point(linewidth=3)+
  expand_limits(y=0)+
  scale_color_jama()+
  labs(x="", col="Substitution", 
       y="Proportion of errors")+
  theme_cowplot()+
  theme(axis.text.x=element_text(angle=90))

# Summarize across libraries, platforms, labs, etc.
cenik_roth_ec_wt_subst_counts_summ_dt<- 
  cenik_roth_ec_wt_subst_counts_dt[
    ,.(Mean_prop=mean(prop), SD_prop=sd(prop)/sqrt(.N)),
    by=.(SUBST)]

cenik_roth_ec_wt_subst_counts_subst_gg<- ggplot(
  data=cenik_roth_ec_wt_subst_counts_summ_dt, 
  aes(x=reorder(SUBST, Mean_prop), y=Mean_prop, 
      ymax=Mean_prop+SD_prop, 
      ymin=Mean_prop-SD_prop))

cenik_roth_ec_wt_subst_counts_subst_gg+
  geom_errorbar(width=0.2, col="grey")+
  geom_point(linewidth=3, col="steelblue")+
  labs(x="Substitution", 
       y="Mean proportion +/- SE")+
  theme_cowplot()


```

```{r Figure 2B: Roth WT nonselected alignments from Song and Weile et al 2020- error correction models}

count_thresh<- 2
roth_ec_dir<- paste(base_dir, "error_correction", "roth_cbs", sep="/")
roth_ec_truth<- paste(roth_ec_dir, "CBS_pAG415GAL_train.truth.vcf", sep="/")
roth_ec<- paste(roth_ec_dir, "CBS_pAG415GAL_train.var.cand.vcf.summary.txt", sep="/")

roth_ec_truth_dt<- fread(roth_ec_truth, skip=18, sep="\t")
roth_ec_truth_dt<- roth_ec_truth_dt[!grepl("CAO=0;", INFO, fixed=TRUE)]

roth_ec_dt<- fread(roth_ec, sep="\t")

# Remove rows where any of the features are NA
# For amplicon data, we only expect data from R1(+)/R2(-) due to primer adapter tailing
# Any variants supported only from pairs on the other sample strand should be discarded
# as they contain NAs for the relevant quality features
roth_ec_filt_dt<- roth_ec_dt[!is.na(R1_PLUS_MED_RP),]

# Add a sample as this is required by model prep function
roth_ec_filt_dt[,Sample:="Roth_ec_wt_nonselect"]

# Get count filter-based performance
roth_ec_filt_nosingletons_dt<- roth_ec_filt_dt[CAO>=count_thresh]
roth_ec_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]
roth_ec_filt_nosingletons_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

roth_ec_filt_nosingletons_dt[,Truth:=FALSE]
roth_ec_filt_nosingletons_dt[VAR_ID%in%roth_ec_truth_dt[,VAR_ID], Truth:=TRUE]
roth_ec_filt_nosingletons_dt[,Truth:=factor(Truth, levels=cont_table_factor_levels)]

roth_ec_filt_nosingletons_dt[,satmut_utils_called:=factor(
  rep(TRUE, nrow(roth_ec_filt_nosingletons_dt)), levels=cont_table_factor_levels)]

roth_ec_filt_cont_table<- create_cont_table(
  roth_ec_filt_nosingletons_dt[,Truth], 
  roth_ec_filt_nosingletons_dt[,satmut_utils_called])

roth_ec_filt_prec<- get_precision(roth_ec_filt_cont_table)
roth_ec_filt_recall<- get_sensitivity(roth_ec_filt_cont_table)

# Compare different supervised binary classifiers
# Nonparametric likely to perform better given features, sharp contours

modeling_features<- c(
  "SUBST", "UP_REF_NT", "DOWN_REF_NT", 
  "TYPE", "log10_CAF", "MATCHES_MUT_SIG",
  "R1_PLUS_MED_RP", "R2_MINUS_MED_RP", 
  "R1_PLUS_MED_BQ", "R2_MINUS_MED_BQ", 
  "R1_PLUS_MED_NM", "R2_MINUS_MED_NM")

# Determine if any features should be dropped to protect against collinearity
#roth_ec_filt_prep_dt[,cor(R1_PLUS_MED_NM, R2_MINUS_MED_NM, method="spearman")]
#roth_ec_filt_prep_dt[,cor(R1_PLUS_MED_RP, R2_MINUS_MED_RP, method="spearman")]

# Because the NM and RP are collinear between R1 and R2, just select the stat for one read
modeling_features_no_collinear<- modeling_features[
  !grepl("R1_PLUS_MED_RP|R1_PLUS_MED_NM", modeling_features)]

# Prep the dataset for modeling
roth_ec_filt_prep_input_dt<- prep_train_dt(
  train_dt=roth_ec_filt_dt, truth_dt=roth_ec_truth_dt, 
  modeling_features=modeling_features_no_collinear)

# Get indices for 10 mutually exclusive folds
roth_ec_filt_prep_input_kfolds<- create_kfolds(roth_ec_filt_prep_input_dt)

roth_ec_filt_prep_input_test_dt<- roth_ec_filt_prep_input_dt[
  roth_ec_filt_prep_input_kfolds[[1]]]

roth_ec_filt_prep_input_train_dt<- roth_ec_filt_prep_input_dt[
  !roth_ec_filt_prep_input_kfolds[[1]]]

# Run 10-fold nested cross validation for performance comparison of models
roth_ec_filt_prep_input_elasticnet_cv_res_dt<- run_nested_cv(
  roth_ec_filt_prep_input_train_dt, model="elasticnet")

roth_ec_filt_prep_input_knn_cv_res_dt<- run_nested_cv(
  roth_ec_filt_prep_input_train_dt, model="knn")

roth_ec_filt_prep_input_rf_cv_res_dt<- run_nested_cv(
  roth_ec_filt_prep_input_train_dt, model="rf")

roth_ec_filt_prep_input_gbm_cv_res_dt<- run_nested_cv(
  roth_ec_filt_prep_input_train_dt, model="gbm")

roth_ec_filt_prep_input_svm_cv_res_dt<- run_nested_cv(
  roth_ec_filt_prep_input_train_dt, model="svm")

roth_ec_filt_prep_input_all_perf_dt<- as.data.table(rbind(
  roth_ec_filt_prep_input_elasticnet_cv_res_dt[[1]], 
  roth_ec_filt_prep_input_knn_cv_res_dt[[1]], 
  roth_ec_filt_prep_input_rf_cv_res_dt[[1]],
  roth_ec_filt_prep_input_gbm_cv_res_dt[[1]],
  roth_ec_filt_prep_input_svm_cv_res_dt[[1]]))

roth_ec_filt_prep_input_all_perf_dt[
  ,Model:=as.factor(rep(c("GLM-elasticnet", "kNN", "RF", "GBM", "SVC"), each=10))]

roth_ec_filt_prep_input_all_perf_melt_dt<- melt(
  data=roth_ec_filt_prep_input_all_perf_dt, 
  id.vars=c("Model"), value.name="Value", 
  variable.name="Measure")

roth_ec_filt_prep_input_all_perf_gg<- ggplot(
  data=roth_ec_filt_prep_input_all_perf_melt_dt, 
  aes(x=Model, y=Value, col=Measure))

roth_ec_filt_prep_input_all_perf_gg+
  geom_boxplot(position="dodge")+
  scale_color_simpsons()+
  scale_y_continuous(breaks=seq(0.9,1,.02))+
  coord_cartesian(ylim=c(0.9,1))+
  labs(y="10-fold CV accuracy")+
  theme_cowplot()

# Combine with count-based performance
roth_ec_filt_acc<- get_accuracy(roth_ec_filt_cont_table)
roth_ec_filt_recall<- get_sensitivity(roth_ec_filt_cont_table)
roth_ec_filt_prec<- get_precision(roth_ec_filt_cont_table)

roth_ec_filt_prep_ct_dt<- data.table(
  Model="Count", Precision=roth_ec_filt_prec, 
  Recall=roth_ec_filt_recall, Accuracy=roth_ec_filt_acc)

roth_ec_filt_prep_input_all_perf_comb_dt<- rbind(
  roth_ec_filt_prep_input_all_perf_dt,
  roth_ec_filt_prep_ct_dt)

# Plot performance of models compared to naive count filtering
roth_ec_filt_prep_input_all_perf_comb_melt_dt<- melt(
  data=roth_ec_filt_prep_input_all_perf_comb_dt, 
  id.vars=c("Model"), value.name="Value", 
  variable.name="Measure")

roth_ec_filt_prep_input_all_perf_comb_melt_summ_dt<- 
  roth_ec_filt_prep_input_all_perf_comb_melt_dt[
    ,.(Mean=mean(Value), SD=sd(Value)), 
    by=.(Measure, Model)]

roth_ec_filt_prep_input_all_perf_comb_melt_summ_cast_dt<- 
  dcast(data=roth_ec_filt_prep_input_all_perf_comb_melt_summ_dt,
        Model~Measure, value.var=c("Mean", "SD"))

roth_ec_filt_prep_input_all_perf_comb_melt_summ_cast_dt[
  is.na(roth_ec_filt_prep_input_all_perf_comb_melt_summ_cast_dt)] <- 0
  
roth_ec_filt_prep_input_all_perf_comb_gg<- ggplot(
  data=roth_ec_filt_prep_input_all_perf_comb_melt_summ_cast_dt,
  aes(x=Mean_Recall, y=Mean_Precision, 
      ymin=Mean_Precision-SD_Precision, ymax=Mean_Precision+SD_Precision, 
      xmin=Mean_Recall-SD_Recall, xmax=Mean_Recall+SD_Recall, col=Model))

roth_ec_filt_prep_input_all_perf_comb_gg+
  geom_point(size=3)+
  geom_errorbar(width=0.0025)+
  geom_errorbarh(height=0.005)+
  scale_x_continuous(breaks=seq(0.9,1,0.02))+
  scale_y_continuous(breaks=seq(0.5,1,0.1))+
  coord_cartesian(xlim=c(0.9,1), ylim=c(0.55,1))+
  scale_color_nejm()+
  labs(x="Recall", y="Precision")+
  theme_cowplot()
  
# RF for validation dataset
rf_model_a<- randomForest(Truth~., roth_ec_filt_prep_input_train_dt, mtry=10)
rf_model_b<- randomForest(Truth~., roth_ec_filt_prep_input_train_dt[,-c("MATCHES_MUT_SIGTrue")], mtry=10)

prob_cutoff<- 0.5
model_predict<- predict(rf_model_a, roth_ec_filt_prep_input_test_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
contigency_table<- create_cont_table(roth_ec_filt_prep_input_test_dt[,Truth], prediction_labels)

#get_accuracy(contigency_table)
#get_sensitivity(contigency_table)
#get_precision(contigency_table)

model_predict<- predict(rf_model_b, roth_ec_filt_prep_input_test_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
contigency_table<- create_cont_table(roth_ec_filt_prep_input_test_dt[,Truth], prediction_labels)

#get_accuracy(contigency_table)
#get_sensitivity(contigency_table)
#get_precision(contigency_table)

roth_ec_filt_prep_input_train_dt[,Truth_int:=as.integer(as.logical(Truth))]
roth_ec_filt_prep_input_test_dt[,Truth_int:=as.integer(as.logical(Truth))]

# GBM all
gbm_model<- gbm(Truth_int~., roth_ec_filt_prep_input_train_dt[,-c("Truth")], distribution="bernoulli", 
                      n.trees=median(roth_ec_filt_prep_input_gbm_cv_res_dt[[2]]$n.trees), 
                      interaction.depth=median(roth_ec_filt_prep_input_gbm_cv_res_dt[[2]]$interaction.depth), 
                      shrinkage=median(roth_ec_filt_prep_input_gbm_cv_res_dt[[2]]$shrinkage),
                      n.minobsinnode=median(roth_ec_filt_prep_input_gbm_cv_res_dt[[2]]$n.minobsinnode))

model_predict<- predict(gbm_model, roth_ec_filt_prep_input_test_dt, 
                        type="response", n.trees=median(roth_ec_filt_prep_input_gbm_cv_res_dt[[2]]$n.trees))

prediction_labels<- factor(model_predict>=prob_cutoff, levels=cont_table_factor_levels)
contigency_table<- create_cont_table(roth_ec_filt_prep_input_test_dt[,Truth], prediction_labels)

#get_accuracy(contigency_table)
#get_sensitivity(contigency_table)
#get_precision(contigency_table)
  
gbm_model_summary_dt<- as.data.table(summary.gbm(gbm_model))

```

### Cenik lab- CBS endogenous WT, NovaSeq 6000, HiSeq 2500, HiSeq 4000

```{r Figure 2C-E: Comparison of RF models between Cenik CBS simulated datasets, cross-generalization of models}

cenik_ec_dir<- paste(base_dir, "error_correction", "cenik_cbs", sep="/")

cenik_ec_cbs31_truth<- paste(cenik_ec_dir, "CBS1_31_pEZY3_train.truth.vcf", sep="/")
cenik_ec_cbs34_truth<- paste(cenik_ec_dir, "CBS1_34_pEZY3_train.truth.vcf", sep="/")
cenik_ec_cbs49_truth<- paste(cenik_ec_dir, "CBS1_49_pEZY3_train.truth.vcf", sep="/")

cenik_ec_cbs31_truth_dt<- fread(cenik_ec_cbs31_truth, skip=18, sep="\t")
cenik_ec_cbs34_truth_dt<- fread(cenik_ec_cbs34_truth, skip=18, sep="\t")
cenik_ec_cbs49_truth_dt<- fread(cenik_ec_cbs49_truth, skip=18, sep="\t")

cenik_ec_cbs31<- paste(cenik_ec_dir, "CBS1_31_pEZY3_train.var.cand.vcf.summary.txt", sep="/")
cenik_ec_cbs34<- paste(cenik_ec_dir, "CBS1_34_pEZY3_train.var.cand.vcf.summary.txt", sep="/")
cenik_ec_cbs49<- paste(cenik_ec_dir, "CBS1_49_pEZY3_train.var.cand.vcf.summary.txt", sep="/")

cenik_ec_cbs31_dt<- fread(cenik_ec_cbs31, sep="\t")
cenik_ec_cbs34_dt<- fread(cenik_ec_cbs34, sep="\t")
cenik_ec_cbs49_dt<- fread(cenik_ec_cbs49, sep="\t")

# Determine performance without any models
count_thresh<- 2
cenik_ec_cbs31_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]
cenik_ec_cbs31_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]
cenik_ec_cbs34_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]
cenik_ec_cbs34_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]
cenik_ec_cbs49_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]
cenik_ec_cbs49_truth_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

# Filter out variants that were not simulated due to constraints
cenik_ec_cbs31_truth_dt<- cenik_ec_cbs31_truth_dt[!grepl("CAO=0;", INFO, fixed=TRUE)]
cenik_ec_cbs34_truth_dt<- cenik_ec_cbs34_truth_dt[!grepl("CAO=0;", INFO, fixed=TRUE)]
cenik_ec_cbs49_truth_dt<- cenik_ec_cbs49_truth_dt[!grepl("CAO=0;", INFO, fixed=TRUE)]

cenik_ec_cbs31_dt[,Truth:=FALSE]
cenik_ec_cbs31_dt[VAR_ID%in%cenik_ec_cbs31_truth_dt[,VAR_ID], Truth:=TRUE]
cenik_ec_cbs31_dt[,Truth:=factor(Truth, levels=cont_table_factor_levels)]
cenik_ec_cbs31_dt[,satmut_utils_called:=factor(
  rep(TRUE, nrow(cenik_ec_cbs31_dt)), levels=cont_table_factor_levels)]
cenik_ec_cbs31_dt[,Sample:="CBS1_31"]

cenik_ec_cbs34_dt[,Truth:=FALSE]
cenik_ec_cbs34_dt[VAR_ID%in%cenik_ec_cbs34_truth_dt[,VAR_ID], Truth:=TRUE]
cenik_ec_cbs34_dt[,Truth:=factor(Truth, levels=cont_table_factor_levels)]
cenik_ec_cbs34_dt[,satmut_utils_called:=factor(
  rep(TRUE, nrow(cenik_ec_cbs34_dt)), levels=cont_table_factor_levels)]
cenik_ec_cbs34_dt[,Sample:="CBS1_34"]

cenik_ec_cbs49_dt[,Truth:=FALSE]
cenik_ec_cbs49_dt[VAR_ID%in%cenik_ec_cbs49_truth_dt[,VAR_ID], Truth:=TRUE]
cenik_ec_cbs49_dt[,Truth:=factor(Truth, levels=cont_table_factor_levels)]
cenik_ec_cbs49_dt[,satmut_utils_called:=factor(
  rep(TRUE, nrow(cenik_ec_cbs49_dt)), levels=cont_table_factor_levels)]
cenik_ec_cbs49_dt[,Sample:="CBS1_49"]

# Apply basic count filters prior to performance evaluation
cenik_ec_cbs31_filt_dt<- cenik_ec_cbs31_dt[CAO>=count_thresh]
cenik_ec_cbs34_filt_dt<- cenik_ec_cbs34_dt[CAO>=count_thresh]
cenik_ec_cbs49_filt_dt<- cenik_ec_cbs49_dt[CAO>=count_thresh]

cenik_ec_cbs31_filt_cont_table<- create_cont_table(
  cenik_ec_cbs31_filt_dt[,Truth], cenik_ec_cbs31_filt_dt[,satmut_utils_called])

cenik_ec_cbs34_filt_cont_table<- create_cont_table(
  cenik_ec_cbs34_filt_dt[,Truth], cenik_ec_cbs34_filt_dt[,satmut_utils_called])

cenik_ec_cbs49_filt_cont_table<- create_cont_table(
  cenik_ec_cbs49_filt_dt[,Truth], cenik_ec_cbs49_filt_dt[,satmut_utils_called])

cenik_ec_cbs31_filt_prec<- get_precision(cenik_ec_cbs31_filt_cont_table)
cenik_ec_cbs31_filt_recall<- get_sensitivity(cenik_ec_cbs31_filt_cont_table)

cenik_ec_cbs34_filt_prec<- get_precision(cenik_ec_cbs34_filt_cont_table)
cenik_ec_cbs34_filt_recall<- get_sensitivity(cenik_ec_cbs34_filt_cont_table)

cenik_ec_cbs49_filt_prec<- get_precision(cenik_ec_cbs49_filt_cont_table)
cenik_ec_cbs49_filt_recall<- get_sensitivity(cenik_ec_cbs49_filt_cont_table)

cenik_roth_ec_all_prec<- c(roth_ec_filt_prec, cenik_ec_cbs31_filt_prec, cenik_ec_cbs34_filt_prec, cenik_ec_cbs49_filt_prec)
cenik_roth_ec_all_recall<- c(roth_ec_filt_recall, cenik_ec_cbs31_filt_recall, cenik_ec_cbs34_filt_recall, cenik_ec_cbs49_filt_recall)

# mean(cenik_roth_ec_all_prec)
# sd(cenik_roth_ec_all_prec)

# Remove variants with incomplete features
# For this particular amplicon chemistry, we don't expect variants on R1(+)/R2(-)
cenik_ec_cbs49_filt_dt<- cenik_ec_cbs49_dt[!is.na(R1_PLUS_MED_RP),]

modeling_features<- c(
  "VAR_ID", "SUBST", "UP_REF_NT", "DOWN_REF_NT", 
  "TYPE", "log10_CAF", "MATCHES_MUT_SIG",
  "R1_PLUS_MED_RP", "R2_MINUS_MED_RP", 
  "R1_PLUS_MED_BQ", "R2_MINUS_MED_BQ", 
  "R1_PLUS_MED_NM", "R2_MINUS_MED_NM")

# Determine if any features should be dropped to protect against collinearity
#cenik_ec_cbs31_dt[,cor(R1_PLUS_MED_NM, R2_MINUS_MED_NM, method="spearman")]
#cenik_ec_cbs31_dt[,cor(R1_PLUS_MED_RP, R2_MINUS_MED_RP, method="spearman")]

# Because the NM and RP are collinear between R1 and R2, just select the stat for one read
modeling_features_no_collinear<- modeling_features[
  !grepl("R1_PLUS_MED_RP|R1_PLUS_MED_NM", modeling_features)]

# Prep the datasets for modeling
cenik_ec_cbs31_prep_input_dt<- prep_train_dt(
  train_dt=cenik_ec_cbs31_dt, truth_dt=cenik_ec_cbs31_truth_dt, 
  modeling_features=modeling_features_no_collinear)

cenik_ec_cbs34_prep_input_dt<- prep_train_dt(
  train_dt=cenik_ec_cbs34_dt, truth_dt=cenik_ec_cbs34_truth_dt, 
  modeling_features=modeling_features_no_collinear)

cenik_ec_cbs49_prep_input_dt<- prep_train_dt(
  train_dt=cenik_ec_cbs49_filt_dt, truth_dt=cenik_ec_cbs49_truth_dt, 
  modeling_features=modeling_features_no_collinear)


# Model selection by 10-fold nested CV

# Get indices for 10 mutually exclusive folds
cenik_ec_cbs31_prep_input_kfolds<- create_kfolds(cenik_ec_cbs31_prep_input_dt)
cenik_ec_cbs34_prep_input_kfolds<- create_kfolds(cenik_ec_cbs34_prep_input_dt)
cenik_ec_cbs49_prep_input_kfolds<- create_kfolds(cenik_ec_cbs49_prep_input_dt)

# CBS1_31, HiSeq 4000
cenik_ec_cbs31_prep_input_test_dt<- cenik_ec_cbs31_prep_input_dt[
  cenik_ec_cbs31_prep_input_kfolds[[1]]]

cenik_ec_cbs31_prep_input_train_dt<- cenik_ec_cbs31_prep_input_dt[
  !cenik_ec_cbs31_prep_input_kfolds[[1]]]

# CBS1_34, NovaSeq 6000
cenik_ec_cbs34_prep_input_test_dt<- cenik_ec_cbs34_prep_input_dt[
  cenik_ec_cbs34_prep_input_kfolds[[1]]]

cenik_ec_cbs34_prep_input_train_dt<- cenik_ec_cbs34_prep_input_dt[
  !cenik_ec_cbs34_prep_input_kfolds[[1]]]

# CBS1_49, HiSeq 2500
cenik_ec_cbs49_prep_input_test_dt<- cenik_ec_cbs49_prep_input_dt[
  cenik_ec_cbs49_prep_input_kfolds[[1]]]

cenik_ec_cbs49_prep_input_train_dt<- cenik_ec_cbs49_prep_input_dt[
  !cenik_ec_cbs49_prep_input_kfolds[[1]]]

# Train RFs on all datasets and validate performance
cenik_ec_cbs31_prep_input_train_cv_res<- run_nested_cv(
  cenik_ec_cbs31_prep_input_train_dt, model="rf")

cenik_ec_cbs34_prep_input_train_cv_res<- run_nested_cv(
  cenik_ec_cbs34_prep_input_train_dt, model="rf")

cenik_ec_cbs49_prep_input_train_cv_res<- run_nested_cv(
  cenik_ec_cbs49_prep_input_train_dt, model="rf")

# Combine RF performance for all models
cenik_roth_ec_filt_prep_input_all_perf_dt<- as.data.table(rbind(
  roth_ec_filt_prep_input_rf_cv_res_dt[[1]],
  cenik_ec_cbs31_prep_input_train_cv_res[[1]],
  cenik_ec_cbs34_prep_input_train_cv_res[[1]],
  cenik_ec_cbs49_prep_input_train_cv_res[[1]]))

cenik_roth_ec_filt_prep_input_all_perf_dt[
  ,Dataset:=as.factor(rep(c("NextSeq 500", "HiSeq 4000", "Novaseq 6000", "HiSeq 2500"), each=10))]

cenik_roth_ec_filt_prep_input_all_perf_melt_dt<- melt(
  data=cenik_roth_ec_filt_prep_input_all_perf_dt, 
  id.vars=c("Dataset"), value.name="Value", 
  variable.name="Measure")

cenik_roth_ec_filt_prep_input_all_perf_melt_dt[
  ,Dataset:=factor(Dataset, levels=c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"))]
  
cenik_roth_ec_filt_prep_input_all_perf_gg<- ggplot(
  data=cenik_roth_ec_filt_prep_input_all_perf_melt_dt, 
  aes(x=Dataset, y=Value, col=Measure))

cenik_roth_ec_filt_prep_input_all_perf_gg+
  geom_boxplot(position="dodge")+
  scale_color_simpsons()+
  scale_y_continuous(breaks=seq(0.9,1,.02))+
  coord_cartesian(ylim=c(0.9,1))+
  labs(y="10-fold CV accuracy")+
  theme_cowplot()

# Test performance of RFs on each validation test set
set.seed(9)

rf_model_nextseq500<- randomForest(
  Truth~., roth_ec_filt_prep_input_train_dt[,-c("Truth_int")], 
  mtry=round(median(roth_ec_filt_prep_input_rf_cv_res_dt[[2]]$mtry)))

rf_model_novaseq6000<- randomForest(
  Truth~., cenik_ec_cbs34_prep_input_train_dt, 
  mtry=round(median(cenik_ec_cbs34_prep_input_train_cv_res[[2]]$mtry)))

rf_model_hiseq2500<- randomForest(
  Truth~., cenik_ec_cbs49_prep_input_train_dt, 
  mtry=round(median(cenik_ec_cbs49_prep_input_train_cv_res[[2]]$mtry)))

rf_model_hiseq4000<- randomForest(
  Truth~., cenik_ec_cbs31_prep_input_train_dt, 
  mtry=round(median(cenik_ec_cbs31_prep_input_train_cv_res[[2]]$mtry)))

# Generate a combined model with all simulated datasets
cenik_roth_ec_all_prep_input_train_dt<- rbind(
  roth_ec_filt_prep_input_train_dt[,-c("Truth_int")], cenik_ec_cbs34_prep_input_train_dt, 
  cenik_ec_cbs49_prep_input_train_dt, cenik_ec_cbs31_prep_input_train_dt)

cenik_roth_ec_all_prep_input_test_dt<- rbind(
  roth_ec_filt_prep_input_test_dt[,-c("Truth_int")], cenik_ec_cbs34_prep_input_test_dt, 
  cenik_ec_cbs49_prep_input_test_dt, cenik_ec_cbs31_prep_input_test_dt)

rf_model_all<- randomForest(Truth~., cenik_roth_ec_all_prep_input_train_dt, mtry=6, importance=TRUE)
par(pch=19)
varImpPlot(rf_model_all, type=1, n.var=15)

prob_cutoff<- 0.5

# wt_nonselect Nextseq 500 perf
model_predict<- predict(rf_model_nextseq500, roth_ec_filt_prep_input_test_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
contigency_table<- create_cont_table(roth_ec_filt_prep_input_test_dt[,Truth], prediction_labels)

roth_ec_filt_prep_input_test_acc<- get_accuracy(contigency_table)
roth_ec_filt_prep_input_test_recall<- get_sensitivity(contigency_table)
roth_ec_filt_prep_input_test_prec<- get_precision(contigency_table)

# CBS1_34, NovaSeq 6000 perf
model_predict<- predict(rf_model_novaseq6000, cenik_ec_cbs34_prep_input_test_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
contigency_table<- create_cont_table(cenik_ec_cbs34_prep_input_test_dt[,Truth], prediction_labels)

cenik_ec_cbs34_prep_input_test_acc<- get_accuracy(contigency_table)
cenik_ec_cbs34_prep_input_test_recall<- get_sensitivity(contigency_table)
cenik_ec_cbs34_prep_input_test_prec<- get_precision(contigency_table)

# CBS1_49, HiSeq 2500 perf
model_predict<- predict(rf_model_hiseq2500, cenik_ec_cbs49_prep_input_test_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
contigency_table<- create_cont_table(cenik_ec_cbs49_prep_input_test_dt[,Truth], prediction_labels)

cenik_ec_cbs49_prep_input_test_acc<- get_accuracy(contigency_table)
cenik_ec_cbs49_prep_input_test_recall<- get_sensitivity(contigency_table)
cenik_ec_cbs49_prep_input_test_prec<- get_precision(contigency_table)

# CBS1_31, HiSeq 4000 perf
model_predict<- predict(rf_model_hiseq4000, cenik_ec_cbs31_prep_input_test_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
contigency_table<- create_cont_table(cenik_ec_cbs31_prep_input_test_dt[,Truth], prediction_labels)

cenik_ec_cbs31_prep_input_test_acc<- get_accuracy(contigency_table)
cenik_ec_cbs31_prep_input_test_recall<- get_sensitivity(contigency_table)
cenik_ec_cbs31_prep_input_test_prec<- get_precision(contigency_table)

all_model_accs<- c(roth_ec_filt_prep_input_test_acc, cenik_ec_cbs34_prep_input_test_acc,
                   cenik_ec_cbs49_prep_input_test_acc, cenik_ec_cbs31_prep_input_test_acc)

# mean(all_model_accs)
# sd(all_model_accs)

# Determine cross-generalization of models
cenik_roth_ec_all_prep_input_train_models<- list(
  rf_model_nextseq500, rf_model_novaseq6000, 
  rf_model_hiseq2500, rf_model_hiseq4000)

cenik_roth_ec_all_prep_input_test_datasets<- list(
  roth_ec_filt_prep_input_test_dt[,-c("Truth_int")], cenik_ec_cbs34_prep_input_test_dt, 
  cenik_ec_cbs49_prep_input_test_dt, cenik_ec_cbs31_prep_input_test_dt)

crossgen_nmodels<- length(cenik_roth_ec_all_prep_input_train_models)
crossgen_ndatasets<- length(cenik_roth_ec_all_prep_input_test_datasets)

cenik_roth_ec_all_prep_input_crossgen<- NULL

for(i in 1:crossgen_nmodels){
  
  model_i<- cenik_roth_ec_all_prep_input_train_models[[i]]
  
  for(j in 1:crossgen_ndatasets){
    
    test_dataset_j<- cenik_roth_ec_all_prep_input_test_datasets[[j]]
    
    model_predict<- predict(model_i, test_dataset_j, type="prob")
    prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
    contigency_table<- create_cont_table(test_dataset_j[,Truth], prediction_labels)

    test_dataset_ij_acc<- get_accuracy(contigency_table)
    test_dataset_ij_precision<- get_precision(contigency_table)
    test_dataset_ij_recall<- get_sensitivity(contigency_table)
    
    cenik_roth_ec_all_prep_input_crossgen<- rbind(
      cenik_roth_ec_all_prep_input_crossgen,
      matrix(data=c(test_dataset_ij_acc, test_dataset_ij_precision, test_dataset_ij_recall),
             nrow=1, ncol=3))
  }
  colnames(cenik_roth_ec_all_prep_input_crossgen)<- c("Accuracy", "Precision", "Recall")
}

cenik_roth_ec_all_prep_input_crossgen_dt<- as.data.table(cenik_roth_ec_all_prep_input_crossgen)

cenik_roth_ec_all_prep_input_crossgen_dt[,Train_dataset:=factor(
  rep(c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"), each=4),
  levels=c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"))]

cenik_roth_ec_all_prep_input_crossgen_dt[,Test_dataset:=factor(
  rep(c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"), times=4),
  levels=c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"))]

cenik_roth_ec_all_prep_input_crossgen_melt_dt<- melt(
  data=cenik_roth_ec_all_prep_input_crossgen_dt, 
  id.vars=c("Train_dataset", "Test_dataset"), value.name="Value", 
  variable.name="Measure")

# Plot the results as heatmaps
cenik_roth_ec_all_prep_input_crossgen_gg<- ggplot(
  data=cenik_roth_ec_all_prep_input_crossgen_melt_dt[Measure!="Accuracy"],
  aes(x=Train_dataset, y=Test_dataset, fill=Value))

cenik_roth_ec_all_prep_input_crossgen_gg+
  facet_wrap(~Measure)+
  geom_tile()+
  scale_fill_viridis(limits=c(0.78,1))+
  labs(x="Train dataset", y="Test dataset")+
  theme_cowplot()

cenik_roth_ec_all_prep_input_crossgen_melt_dt[
  Train_dataset=="NextSeq 500" & Test_dataset!="NextSeq 500" & Measure=="Accuracy", 
  .(Mean_acc=mean(Value), SD_acc=sd(Value))]

cenik_roth_ec_all_prep_input_crossgen_melt_dt[
  Test_dataset!="NextSeq 500" & Train_dataset!="NextSeq 500" & Measure=="Accuracy",
  .(Mean_acc=mean(Value), SD_acc=sd(Value))]

```

```{r Figure 2F: Filtering of negative control calls with models}

roth_ec_dir<- paste(base_dir, "error_correction", "roth_cbs", sep="/")
roth_wt<- paste(roth_ec_dir, "wt_nonselect_r.var.cand.vcf.summary.txt", sep="/")
roth_wt_dt<- fread(roth_wt, sep="\t")

cenik_ec_dir<- paste(base_dir, "error_correction", "cenik_cbs", sep="/")
cenik_cbs31_wt<- paste(cenik_ec_dir, "CBS1_31_S10_L006_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs34_wt<- paste(cenik_ec_dir, "CBS1_34_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs49_wt<- paste(cenik_ec_dir, "CBS1_49_R.var.cand.vcf.summary.txt", sep="/")
cenik_cbs31_wt_dt<- fread(cenik_cbs31_wt, sep="\t")
cenik_cbs34_wt_dt<- fread(cenik_cbs34_wt, sep="\t")
cenik_cbs49_wt_dt<- fread(cenik_cbs49_wt, sep="\t")

roth_wt_dt[,Sample:="Roth_WT"]
cenik_cbs31_wt_dt[,Sample:="CBS1_31"]
cenik_cbs34_wt_dt[,Sample:="CBS1_34"]
cenik_cbs49_wt_dt[,Sample:="CBS1_49"]

# Note that since now we are not training models we do not include the label Truth
filter_modeling_features<- c(
  "SUBST", "UP_REF_NT", "DOWN_REF_NT", 
  "TYPE", "log10_CAF", "MATCHES_MUT_SIG",
  "R2_MINUS_MED_RP", "R1_PLUS_MED_BQ", 
  "R2_MINUS_MED_BQ", "R2_MINUS_MED_NM")

# It is up to the user to ensure all variant filters occurring in prep_train_dt()
# (e.g. CAF filter) are mirrored and applied to the input data.table prior to model predictions
# as the nrow of the input needs to exactly match the output for proper model-based filtering
# the last filter is specific to certain datasets where variant calls are made from reads
# on the unexpected strand (due tovariant library and/or alignment artifacts)
caf_thresh<- 0.3
roth_wt_filt_dt<- roth_wt_dt[LOCATION=="CDS" & CAF < caf_thresh & !is.na(R1_PLUS_MED_RP)]
cenik_cbs31_wt_filt_dt<- cenik_cbs31_wt_dt[LOCATION=="CDS" & CAF < caf_thresh]
cenik_cbs34_wt_filt_dt<- cenik_cbs34_wt_dt[LOCATION=="CDS" & CAF < caf_thresh]
cenik_cbs49_wt_filt_dt<- cenik_cbs49_wt_dt[LOCATION=="CDS" & CAF < caf_thresh & !is.na(R1_PLUS_MED_RP)]

# Now annotate the modeling features not already provided and one-hot encode
roth_wt_filt_prep_input_dt<- prep_train_dt(
  train_dt=roth_wt_filt_dt, modeling_features=filter_modeling_features, caf_thresh=caf_thresh)
  
cenik_cbs31_wt_prep_input_dt<- prep_train_dt(
  train_dt=cenik_cbs31_wt_filt_dt, modeling_features=filter_modeling_features, caf_thresh=caf_thresh)

cenik_cbs34_wt_prep_input_dt<- prep_train_dt(
  train_dt=cenik_cbs34_wt_filt_dt, modeling_features=filter_modeling_features, caf_thresh=caf_thresh)

cenik_cbs49_wt_prep_input_dt<- prep_train_dt(
  train_dt=cenik_cbs49_wt_filt_dt, modeling_features=filter_modeling_features, caf_thresh=caf_thresh)

# Validate we have the same number of records before and after annotation/prep
#nrow(roth_wt_filt_prep_input_dt) == nrow(roth_wt_filt_dt)
#nrow(cenik_cbs31_wt_prep_input_dt) == nrow(cenik_cbs31_wt_filt_dt)
#nrow(cenik_cbs34_wt_prep_input_dt) == nrow(cenik_cbs34_wt_filt_dt)
#nrow(cenik_cbs49_wt_prep_input_dt) == nrow(cenik_cbs49_wt_filt_dt)

# Now filter negative control datasets with each model
prob_cutoff<- 0.5

# wt_nonselect Nextseq 500 perf
model_predict<- predict(rf_model_nextseq500, roth_wt_filt_prep_input_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
roth_wt_filt_prep_input_mfilt_dt<- roth_wt_filt_prep_input_dt[as.logical(prediction_labels)]

# CBS1_34, NovaSeq 6000 perf
model_predict<- predict(rf_model_novaseq6000, cenik_cbs34_wt_prep_input_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
cenik_cbs34_wt_prep_input_mfilt_dt<- cenik_cbs34_wt_prep_input_dt[as.logical(prediction_labels)]

# CBS1_49, HiSeq 2500 perf
model_predict<- predict(rf_model_hiseq2500, cenik_cbs49_wt_prep_input_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
cenik_cbs49_wt_prep_input_mfilt_dt<- cenik_cbs49_wt_prep_input_dt[as.logical(prediction_labels)]

# CBS1_31, HiSeq 4000 perf
model_predict<- predict(rf_model_hiseq4000, cenik_cbs31_wt_prep_input_dt, type="prob")
prediction_labels<- factor(model_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)
cenik_cbs31_wt_prep_input_mfilt_dt<- cenik_cbs31_wt_prep_input_dt[as.logical(prediction_labels)]

# As expected, we saw a large reduction in the number of mismatched bases following model filtering
model_filt_props_dt<- data.table(
  Pre_filter=c(nrow(roth_wt_filt_prep_input_dt), nrow(cenik_cbs34_wt_prep_input_dt), 
               nrow(cenik_cbs49_wt_prep_input_dt), nrow(cenik_cbs31_wt_prep_input_dt)),
  Post_filter=c(nrow(roth_wt_filt_prep_input_mfilt_dt), nrow(cenik_cbs34_wt_prep_input_mfilt_dt), 
                nrow(cenik_cbs49_wt_prep_input_mfilt_dt), nrow(cenik_cbs31_wt_prep_input_mfilt_dt)),
  Dataset=factor(c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"), 
                 levels=c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000")))

model_filt_props_melt_dt<- melt(
  data=model_filt_props_dt, 
  id.vars=c("Dataset"), value.name="Count", variable.name="Filter")

model_filt_props_melt_dt[,Filter:=factor(Filter, levels=c("Pre_filter", "Post_filter"))]

model_filt_props_melt_gg<- ggplot(
  data=model_filt_props_melt_dt, 
  aes(y=Dataset, x=Count, fill=Filter, label=as.character(Count)))

model_filt_props_melt_gg+
  geom_bar(stat="identity", position="dodge")+
  geom_text(position=position_dodge(0.9))+
  scale_fill_d3()+
  labs(y="Number variants")+
  theme_cowplot()+
  theme(legend.position="bottom")

```

## Effect of primer masking on simulated datasets

```{r Figure 3C: Effect of primer masking}

# Read in sim datasets with and without primer masking
roth_ec_dir<- paste(base_dir, "error_correction", "roth_cbs", sep="/")
mask_dir<- "no_primer_masking"
roth_ec_truth<- paste(roth_ec_dir, "CBS_pAG415GAL_train.truth.vcf", sep="/")
roth_ec<- paste(roth_ec_dir, "CBS_pAG415GAL_train.var.cand.vcf.summary.txt", sep="/")
roth_ec_nomask<- paste(roth_ec_dir, mask_dir, "CBS_pAG415GAL_train.var.cand.vcf.summary.txt", sep="/")

roth_ec_truth_dt<- fread(roth_ec_truth, skip=18, sep="\t")
roth_ec_dt<- fread(roth_ec, sep="\t")
roth_ec_nomask_dt<- fread(roth_ec_nomask, sep="\t")

cenik_ec_dir<- paste(base_dir, "error_correction", "cenik_cbs", sep="/")

cenik_ec_cbs31_truth<- paste(cenik_ec_dir, "CBS1_31_pEZY3_train.truth.vcf", sep="/")
cenik_ec_cbs34_truth<- paste(cenik_ec_dir, "CBS1_34_pEZY3_train.truth.vcf", sep="/")
cenik_ec_cbs49_truth<- paste(cenik_ec_dir, "CBS1_49_pEZY3_train.truth.vcf", sep="/")

cenik_ec_cbs31_truth_dt<- fread(cenik_ec_cbs31_truth, skip=18, sep="\t")
cenik_ec_cbs34_truth_dt<- fread(cenik_ec_cbs34_truth, skip=18, sep="\t")
cenik_ec_cbs49_truth_dt<- fread(cenik_ec_cbs49_truth, skip=18, sep="\t")

cenik_ec_cbs31<- paste(cenik_ec_dir, "CBS1_31_pEZY3_train.var.cand.vcf.summary.txt", sep="/")
cenik_ec_cbs34<- paste(cenik_ec_dir, "CBS1_34_pEZY3_train.var.cand.vcf.summary.txt", sep="/")
cenik_ec_cbs49<- paste(cenik_ec_dir, "CBS1_49_pEZY3_train.var.cand.vcf.summary.txt", sep="/")

cenik_ec_cbs31_nomask<- paste(cenik_ec_dir, mask_dir, "CBS1_31_pEZY3_train.var.cand.vcf.summary.txt", sep="/")
cenik_ec_cbs34_nomask<- paste(cenik_ec_dir, mask_dir, "CBS1_34_pEZY3_train.var.cand.vcf.summary.txt", sep="/")
cenik_ec_cbs49_nomask<- paste(cenik_ec_dir, mask_dir, "CBS1_49_pEZY3_train.var.cand.vcf.summary.txt", sep="/")

cenik_ec_cbs31_dt<- fread(cenik_ec_cbs31, sep="\t")
cenik_ec_cbs34_dt<- fread(cenik_ec_cbs34, sep="\t")
cenik_ec_cbs49_dt<- fread(cenik_ec_cbs49, sep="\t")

cenik_ec_cbs31_nomask_dt<- fread(cenik_ec_cbs31_nomask, sep="\t")
cenik_ec_cbs34_nomask_dt<- fread(cenik_ec_cbs34_nomask, sep="\t")
cenik_ec_cbs49_nomask_dt<- fread(cenik_ec_cbs49_nomask, sep="\t")

# Annotate Truth so we can see primer masking effects on both TPs, FPs
roth_ec_merge_dt<- merge_truth_dt(roth_ec_dt, roth_ec_truth_dt)
roth_ec_nomask_merge_dt<- merge_truth_dt(roth_ec_nomask_dt, roth_ec_truth_dt)

cenik_ec_cbs31_merge_dt<- merge_truth_dt(cenik_ec_cbs31_dt, cenik_ec_cbs31_truth_dt)
cenik_ec_cbs31_nomask_merge_dt<- merge_truth_dt(cenik_ec_cbs31_nomask_dt, cenik_ec_cbs31_truth_dt)

cenik_ec_cbs34_merge_dt<- merge_truth_dt(cenik_ec_cbs34_dt, cenik_ec_cbs34_truth_dt)
cenik_ec_cbs34_nomask_merge_dt<- merge_truth_dt(cenik_ec_cbs34_nomask_dt, cenik_ec_cbs34_truth_dt)

cenik_ec_cbs49_merge_dt<- merge_truth_dt(cenik_ec_cbs49_dt, cenik_ec_cbs49_truth_dt)
cenik_ec_cbs49_nomask_merge_dt<- merge_truth_dt(cenik_ec_cbs49_nomask_dt, cenik_ec_cbs49_truth_dt)

# Because primers were all the same for these datasets, combine them
# and annotate the primer regions
cenik_roth_ec_all_mask_comp_dt<- rbind(
  roth_ec_merge_dt, roth_ec_nomask_merge_dt, 
  cenik_ec_cbs34_merge_dt, cenik_ec_cbs34_nomask_merge_dt, 
  cenik_ec_cbs49_merge_dt, cenik_ec_cbs49_nomask_merge_dt, 
  cenik_ec_cbs31_merge_dt, cenik_ec_cbs31_nomask_merge_dt)

cenik_roth_ec_all_mask_comp_dt[,Truth:=factor(Truth, levels=cont_table_factor_levels)]

# Annotate dataset
cenik_roth_ec_all_mask_comp_dt[,Dataset:=rep(
  rep(c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"), each=2), times=c(
    nrow(roth_ec_merge_dt), nrow(roth_ec_nomask_merge_dt), 
    nrow(cenik_ec_cbs34_merge_dt), nrow(cenik_ec_cbs34_nomask_merge_dt),
    nrow(cenik_ec_cbs49_merge_dt), nrow(cenik_ec_cbs49_nomask_merge_dt),
    nrow(cenik_ec_cbs31_merge_dt), nrow(cenik_ec_cbs31_nomask_merge_dt)))]

cenik_roth_ec_all_mask_comp_dt[,Dataset:=factor(
  Dataset, levels=c("NextSeq 500", "Novaseq 6000", "HiSeq 2500", "HiSeq 4000"))]

# Annotate primer masking
cenik_roth_ec_all_mask_comp_dt[,Primer_mask:=rep(rep(c("Mask", "Nomask"), 4), times=c(
    nrow(roth_ec_merge_dt), nrow(roth_ec_nomask_merge_dt), 
    nrow(cenik_ec_cbs34_merge_dt), nrow(cenik_ec_cbs34_nomask_merge_dt),
    nrow(cenik_ec_cbs49_merge_dt), nrow(cenik_ec_cbs49_nomask_merge_dt),
    nrow(cenik_ec_cbs31_merge_dt), nrow(cenik_ec_cbs31_nomask_merge_dt)))]

cenik_roth_ec_all_mask_comp_dt[,Primer_mask:=factor(Primer_mask, levels=c("Nomask", "Mask"))]

# Finally annotate whether or not the variant intersects a primer span
# Primer coordinates
ref_dir<- "/Users/ianhoskins/reference_files"
cenik_roth_primer_coords_dt<- fread(paste(ref_dir,
  "CBS_pEZY3_primers_coding.bed", sep="/"), sep="\t")

primer_coords<- cenik_roth_primer_coords_dt[,seq(V2,V3),by=.(V4)]
primer_coords_pos<- primer_coords[,unique(V1)]

cenik_roth_ec_all_mask_comp_dt[POS%in%primer_coords_pos, Overlaps_primer:=TRUE]
cenik_roth_ec_all_mask_comp_dt[!POS%in%primer_coords_pos, Overlaps_primer:=FALSE]
cenik_roth_ec_all_mask_comp_dt[,Overlaps_primer:=factor(Overlaps_primer, levels=c("TRUE", "FALSE"))]

cenik_roth_ec_all_mask_comp_filt_dt<- cenik_roth_ec_all_mask_comp_dt[TYPE=="SNP"]

two_cols<- viridis(option="inferno", n=2, begin=0.2, end=0.8)

cenik_roth_ec_all_mask_comp_gg<- ggplot(
  data=cenik_roth_ec_all_mask_comp_filt_dt, 
  aes(x=log10(CAO.x + 1), y=log10(CAO.y + 1), 
      col=Overlaps_primer))

# col=two_cols[1]
cenik_roth_ec_all_mask_comp_gg+
  facet_wrap(~Primer_mask)+
  geom_abline(slope=1, linetype=2, col="grey")+
  geom_smooth(method="loess", se=TRUE)+
  scale_color_manual(values=rev(two_cols))+
  coord_cartesian(xlim=c(0,4.2), ylim=c(0,4.2))+
  labs(x="Expected log10 count + 1", 
       y="Observed log10 count + 1", 
       col="Overlaps primer")+
  theme_cowplot()


# Determine the reduction if false positives
cenik_roth_ec_all_mask_comp_filt_counts_dt<- 
  cenik_roth_ec_all_mask_comp_filt_dt[
    Truth=="FALSE", .N, by=.(Dataset, Primer_mask)]

cenik_roth_ec_all_mask_comp_filt_counts_cast_dt<- dcast(
  data=cenik_roth_ec_all_mask_comp_filt_counts_dt, 
  Dataset~Primer_mask, value.var="N")

cenik_roth_ec_all_mask_comp_filt_counts_cast_dt[
  ,`:=`(Diff=Nomask-Mask, Proportion=Mask/Nomask)]

#cenik_roth_ec_all_mask_comp_filt_counts_cast_dt[,mean(Proportion)]
#cenik_roth_ec_all_mask_comp_filt_counts_cast_dt[,sd(Proportion)]

```

## Effect of UMI-based consensus deduplication on simulated datasets

```{r Figure 3D-E: Consensus deduplication}

cenik_ec_dir<- paste(base_dir, "error_correction", "cenik_cbs", sep="/")
dedup_dir<- paste(cenik_ec_dir, "amp", sep="/")

cenik_wt_cbs50_wt_nondedup<- paste(dedup_dir, "CBS1_50_R.var.cand.vcf.summary.txt", sep="/")
cenik_wt_cbs50_wt_nondedup_dt<- fread(cenik_wt_cbs50_wt_nondedup, sep="\t")

cenik_wt_cbs50_wt_nondedup_cov<- paste(dedup_dir, "CBS1_50_R.cov.bedgraph", sep="/")
cenik_wt_cbs50_wt_nondedup_cov_dt<- fread(cenik_wt_cbs50_wt_nondedup_cov, sep="\t")

cenik_wt_cbs50_wt_cdedup<- paste(dedup_dir, "cdedup", "CBS1_50_R.var.cand.vcf.summary.txt", sep="/")
cenik_wt_cbs50_wt_cdedup_dt<- fread(cenik_wt_cbs50_wt_cdedup, sep="\t")

cenik_wt_cbs50_wt_cdedup_cov<- paste(dedup_dir, "cdedup", "CBS1_50_R.cov.bedgraph", sep="/")
cenik_wt_cbs50_wt_cdedup_cov_dt<- fread(cenik_wt_cbs50_wt_cdedup_cov, sep="\t")

cenik_wt_cbs50_wt_both_dt<- rbind(
  cenik_wt_cbs50_wt_nondedup_dt, cenik_wt_cbs50_wt_cdedup_dt)

cenik_wt_cbs50_wt_both_dt[,Processing:=as.factor(
  rep(c("Nondeduplicated", "Deduplicated"), 
      c(nrow(cenik_wt_cbs50_wt_nondedup_dt), nrow(cenik_wt_cbs50_wt_cdedup_dt))))]

cenik_wt_cbs50_wt_both_cov_dt<- rbind(
  cenik_wt_cbs50_wt_nondedup_cov_dt, cenik_wt_cbs50_wt_cdedup_cov_dt)

cenik_wt_cbs50_wt_both_cov_dt[,Processing:=as.factor(
  rep(c("Nondeduplicated", "Deduplicated"), 
      c(nrow(cenik_wt_cbs50_wt_nondedup_cov_dt), nrow(cenik_wt_cbs50_wt_cdedup_cov_dt))))]

# First plot coverage over the target by nondedup/dedup
target_coords<- seq(261,1916)

cenik_wt_cbs50_wt_both_cov_filt_dt<- 
  cenik_wt_cbs50_wt_both_cov_dt[(V2+1)%in%target_coords]

cenik_wt_cbs50_wt_both_cov_waterfall_dt<- 
  waterfall_ecdf(cenik_wt_cbs50_wt_both_cov_filt_dt, 
                 key_features="Processing")

cenik_wt_cbs50_wt_both_cov_waterfall_gg<- ggplot(
  data=cenik_wt_cbs50_wt_both_cov_waterfall_dt,
  aes(x=log10(Coverage+1), y=Proportion_greater_than, col=Processing))

cenik_wt_cbs50_wt_both_cov_waterfall_gg+
  geom_line()+
  scale_color_uchicago()+
  coord_cartesian(xlim=c(4.4,5.6))+
  scale_x_continuous(breaks=seq(4.6,5.6,0.4))+
  labs(x="log10(coverage + 1)", 
       y="Proportion bases > coverage")+
  theme_cowplot()

cenik_wt_cbs50_wt_both_dt[,VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

cenik_wt_cbs50_wt_both_collapse_dt<- collapse_mnps(
  cenik_wt_cbs50_wt_both_dt, key_features=c("Processing", "VAR_ID"))

cenik_wt_cbs50_wt_both_collapse_dt[
  ,TYPE:=get_var_type(REF, ALT), by=seq_len(nrow(cenik_wt_cbs50_wt_both_collapse_dt))]

cenik_wt_cbs50_wt_both_collapse_filt_dt<- 
  cenik_wt_cbs50_wt_both_collapse_dt[LOCATION=="CDS"]

cenik_wt_cbs50_wt_both_collapse_filt_type_counts_dt<- 
  cenik_wt_cbs50_wt_both_collapse_filt_dt[,.N, by=.(Processing, TYPE)]

cenik_wt_cbs50_wt_both_collapse_filt_type_counts_dt[,TYPE:=factor(
  TYPE, levels=c("SNP", "di_nt_MNP", "tri_nt_MNP"))]

cenik_wt_cbs50_wt_both_collapse_filt_type_counts_gg<- ggplot(
  data=cenik_wt_cbs50_wt_both_collapse_filt_type_counts_dt,
  aes(x=TYPE, y=N, fill=Processing))

cenik_wt_cbs50_wt_both_collapse_filt_type_counts_gg+
  geom_bar(stat="identity", position="dodge")+
  scale_fill_uchicago()+
  labs(x="Variant type", 
       y="False positive counts")+
  theme_cowplot()

```

## CBS mNA abundance variant effect analysis:

Figures 5-6

```{r Pico RNA quality check}

cbs_02dec2021_rna_qual_dir<- paste(base_dir, "CBS_02DEC2021", "RNA_quality", sep="/")

rep1_rna<- paste(cbs_02dec2021_rna_qual_dir, "IH_26JAN2022_CBS_Eukaryote Total RNA Pico_DE13806310_2022-01-26_17-00-20_Sample6.csv", sep="/")

rep2_rna<- paste(cbs_02dec2021_rna_qual_dir, "IH_26JAN2022_CBS_Eukaryote Total RNA Pico_DE13806310_2022-01-26_17-00-20_Sample7.csv", sep="/")

rep3_rna<- paste(cbs_02dec2021_rna_qual_dir, "IH_26JAN2022_CBS_Eukaryote Total RNA Pico_DE13806310_2022-01-26_17-00-20_Sample8.csv", sep="/")

rep1_dt<- fread(rep1_rna, sep=",", skip=17)
rep2_dt<- fread(rep2_rna, sep=",", skip=17)
rep3_dt<- fread(rep3_rna, sep=",", skip=17)
rna_qual_all_dt<- rbind(rep1_dt, rep2_dt, rep3_dt)

rna_qual_all_dt[,Rep:=as.factor(rep(c(
  "Rep 1", "Rep 2", "Rep 3"), 
  times=c(nrow(rep1_dt), nrow(rep2_dt), nrow(rep3_dt))))]

rm(list=c("rep1_dt", "rep2_dt", "rep3_dt"))

rna_qual_all_filt_dt<- rna_qual_all_dt[Time > 35 & Time < 55]
rna_qual_all_filt_dt[,Proportion_FU:=Value/sum(Value), by=.(Rep)]

rna_qual_all_gg<- ggplot(
  data=rna_qual_all_filt_dt, 
  aes(x=Time, y=Proportion_FU, col=Rep))

rna_qual_all_gg+
  geom_line()+
  expand_limits(y=0)+
  scale_color_npg()+
  labs(x="Time", y="Normalized fluorescence")+
  theme_cowplot()
```

```{r Library prep- amplicon PCR1 landing pad amplification}

cbs_02dec2021_pcr1_dir<- paste(base_dir, "CBS_02DEC2021", "Library_prep", "Amplicon", "PCR1", sep="/")
pcr1_csv_files<- dir(cbs_02dec2021_pcr1_dir, pattern=".csv", full.names=TRUE)
pcr1_csv_filenames<- dir(cbs_02dec2021_pcr1_dir, pattern=".csv")

pcr1_csv_list<- lapply(pcr1_csv_files, fread, skip=17)
pcr1_csv_list_nrows<- sapply(pcr1_csv_list, nrow)

pcr1_csv_all_dt<- rbindlist(pcr1_csv_list)

pcr1_csv_all_dt[,Sample:=as.factor(rep(
  pcr1_csv_filenames, times=pcr1_csv_list_nrows))]

pcr1_csv_all_dt[,Input:=as.factor(rep(
  c("gDNA", "cDNA"), times=c(
    sum(pcr1_csv_list_nrows[1:3]),
    sum(pcr1_csv_list_nrows[4:6]))))]

rm(pcr1_csv_list)

pcr1_csv_all_filt_dt<- pcr1_csv_all_dt[Time > 80 & Time < 120]
pcr1_csv_all_filt_dt[,Proportion_FU:=Value/sum(Value), by=.(Sample, Input)]

pcr1_csv_all_filt_summ_dt<- pcr1_csv_all_filt_dt[
  ,.(Mean_prop_FU=mean(Proportion_FU)), by=.(Input, Time)]

pcr1_csv_all_filt_summ_gg<- ggplot(
  data=pcr1_csv_all_filt_summ_dt, 
  aes(x=Time, y=Mean_prop_FU, col=Input))

pcr1_csv_all_filt_summ_gg+
  geom_line()+
  expand_limits(y=0)+
  scale_color_aaas()+
  scale_y_continuous(breaks=seq(0,0.075,0.025))+
  labs(x="Time", y="Normalized fluorescence")+
  theme_cowplot()

```

```{r Library prep- amplicon PCR3 final library}

cbs_02dec2021_flq_dir<- paste(base_dir, "CBS_02DEC2021", "Library_prep", "Amplicon", "PCR3", sep="/")
flq_csv_files<- dir(cbs_02dec2021_flq_dir, pattern=".csv", full.names=TRUE)
flq_csv_filenames<- dir(cbs_02dec2021_flq_dir, pattern=".csv")

flq_csv_list<- lapply(flq_csv_files, fread, skip=17)
flq_csv_list_nrows<- sapply(flq_csv_list, nrow)

flq_csv_all<- rbindlist(flq_csv_list)

flq_csv_all[,Sample:=as.factor(rep(
  flq_csv_filenames, times=flq_csv_list_nrows))]

flq_csv_all[,Input:=as.factor(rep(
  c("cDNA", "gDNA", "Negative_control", "gDNA-AMP"), times=c(
    sum(flq_csv_list_nrows[1:3]),
    sum(flq_csv_list_nrows[4:6]),
    flq_csv_list_nrows[7], flq_csv_list_nrows[8])))]

rm(flq_csv_list)

flq_csv_amplicon_filt_dt<- flq_csv_all[Input!="gDNA-AMP" & Time > 45 & Time < 95]
flq_csv_amplicon_filt_dt[,Proportion_FU:=Value/sum(Value), by=.(Sample, Input)]

flq_csv_amplicon_filt_summ_dt<- flq_csv_amplicon_filt_dt[
  ,.(Mean_prop_FU=mean(Proportion_FU)), by=.(Input, Time)]

flq_csv_amplicon_filt_summ_gg<- ggplot(
  data=flq_csv_amplicon_filt_summ_dt, 
  aes(x=Time, y=Mean_prop_FU, col=Input))

flq_csv_amplicon_filt_summ_gg+
  geom_line()+
  expand_limits(y=0)+
  scale_color_aaas()+
  labs(x="Time", y="Normalized fluorescence")+
  theme_cowplot()


```

```{r Library prep- RACE-like final library}

cbs_02dec2021_racelike_flq_dir<- paste(base_dir, "CBS_02DEC2021", "Library_prep", "RACElike", sep="/")
racelike_flq_csv_files<- dir(cbs_02dec2021_racelike_flq_dir, pattern=".csv", full.names=TRUE)
racelike_flq_csv_filenames<- dir(cbs_02dec2021_racelike_flq_dir, pattern=".csv")

racelike_flq_csv_list<- lapply(racelike_flq_csv_files, fread, skip=17)
racelike_flq_csv_list_nrows<- sapply(racelike_flq_csv_list, nrow)

racelike_flq_csv_all_dt<- rbindlist(racelike_flq_csv_list)

racelike_flq_csv_all_dt[,Sample:=as.factor(rep(
  racelike_flq_csv_filenames, times=racelike_flq_csv_list_nrows))]

racelike_flq_csv_all_dt[,Condition:=as.factor(rep(
  c("Non-size-selected", "Size-selected"), 
  times=c(racelike_flq_csv_list_nrows[1], racelike_flq_csv_list_nrows[2])))]

rm(racelike_flq_csv_list)

racelike_flq_csv_amplicon_filt_dt<- racelike_flq_csv_all_dt[Time > 50 & Time < 120]
racelike_flq_csv_amplicon_filt_dt[,Proportion_FU:=Value/sum(Value), by=.(Sample, Condition)]

racelike_flq_csv_amplicon_filt_gg<- ggplot(
  data=racelike_flq_csv_amplicon_filt_dt, 
  aes(x=Time, y=Proportion_FU, col=Condition))

racelike_flq_csv_amplicon_filt_gg+
  geom_line()+
  expand_limits(y=0)+
  scale_color_jco()+
  scale_x_continuous(breaks=seq(50,120,10))+
  scale_y_continuous(breaks=seq(0,0.012,0.002))+
  coord_cartesian(xlim=c(50,120))+
  labs(x="Time", y="Normalized fluorescence")+
  theme_cowplot()


```

```{r CBS clinically relevant mutations}

sun_weile_roth_2020_roi<- seq(328,453)
remediable_residues<- c(50,197,266,289,312)

# We can also plot some of the important residues interfacing with cofactors
heme_domain<- seq(1,69)
catalytic_domain<- seq(70,381)
cbs_linker<- seq(382,411)
cbs_linker_truncating<- c(409,410)  # check for nonsense
bateman_module<- seq(412,551)
heme_binding<- c(52,65)
plp_schiff_base<- c(119)
plp_binding<- c(149,349)
plp_binding2<- seq(256,260)
tetramerization_req<- seq(516,525)
dimerization_interface<- c(111,112)
dimerization_interface2<- c(77,80,87,90,92,152,156,160,180,183,184,339,340,344,35,378,382,386)

catalytic_residues<- c(305, 307)
ptm_residues<- c(27, 119, 199, 211)
uniprot_domain<- seq(418, 476)
uniprot_disordered<- seq(1, 74)
uniprot_acidic_basic<- seq(28, 44)
uniprot_reduced_heme_activity<- c(272, 275)

uniprot_variants<- c(49, 58, 65, 78, 85, 87, 88, 101, 102, 109, 114, 116, 121, 125, 126, 128, 131, 139, 143, 144, 145, 148, 151, 152, 154, 155, 165, 168, 173, 176, 180, 191, 198, 200, 224, 226, 228, 231, 234, 239, 257, 262, 266, 269, 270, 275, 278, 281, 288, 290, 302, 305, 307, 320, 321, 331, 336, 338, 347, 349, 352, 353, 354, 355, 361, 369, 370, 371, 376, 379, 384, 391, 422, 427, 434, 435, 439, 444, 446, 449, 454, 456, 466, 491, 500, 526, 534, 539, 540)

# A360A, C699T, I278T, N212N, and T42N
important_cbs_pos<- c(heme_binding, plp_binding, plp_binding2, tetramerization_req)
multimer_pos<- c(dimerization_interface, dimerization_interface2, tetramerization_req)
cbs_variants_pos<- c(uniprot_variants, uniprot_reduced_heme_activity, ptm_residues)
slop_int<- 2

 # Casique et al., 2013
# (c.260C>A (p.T87N)) and a previously reported variant (c.700G>A (p.D234N))
# Lower expression of CBS protein
casique_vars<- c("p.T87N", "p.D234N")
urreizti_path_changes<- c("p.179M",  "p.C257Y", "p.L338P", "p.S349N", "p.R379Q", "p.L456P")
urreizti_unc_changes<- c("p.I278T", "p.G307S")

important_cbs_pos_slop<- unique(
  c(important_cbs_pos, important_cbs_pos - slop_int, important_cbs_pos + slop_int))

agg_prone_alleles<- c("p.T87N", "p.A114V", "p.A155T", "p.E176K", "p.I278T")

```

```{r CBS Uniprot mutations}

cbs_uniprot_mut_file<- paste(base_dir, "CBS_02DEC2021", "Uniprot_CBS_variants.txt", sep="/")
cbs_uniprot_mut_dt<- fread(cbs_uniprot_mut_file, sep="\t", header=T)

# Determine amino acid types
positive_charged_aas<- c("R", "H", "K")
negative_charged_aas<- c("D", "E")
polar_uncharged_aas<- c("S", "T", "N", "Q") # C
nonpolar_uncharged_aas<- c("A", "V", "I", "L", "M", "F", "Y", "W") # G, P
other_aas<- c("C", "G", "P")

cbs_uniprot_mut_dt[,`:=`(REF_type=get_aa_change_type(REF_AA), 
                         ALT_type=get_aa_change_type(ALT_AA)), 
                   by=seq_len(nrow(cbs_uniprot_mut_dt))]

cbs_uniprot_mut_dt[,Conserved:=FALSE]
cbs_uniprot_mut_dt[REF_type==ALT_type, Conserved:=TRUE]
cbs_uniprot_mut_dt[,Conserved:=as.factor(Conserved)]

cbs_uniprot_mut_summ_dt<- cbs_uniprot_mut_dt[!(is.na(REF_AA) | is.na(ALT_AA)), .N, by=.(Positions, Conserved)]
cbs_uniprot_mut_summ_dt[,Positions:=as.integer(Positions)]

cbs_uniprot_mut_gg<- ggplot(
  data=cbs_uniprot_mut_summ_dt, 
  aes(x=Positions, y=N, col=Conserved))

cbs_uniprot_mut_gg+
  scale_x_continuous(breaks=seq(0,550, 50))+
  scale_color_aaas()+
  geom_point()+
  geom_segment(aes(x=Positions, xend=Positions, y=0, yend=N))+
  theme_cowplot()

```

```{r CBS libraries}

cbs_02dec2021_dir<- paste(base_dir, "CBS_02DEC2021", sep="/")
cbs_02dec2021_dt<- read_vcf_summary_files(cbs_02dec2021_dir)

amplicon_tileseq_samples<- paste("CBS1", seq(68, 75), "R", sep="_")
amp_racelike_samples<- paste("CBS1", seq(76, 82), "R", sep="_")

cdna_samples<- c(paste("CBS1", seq(68, 70), "R", sep="_"),
                 paste("CBS1", seq(79, 81), "R", sep="_"))

gdna_samples<- c(paste("CBS1", seq(71, 73), "R", sep="_"),
                 paste("CBS1", seq(76, 78), "R", sep="_"))

nc_samples<- paste("CBS1", c(75, 82), "R", sep="_")
input_lib_samples<- paste("CBS1", 74, "R", sep="_")

gdna_samples_replace<- copy(gdna_samples)
gdna_samples_replace[3]<- "CBS1_74_R"

cbs_02dec2021_dt[Sample%in%amplicon_tileseq_samples, Chemistry:="Amplicon"]
cbs_02dec2021_dt[Sample%in%amp_racelike_samples, Chemistry:="RACE-like"]

cbs_02dec2021_dt[Sample%in%cdna_samples, Input:="cDNA"]
cbs_02dec2021_dt[Sample%in%gdna_samples, Input:="gDNA"]
cbs_02dec2021_dt[Sample%in%nc_samples, Input:="Negative_control"]
cbs_02dec2021_dt[Sample%in%input_lib_samples, Input:="Input_library"]

cbs_02dec2021_dt[,Input:=as.factor(Input)]
cbs_02dec2021_dt[,Chemistry:=as.factor(Chemistry)]

cbs_02dec2021_amplicon_dt<- cbs_02dec2021_dt[Chemistry=="Amplicon"]
cbs_02dec2021_racelike_dt<- cbs_02dec2021_dt[Chemistry=="RACE-like"]
cbs_02dec2021_racelike_dt[,R1_AO_diff:=abs(R1_PLUS_AO-R1_MINUS_AO)]

# Amplicon
# Only count variants on the expected strand for amplicon method
cbs_02dec2021_amplicon_filt_dt<- cbs_02dec2021_amplicon_dt[!R1_MINUS_AO>0]
cbs_02dec2021_amplicon_filt_dt[,CAO:=R1_PLUS_AO]
cbs_02dec2021_amplicon_filt_dt[,VAR_ID:=as.factor(paste(`#CHROM`,POS,REF,ALT,sep=":"))]
cbs_02dec2021_amplicon_filt_dt[,CAF:=max(CAO/DP), by=.(Input, Sample, VAR_ID)]

cbs_02dec2021_amplicon_postprocess_dt<- postprocess_vcf_summary_dt(
  cbs_02dec2021_amplicon_filt_dt, key_features=c("Sample", "Input", "VAR_ID"))

cbs_02dec2021_amplicon_postprocess_subtract_dt<- subtract_af(
  cbs_02dec2021_amplicon_postprocess_dt)

# Filter variants outside of the amplicons
amplicon_tile2_pos<- seq(4648, 4797)
amplicon_tile4_pos<- seq(4864, 5018)
amplicon_tile2_4_pos<- c(amplicon_tile2_pos, amplicon_tile4_pos)

amplicon_tile2_target<- seq(4670, 4771)
amplicon_tile4_target<- seq(4886, 5000)
amplicon_tile2_4_target<- unique(c(amplicon_tile2_target, amplicon_tile4_target))

cbs_02dec2021_amplicon_postprocess_subtract_filt_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_dt[
    LOCATION=="CDS" & POS%in%amplicon_tile2_4_pos & 
      MATCHES_MUT_SIG=="True" & (TYPE!="SNP" | CAO>1)]

# RACE-like
cbs_02dec2021_racelike_postprocess_dt<- postprocess_vcf_summary_dt(
  cbs_02dec2021_racelike_dt, key_features=c("Sample", "Input", "VAR_ID"))

cbs_02dec2021_racelike_postprocess_subtract_dt<- subtract_af(
  cbs_02dec2021_racelike_postprocess_dt)

cbs_02dec2021_racelike_postprocess_subtract_filt_dt<- 
  cbs_02dec2021_racelike_postprocess_subtract_dt[
    LOCATION=="CDS" & MATCHES_MUT_SIG=="True" & (TYPE!="SNP" | CAO>1)]

# Assess strand-specific errors or sample-strand specific errors
# For the terminal tiles, we expect strand bias because the original 
# custom CBS primer designs did not have a primer for the new stable expression
# vector (pDEST_HC_Rec_Bxb_v2)
# Thus only filter out variants with high strand bias in regions that
# are bidirectionally covered
cbs_02dec2021_racelike_postprocess_subtract_filt_dt[
  ,R1_AO_diff:=log2(abs(R1_PLUS_AO-R1_MINUS_AO))]

cbs_02dec2021_racelike_postprocess_subtract_filt_dt[,POS:=as.integer(POS)]

racelike_terminal_tile_pos<- c(seq(4562, 4670), seq(6125, 6217))

cbs_02dec2021_racelike_postprocess_subtract_filt_dt[
    POS%in%racelike_terminal_tile_pos, Terminal_tile:=TRUE]

cbs_02dec2021_racelike_postprocess_subtract_filt_dt[
    is.na(Terminal_tile), Terminal_tile:=FALSE]

cbs_02dec2021_racelike_postprocess_subtract_filt_dt[
  , Terminal_tile:=as.factor(Terminal_tile)]

strand_diff_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_dt, 
  aes(x=R1_AO_diff, col=Terminal_tile))

strand_diff_gg+
  geom_density()+
  labs(x="log2 abs strand diff", col="Terminal tile")+
  geom_vline(xintercept=6, col="grey", linetype=2)+
  theme_cowplot()

cbs_02dec2021_racelike_postprocess_subtract_filt_dt<-
      cbs_02dec2021_racelike_postprocess_subtract_filt_dt[
        R1_AO_diff <= 6]

cbs_02dec2021_racelike_postprocess_subtract_filt_dt[,Terminal_tile:=NULL]

# Combine datasets
cbs_02dec2021_both_postprocess_subtract_dt<- rbind(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_dt, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_dt, fill=TRUE)

cbs_02dec2021_both_postprocess_subtract_dt[,Chemistry:=as.factor(
  rep(c("Amplicon", "RACE-like"), times=c(
    nrow(cbs_02dec2021_amplicon_postprocess_subtract_filt_dt), 
    nrow(cbs_02dec2021_racelike_postprocess_subtract_filt_dt))))]

```

```{r CBS libraries RF filtering}

set.seed(9)

cbs_02dec2021_ec_dir<- paste(base_dir, "error_correction", "cenik_cbs", "CBS_02DEC2021", sep="/")
cbs_02dec2021_ec_dt<-  read_vcf_summary_files(cbs_02dec2021_ec_dir)
cbs_02dec2021_ec_dt[,R1_AO_diff:=abs(R1_PLUS_AO-R1_MINUS_AO)]

cbs_02dec2021_ec_75_truth_vcf<- paste(cbs_02dec2021_ec_dir, "CBS1_75_pDEST_train.truth.vcf", sep="/")
cbs_02dec2021_ec_82_truth_vcf<- paste(cbs_02dec2021_ec_dir, "CBS1_82_pDEST_train.truth.vcf", sep="/")

cbs_02dec2021_ec_75_truth_dt<- fread(cbs_02dec2021_ec_75_truth_vcf, skip=20, header=T, sep="\t")
cbs_02dec2021_ec_82_truth_dt<- fread(cbs_02dec2021_ec_82_truth_vcf, skip=20, header=T, sep="\t")

# Prep training dataset for amplicon
modeling_features<- c(
  "VAR_ID", "SUBST", "UP_REF_NT", "DOWN_REF_NT", 
  "TYPE", "log10_CAF", "MATCHES_MUT_SIG",
  "R1_PLUS_MED_RP", "R2_MINUS_MED_RP", 
  "R1_PLUS_MED_BQ", "R2_MINUS_MED_BQ", 
  "R1_PLUS_MED_NM", "R2_MINUS_MED_NM")

# Because the NM and RP are collinear between R1 and R2, just select the stat for one read
modeling_features_no_collinear<- modeling_features[
  !grepl("R1_PLUS_MED_RP|R1_PLUS_MED_NM", modeling_features)]

# Prep the datasets for modeling
cbs_02dec2021_ec_75_dt<- prep_train_dt(
  train_dt=cbs_02dec2021_ec_dt[
    Sample=="CBS1_75_pDEST_train" & POS%in%amplicon_tile2_4_pos], 
  truth_dt=cbs_02dec2021_ec_75_truth_dt, 
  modeling_features=modeling_features_no_collinear)

# Prep training dataset for RACE-like
modeling_features_racelike<- c(
  "VAR_ID", "SUBST", "UP_REF_NT", "DOWN_REF_NT", 
  "TYPE", "log10_CAF", "MATCHES_MUT_SIG", "R1_AO_diff",
  "R1_PLUS_MED_RP", "R2_MINUS_MED_RP", "R2_PLUS_MED_RP", "R1_MINUS_MED_RP", 
  "R1_PLUS_MED_BQ", "R2_MINUS_MED_BQ", "R2_PLUS_MED_BQ", "R1_MINUS_MED_BQ",
  "R1_PLUS_MED_NM", "R2_MINUS_MED_NM", "R2_PLUS_MED_NM", "R1_MINUS_MED_NM")

# Because the NM and RP are collinear between R1 and R2, just select the stat for one read
modeling_features_no_collinear_racelike<- modeling_features_racelike[
  !grepl("R1_PLUS_MED_RP|R1_PLUS_MED_NM|R1_MINUS_MED_RP|R1_MINUS_MED_NM", modeling_features_racelike)]

options(na.action="na.pass")

cbs_02dec2021_ec_82_dt<- prep_train_dt(
  train_dt=cbs_02dec2021_ec_dt[Sample=="CBS1_82_pDEST_train"], 
  truth_dt=cbs_02dec2021_ec_82_truth_dt, 
  modeling_features=modeling_features_racelike)

options(na.action="na.omit")

prob_cutoff<- 0.49

# Amplicon model
cbs_02dec2021_ec_75_folds<- create_kfolds(cbs_02dec2021_ec_75_dt)

cbs_02dec2021_ec_75_train_dt<- cbs_02dec2021_ec_75_dt[!cbs_02dec2021_ec_75_folds[[1]]]
cbs_02dec2021_ec_75_test_dt<- cbs_02dec2021_ec_75_dt[cbs_02dec2021_ec_75_folds[[1]]]

#cbs_02dec2021_ec_75_train_cv_res<- run_nested_cv(in_dt=cbs_02dec2021_ec_75_train_dt, hyperparam_prop=1.0)

cbs_02dec2021_ec_75_train_rf<- randomForest(Truth~., cbs_02dec2021_ec_75_train_dt, mtry=9)

cbs_02dec2021_ec_75_train_predict<- predict(
  cbs_02dec2021_ec_75_train_rf, cbs_02dec2021_ec_75_test_dt, type="prob")

cbs_02dec2021_ec_75_train_predict_labels<- factor(
  cbs_02dec2021_ec_75_train_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)

cbs_02dec2021_ec_75_test_contigency_table<- create_cont_table(
  cbs_02dec2021_ec_75_test_dt[,Truth], cbs_02dec2021_ec_75_train_predict_labels)

cbs_02dec2021_ec_75_test_acc<- get_accuracy(cbs_02dec2021_ec_75_test_contigency_table)
cbs_02dec2021_ec_75_test_recall<- get_sensitivity(cbs_02dec2021_ec_75_test_contigency_table)
cbs_02dec2021_ec_75_test_prec<- get_precision(cbs_02dec2021_ec_75_test_contigency_table)


# RACE-like model
cbs_02dec2021_ec_82_folds<- create_kfolds(cbs_02dec2021_ec_82_dt)

cbs_02dec2021_ec_82_train_dt<- cbs_02dec2021_ec_82_dt[!cbs_02dec2021_ec_82_folds[[1]]]
cbs_02dec2021_ec_82_test_dt<- cbs_02dec2021_ec_82_dt[cbs_02dec2021_ec_82_folds[[1]]]

#cbs_02dec2021_ec_82_train_cv_res<- run_nested_cv(in_dt=cbs_02dec2021_ec_82_train_dt)

# For RACE-like libraries, since counts may be absent from one sample strand, there may be NAs
# in the data; use na.roughfix for imputation 
cbs_02dec2021_ec_82_train_rf<- randomForest(
  Truth~., cbs_02dec2021_ec_82_train_dt, mtry=8, na.action=na.roughfix)

cbs_02dec2021_ec_82_train_predict<- predict(
  cbs_02dec2021_ec_82_train_rf, cbs_02dec2021_ec_82_test_dt, type="prob")

cbs_02dec2021_ec_82_train_predict_labels<- factor(
  cbs_02dec2021_ec_82_train_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)

cbs_02dec2021_ec_82_test_contigency_table<- create_cont_table(
  cbs_02dec2021_ec_82_test_dt[,Truth], cbs_02dec2021_ec_82_train_predict_labels)

cbs_02dec2021_ec_82_test_acc<- get_accuracy(cbs_02dec2021_ec_82_test_contigency_table)
cbs_02dec2021_ec_82_test_recall<- get_sensitivity(cbs_02dec2021_ec_82_test_contigency_table)
cbs_02dec2021_ec_82_test_prec<- get_precision(cbs_02dec2021_ec_82_test_contigency_table)


# Now filter the datasets with the models and compare correlation between replicates
# with and without filtering
# Generate a final model with the whole simulation dataset

cbs_02dec2021_ec_75_final_rf<- randomForest(Truth~., cbs_02dec2021_ec_75_dt, mtry=9)

# Additionally ensure removal of non-CDS variants at those outside the amplicons
cbs_02dec2021_amplicon_filt_prep_dt<- prep_train_dt(
  train_dt=cbs_02dec2021_amplicon_filt_dt[LOCATION=="CDS" & POS%in%amplicon_tile2_4_pos],
  modeling_features=modeling_features_no_collinear)

cbs_02dec2021_amplicon_filt_prep_predict<- predict(
  cbs_02dec2021_ec_75_final_rf, cbs_02dec2021_amplicon_filt_prep_dt, type="prob")

cbs_02dec2021_amplicon_filt_prep_predict_labels<- factor(
  cbs_02dec2021_amplicon_filt_prep_predict[,2]>=prob_cutoff, 
  levels=cont_table_factor_levels)

cbs_02dec2021_amplicon_filt_prep_dt[
  ,Prediction:=cbs_02dec2021_amplicon_filt_prep_predict_labels]

# Count false positive predictions for input variant library, gDNA, cDNA samples only
cbs_02dec2021_amplicon_filt_prep_pred_fp_counts_dt<- 
  cbs_02dec2021_amplicon_filt_prep_dt[
    Prediction=="FALSE" & Sample!="CBS1_75_R", .N, by=.(VAR_ID, TYPE)]

# Select variants that are predicted false in all samples
cbs_02dec2021_amplicon_filt_prep_fp_varids<- 
  cbs_02dec2021_amplicon_filt_prep_pred_fp_counts_dt[
    (TYPE=="SNP" & N == 7) | (TYPE=="di_nt_MNP" & N == 14) | 
      (TYPE=="tri_nt_MNP" & N == 21), unique(VAR_ID)]

#c109g_varid<- cbs_02dec2021_amplicon_filt_prep_dt[,which(VAR_ID=="pDEST_HC_Rec_Bxb_v2_CBS_WT:4886:T:G")]

#cbs_02dec2021_amplicon_filt_prep_predict[c109g_varid,]

# RACE-like filtering
cbs_02dec2021_ec_82_final_rf<- randomForest(
  Truth~., cbs_02dec2021_ec_82_dt[
    ,-c("R1_PLUS_MED_RP", "R1_PLUS_MED_NM", "R1_MINUS_MED_RP", "R1_MINUS_MED_NM")], 
  mtry=8, na.action=na.roughfix)

options(na.action="na.pass")

cbs_02dec2021_racelike_filt_prep_dt<- prep_train_dt(
  train_dt=cbs_02dec2021_racelike_dt[LOCATION=="CDS"],
  modeling_features=modeling_features_no_collinear_racelike)

options(na.action="na.omit")

cbs_02dec2021_racelike_filt_prep_predict<- predict(
  cbs_02dec2021_ec_82_final_rf, cbs_02dec2021_racelike_filt_prep_dt, type="prob")

cbs_02dec2021_racelike_filt_prep_predict_labels<- factor(
  cbs_02dec2021_racelike_filt_prep_predict[,2]>=prob_cutoff, 
  levels=cont_table_factor_levels)

cbs_02dec2021_racelike_filt_prep_dt[
  ,Prediction:=cbs_02dec2021_racelike_filt_prep_predict_labels]

# Note that for RACE-like data, some prediction labels are NA;
# this is due to NA data for one sample strand; these are either
# 1) variants that are low in abundance and only observed on one 
# sample strand by chance; 2) sample-strand specific chemical changes
# For this dataset, since there were no vector primers in the 
# custom primer mix from ArcherDX for the landing pad vector,
# terminal PCR tiles are guaranteed to only show counts from
# one strand (emanating from GSP2 primer read towards the vector junction)

# Filtering strategy: filter predicted false variants, and NA predictions
# for SNPs that are not in the terminal PCR tiles (keep MNPs observed on
# one sample strand as these are more likely to be true variants)

# Count false positive predictions for input variant library, gDNA, cDNA samples only
cbs_02dec2021_racelike_filt_prep_pred_fp_counts_dt<- 
  cbs_02dec2021_racelike_filt_prep_dt[
    Sample!="CBS1_82_R" & Prediction=="FALSE", 
    .N, by=.(VAR_ID, TYPE)]

# Select variants that are predicted false in all samples
cbs_02dec2021_racelike_filt_prep_fp_varids<- 
  cbs_02dec2021_racelike_filt_prep_pred_fp_counts_dt[
    (TYPE=="SNP" & N == 6) | (TYPE=="di_nt_MNP" & N == 12) | 
      (TYPE=="tri_nt_MNP" & N == 18), unique(VAR_ID)]

# Add in SNPs that are only found in one strand and are not in terminal tiles (see filtering above)
# Note this is a compromise due to the primer design and the terminal tiles may see higher rates
# of false positives compared to other tiles
# cbs_02dec2021_racelike_filt_prep_dt[(is.na(Prediction) & TYPE=="SNP"), unique(VAR_ID)]
#cbs_02dec2021_racelike_filt_prep_fp_varids<- c(cbs_02dec2021_racelike_filt_prep_fp_varids)

# Now that we have the likely false positives, 
# filter them out of the MNP-collapsed and 
# negative control library freq-subtracted dt
cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_dt[
    !VAR_ID%in%cbs_02dec2021_amplicon_filt_prep_fp_varids]

cbs_02dec2021_racelike_postprocess_subtract_filt_rf_dt<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_dt[
    !VAR_ID%in%cbs_02dec2021_racelike_filt_prep_fp_varids]
  
cbs_02dec2021_both_postprocess_subtract_rf_dt<- rbind(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_dt, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_rf_dt, fill=TRUE)

cbs_02dec2021_both_postprocess_subtract_rf_dt[,Chemistry:=as.factor(
  rep(c("Amplicon", "RACE-like"), times=c(
    nrow(cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_dt), 
    nrow(cbs_02dec2021_racelike_postprocess_subtract_filt_rf_dt))))]

cbs_02dec2021_both_postprocess_subtract_rf_dt[,log_CAF:=log(CAF)]

cbs_02dec2021_both_postprocess_subtract_rf_drop_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_dt[Sample!="CBS1_73_R"]

cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[Sample=="CBS1_74_R", Input:="gDNA"]

```

```{r CBS RF filtering comparison}

# Amplicon
# Remove the gDNA outlier and use the plasmid library in place for it
cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_dt<- 
  rbind(cbs_02dec2021_amplicon_postprocess_subtract_filt_dt[Sample!="CBS1_73_R"], 
        cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_dt[Sample!="CBS1_73_R"], 
        fill=TRUE)

cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_dt[Sample=="CBS1_74_R", Input:="gDNA"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_dt[,Filter:=as.factor(
  rep(c("Pre-filter", "Post-filter"), times=c(
    nrow(cbs_02dec2021_amplicon_postprocess_subtract_filt_dt[Sample!="CBS1_73_R"]), 
    nrow(cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_dt[Sample!="CBS1_73_R"]))))]

cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_dt[
  TYPE=="SNP", TYPE_SHORT:="SNP"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_dt[
  is.na(TYPE_SHORT), TYPE_SHORT:="MNP"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_dt[
  ,TYPE_SHORT:=factor(TYPE_SHORT, levels=c("SNP", "MNP"))]

cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_counts_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_dt[
    ,.(N=length(unique(VAR_ID))), by=.(Chemistry, Input, TYPE_SHORT, Filter, Sample)]

# RACE-like
cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_dt<- 
  rbind(cbs_02dec2021_racelike_postprocess_subtract_filt_dt, 
        cbs_02dec2021_racelike_postprocess_subtract_filt_rf_dt, fill=TRUE)

cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_dt[,Filter:=as.factor(
  rep(c("Pre-filter", "Post-filter"), times=c(
    nrow(cbs_02dec2021_racelike_postprocess_subtract_filt_dt), 
    nrow(cbs_02dec2021_racelike_postprocess_subtract_filt_rf_dt))))]

cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_dt[
  TYPE=="SNP", TYPE_SHORT:="SNP"]

cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_dt[
  is.na(TYPE_SHORT), TYPE_SHORT:="MNP"]

cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_dt[
  ,TYPE_SHORT:=factor(TYPE_SHORT, levels=c("SNP", "MNP"))]

cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_counts_dt<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_dt[
    ,.(N=length(unique(VAR_ID))), by=.(Chemistry, Input, TYPE_SHORT, Filter, Sample)]

# Combine and plot
cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_dt<- 
  rbind(cbs_02dec2021_amplicon_postprocess_subtract_filt_rf_comp_counts_dt,
        cbs_02dec2021_racelike_postprocess_subtract_filt_rf_comp_counts_dt)

cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_dt<- 
  cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_dt[
    ,.(Mean_N=mean(N), SD_N=sd(N)), by=.(Chemistry, Input, TYPE_SHORT, Filter)]

cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_dt[
  ,Filter:=factor(Filter, levels=c("Pre-filter", "Post-filter"))]

cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_amplicon_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_dt[Chemistry=="Amplicon"], 
  aes(x=TYPE_SHORT, y=Mean_N, ymax=Mean_N+SD_N, ymin=Mean_N-SD_N, col=Filter))

cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_amplicon_gg+
  facet_wrap(~Input)+
  expand_limits(y=0)+
  geom_point(position=position_dodge(0.9), size=1.5)+
  geom_errorbar(position=position_dodge(0.9), width=0.1)+
  scale_y_continuous(breaks=seq(0,1000,250))+
  coord_cartesian(ylim=c(0,1000))+
  scale_color_d3()+
  labs(x="Variant type", y="Variants", col="RF model")+
  theme_cowplot()

cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_racelike_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_dt[Chemistry=="RACE-like"], 
  aes(x=TYPE_SHORT, y=Mean_N, ymax=Mean_N+SD_N, ymin=Mean_N-SD_N, col=Filter))

cbs_02dec2021_both_postprocess_subtract_filt_rf_comp_counts_summ_racelike_gg+
  facet_wrap(~Input)+
  expand_limits(y=0)+
  geom_point(position=position_dodge(0.9), size=1.5)+
  geom_errorbar(position=position_dodge(0.9), width=0.1)+
  scale_y_continuous(breaks=seq(0,3500,500))+
  coord_cartesian(ylim=c(0,3700))+
  scale_color_d3()+
  labs(x="Variant type", y="Variants", col="RF model")+
  theme_cowplot()

```

```{r Generate WT-normalized log frequencies}

cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[CAO>=0]

cbs_02dec2021_both_postprocess_subtract_rf_drop_wt_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    ,.(CAO_sum=sum(CAO), DP=unique(DP)), by=.(Sample, POS)]

cbs_02dec2021_both_postprocess_subtract_rf_drop_wt_dt[,WT_CAO:=DP-CAO_sum]

# Only measure counts in the target regions
cbs_02dec2021_both_postprocess_subtract_rf_drop_wt_summ_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_wt_dt[
    POS%in%seq(4562, 6217),.(WT_CAO=unique(WT_CAO)), by=.(Sample, POS)]

cbs_02dec2021_both_postprocess_subtract_rf_drop_wt_summ_dt[,Key:=paste(Sample, POS, sep=":")]

cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[,Key:=paste(Sample, POS, sep=":")]

# Generate a WT-normalized log ratio
cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
  ,WT_CAO:=cbs_02dec2021_both_postprocess_subtract_rf_drop_wt_summ_dt[
    match(cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[,Key], Key), WT_CAO]]

# log_CAF_norm:=log(CAO/DP) - log(WT_CAO/DP)
cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
  ,log_CAF_norm:=log(CAO+0.5) - log(WT_CAO+0.5)]

cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[!is.na(log_CAF_norm)]

```

```{r Mutation type comparison and positional effects}

cbs_02dec2021_both_postprocess_subtract_rf_drop_counts_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    ,.N, by=.(Chemistry, VAR_ID)]

cbs_02dec2021_amplicon_postprocess_subtract_rf_drop_counts_nrows<- 
  nrow(cbs_02dec2021_both_postprocess_subtract_rf_drop_counts_dt[Chemistry=="Amplicon"])

cbs_02dec2021_racelike_postprocess_subtract_rf_drop_counts_nrows<- 
  nrow(cbs_02dec2021_both_postprocess_subtract_rf_drop_counts_dt[Chemistry=="RACE-like"])

amplicon_singletons<- cbs_02dec2021_both_postprocess_subtract_rf_drop_counts_dt[
  Chemistry=="Amplicon" & N==1, unique(VAR_ID)]

# length(amplicon_singletons)/cbs_02dec2021_amplicon_postprocess_subtract_rf_drop_counts_nrows

racelike_singletons<- cbs_02dec2021_both_postprocess_subtract_rf_drop_counts_dt[
  Chemistry=="RACE-like" & N==1, unique(VAR_ID)]

# length(racelike_singletons)/cbs_02dec2021_racelike_postprocess_subtract_rf_drop_counts_nrows

amplicon_complete<- cbs_02dec2021_both_postprocess_subtract_rf_drop_counts_dt[
  Chemistry=="Amplicon" & N==6, unique(VAR_ID)]

# length(amplicon_complete)/cbs_02dec2021_amplicon_postprocess_subtract_rf_drop_counts_nrows

racelike_complete<- cbs_02dec2021_both_postprocess_subtract_rf_drop_counts_dt[
  Chemistry=="RACE-like" & N==6, unique(VAR_ID)]

# length(racelike_complete)/cbs_02dec2021_racelike_postprocess_subtract_rf_drop_counts_nrows

cbs_02dec2021_both_postprocess_subtract_rf_mut_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    (Chemistry=="Amplicon" & VAR_ID%in%amplicon_complete) | 
      (Chemistry=="RACE-like" & VAR_ID%in%racelike_complete) ,
    .(Median_log_CAF_norm=median(log_CAF_norm, na.rm=TRUE)),
    by=.(Chemistry, Input, VAR_ID, TYPE, VAR_TYPE, AA_POS)]

cbs_02dec2021_both_postprocess_subtract_rf_mut_dt[
  ,AA_POS_int:=as.integer(AA_POS)]

cbs_02dec2021_both_postprocess_subtract_rf_mut_cast_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_mut_dt, 
  Chemistry+VAR_ID+TYPE+VAR_TYPE+AA_POS_int~Input, value.var="Median_log_CAF_norm")

cbs_02dec2021_both_postprocess_subtract_rf_mut_cast_dt[,log_diff:=cDNA-gDNA]

mut_type_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_mut_cast_dt, 
  aes(y=log_diff, x=VAR_TYPE, col=VAR_TYPE))

mut_type_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  facet_wrap(~Chemistry)+
  geom_boxplot()+
  scale_color_startrek()+
  theme_cowplot()+
  labs(y="log ratio, cDNA to gDNA", col="Mutation")

# Add some test results
mut_type_silent_vs_missense_wilcox_res<- 
  cbs_02dec2021_both_postprocess_subtract_rf_mut_cast_dt[
  VAR_TYPE%in%c("Silent", "Missense"), wilcox.test(
    log_diff~VAR_TYPE, .SD, alternative="greater")$p.value, by=.(Chemistry)]

mut_type_silent_vs_nonsense_wilcox_res<- 
  cbs_02dec2021_both_postprocess_subtract_rf_mut_cast_dt[
  VAR_TYPE%in%c("Silent", "Nonsense"), wilcox.test(
    log_diff~VAR_TYPE, .SD, alternative="greater")$p.value, by=.(Chemistry)]

# Using RACE-like data for the whole coding sequence, determine if there is a distance
# dependent effect for nonsense variants
cbs_02dec2021_racelike_postprocess_subtract_rf_mut_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    Chemistry=="RACE-like" & !VAR_ID%in%racelike_singletons,
    .(Median_log_CAF_norm=median(log_CAF_norm, na.rm=TRUE)),
    by=.(Input, VAR_ID, TYPE, VAR_TYPE, AA_POS)]

cbs_02dec2021_racelike_postprocess_subtract_rf_mut_dt[
  ,AA_POS_int:=as.integer(AA_POS)]

cbs_02dec2021_racelike_postprocess_subtract_rf_mut_cast_dt<- dcast(
  data=cbs_02dec2021_racelike_postprocess_subtract_rf_mut_dt, 
  VAR_ID+TYPE+VAR_TYPE+AA_POS_int~Input, value.var="Median_log_CAF_norm")

cbs_02dec2021_racelike_postprocess_subtract_rf_mut_cast_dt[,log_diff:=cDNA-gDNA]

# Plot silent variants for control
mut_type_racelike_silent_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_rf_mut_cast_dt[VAR_TYPE%in%c("Silent", "Nonsense")],
  aes(x=AA_POS_int, y=log_diff, col=VAR_TYPE))

mut_type_racelike_silent_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point()+
  geom_smooth()+
  scale_x_continuous(breaks=seq(0,550,100))+
  scale_y_continuous(breaks=seq(-2.5,2.5,1))+
  coord_cartesian(xlim=c(0,550), ylim=c(-2.5,2.5))+
  scale_color_jama()+
  theme_cowplot()+
  labs(x="CBS amino acid position", 
       y="log ratio, cDNA to gDNA")

# Missense variants
mut_type_racelike_missense_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_rf_mut_cast_dt[VAR_TYPE=="Missense"],
  aes(x=AA_POS_int, y=log_diff))

mut_type_racelike_missense_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  scale_x_continuous(breaks=seq(0,550,100))+
  scale_y_continuous(breaks=seq(-4,4,1))+
  coord_cartesian(xlim=c(0,550), ylim=c(-3,4))+
  theme_cowplot()+
  labs(x="CBS amino acid position", 
       y="log ratio, cDNA to gDNA")
  
```

```{r Amplicon analysis}

amplicon_dt<- cbs_02dec2021_both_postprocess_subtract_rf_dt[Chemistry=="Amplicon"]

cbs_02dec2021_amplicon_postprocess_filt_cast_dt<- dcast(
  data=amplicon_dt, formula=VAR_ID~Sample, value.var="log10_CAF")

cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_71_R, y=CBS1_72_R))

cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg<- 
  cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="gDNA rep 1",y="gDNA rep 2")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_71_R)|is.na(CBS1_72_R)), cor(CBS1_71_R, CBS1_72_R)]


cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_71_R, y=CBS1_73_R))

cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg<-
  cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="gDNA rep 1", y="gDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_71_R)|is.na(CBS1_73_R)), cor(CBS1_71_R, CBS1_73_R)]


cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_72_R, y=CBS1_73_R))

cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg<- 
  cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="gDNA rep 2", y="gDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_72_R)|is.na(CBS1_73_R)), cor(CBS1_72_R, CBS1_73_R)]


# gDNA-plasmid library
cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_74_R, y=CBS1_71_R))

cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg<- 
  cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="Plasmid library", y="gDNA rep 1")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_74_R)|is.na(CBS1_71_R)), cor(CBS1_74_R, CBS1_71_R)]


# cDNA
cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_68_R, y=CBS1_69_R))

cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg<- 
  cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="cDNA rep 1", y="cDNA rep 2")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_68_R)|is.na(CBS1_69_R)), cor(CBS1_68_R, CBS1_69_R)]


cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_68_R, y=CBS1_70_R))

cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg<- 
  cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="cDNA rep 1", y="cDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_68_R)|is.na(CBS1_70_R)), cor(CBS1_68_R, CBS1_70_R)]


cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_69_R, y=CBS1_70_R))

cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg<- 
  cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="cDNA rep 2", y="cDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_69_R)|is.na(CBS1_70_R)), cor(CBS1_69_R, CBS1_70_R)]

# Impute missing values and construct a heatmap
cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat<- as.matrix(
  cbs_02dec2021_amplicon_postprocess_filt_cast_dt)

# Use KNN imputation to fill in CAF values for incomplete variants
# (i.e. those variants not found across all fractions)
# drop singleton variants in the process (variants found in only one fraction)

# Only do imputation on variants that are found in the plasmid library?  
# Only impute if at least two samples (even if 1 in sample and 1 in plasmid library)
row_is_observed<- function(x){!is.na(x[length(x)]) & sum(!is.na(x))>=2}
obs_rows<- apply(cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat[,-1], 1, row_is_observed)

cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_obs<- 
  cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat[obs_rows,]

# Impute 
cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res<- knn.impute(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_obs[,-1],
  k=10, cat.var=NULL)

cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res_dt<- cbind(
  cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_obs[,"VAR_ID"],
  as.data.table(cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res))

# Plot the impute frequencies in a heatmap
input_colors<- brewer.pal(n=3, "Set2")

heatmap_annotations<- data.frame(
    Input=as.factor(rep(c("cDNA", "gDNA", "Plasmid"), times=c(3,3,1))))

heatmap_colors<- list(
  Input=c("cDNA"=input_colors[1], "gDNA"=input_colors[2], "Plasmid"=input_colors[3]))

pheatmap(
  mat=cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res,
  clustering_method="ward.D2",
  clustering_distance_rows="euclidean",
  clustering_distance_cols="euclidean")

```

```{r Amplicon analysis normalized}

amplicon_dt<- cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[Chemistry=="Amplicon"]

cbs_02dec2021_amplicon_postprocess_filt_cast_dt<- dcast(
  data=amplicon_dt, formula=VAR_ID~Sample, value.var="log_CAF_norm")

cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_71_R, y=CBS1_72_R))

cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg<- 
  cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  labs(x="gDNA rep 1",y="gDNA rep 2")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_71_R)|is.na(CBS1_72_R)), cor(CBS1_71_R, CBS1_72_R)]


cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_71_R, y=CBS1_74_R))

cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg<-
  cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  labs(x="gDNA rep 1", y="gDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_71_R)|is.na(CBS1_74_R)), cor(CBS1_71_R, CBS1_74_R)]


cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_72_R, y=CBS1_74_R))

cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg<- 
  cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  labs(x="gDNA rep 2", y="gDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_72_R)|is.na(CBS1_74_R)), cor(CBS1_72_R, CBS1_74_R)]


# gDNA-plasmid library
cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_74_R, y=CBS1_71_R))

cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg<- 
  cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  labs(x="Plasmid library", y="gDNA rep 1")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_74_R)|is.na(CBS1_71_R)), cor(CBS1_74_R, CBS1_71_R)]


# cDNA
cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_68_R, y=CBS1_69_R))

cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg<- 
  cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  labs(x="cDNA rep 1", y="cDNA rep 2")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_68_R)|is.na(CBS1_69_R)), cor(CBS1_68_R, CBS1_69_R)]


cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_68_R, y=CBS1_70_R))

cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg<- 
  cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  labs(x="cDNA rep 1", y="cDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_68_R)|is.na(CBS1_70_R)), cor(CBS1_68_R, CBS1_70_R)]


cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_dt, 
  aes(x=CBS1_69_R, y=CBS1_70_R))

cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg<- 
  cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  labs(x="cDNA rep 2", y="cDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg)

cbs_02dec2021_amplicon_postprocess_filt_cast_dt[
  !(is.na(CBS1_69_R)|is.na(CBS1_70_R)), cor(CBS1_69_R, CBS1_70_R)]

# Impute missing values and construct a heatmap
cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat<- as.matrix(
  cbs_02dec2021_amplicon_postprocess_filt_cast_dt)

# Use KNN imputation to fill in CAF values for incomplete variants
# (i.e. those variants not found across all fractions)
# drop singleton variants in the process (variants found in only one fraction)

# Only do imputation on variants that are found in the plasmid library?  
# Only impute if at least two samples (even if 1 in sample and 1 in plasmid library)
row_is_observed<- function(x){!is.na(x[length(x)]) & sum(!is.na(x))>=2}
obs_rows<- apply(cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat[,-1], 1, row_is_observed)

cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_obs<- 
  cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat[obs_rows,]

# Impute 
cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res<- knn.impute(
  data=cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_obs[,-1],
  k=10, cat.var=NULL)

cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res_dt<- cbind(
  cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_obs[,"VAR_ID"],
  as.data.table(cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res))

input_colors<- brewer.pal(n=3, "Set2")

heatmap_annotations<- data.frame(
    Input=as.factor(rep(c("cDNA", "gDNA", "Plasmid"), times=c(3,3,1))))

heatmap_colors<- list(
  Input=c("cDNA"=input_colors[1], "gDNA"=input_colors[2], "Plasmid"=input_colors[3]))

pheatmap(
  mat=cbs_02dec2021_amplicon_postprocess_filt_cast_impute_mat_res,
  clustering_method="ward.D2",
  clustering_distance_rows="euclidean",
  clustering_distance_cols="euclidean")

```

gDNA and plasmid library cluster together, and all cDNA samples cluster together. It appears gDNA rep 3 is an outlier. Remove it and use the plasmid library for gDNA rep 3.

```{r RACE-like analysis}

racelike_dt<- cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[Chemistry=="RACE-like"]

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt<- dcast(
  data=racelike_dt, 
  formula=VAR_ID~Sample, value.var="log10_CAF")

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt[
  !(is.na(CBS1_76_R)|is.na(CBS1_77_R)), cor(CBS1_76_R, CBS1_77_R)]

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt[
  !(is.na(CBS1_76_R)|is.na(CBS1_78_R)), cor(CBS1_76_R, CBS1_78_R)]

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt[
  !(is.na(CBS1_77_R)|is.na(CBS1_78_R)), cor(CBS1_77_R, CBS1_78_R)]

# gDNA rep 1 vs rep 2
cbs_02dec2021_racelike_gDNA_rep1_vs_rep2_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt, 
  aes(x=CBS1_76_R, y=CBS1_77_R))

cbs_02dec2021_racelike_gDNA_rep1_vs_rep2_gg<- 
  cbs_02dec2021_racelike_gDNA_rep1_vs_rep2_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="gDNA rep 1", y="gDNA rep 2")+
  theme_cowplot()

print(cbs_02dec2021_racelike_gDNA_rep1_vs_rep2_gg)

# gDNA rep 1 vs rep 3
cbs_02dec2021_racelike_gDNA_rep1_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt, 
  aes(x=CBS1_76_R, y=CBS1_78_R))

cbs_02dec2021_racelike_gDNA_rep1_vs_rep3_gg<- 
  cbs_02dec2021_racelike_gDNA_rep1_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="gDNA rep 1", y="gDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_racelike_gDNA_rep1_vs_rep3_gg)

# gDNA rep 2 vs rep 3
cbs_02dec2021_racelike_gDNA_rep2_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt, 
  aes(x=CBS1_77_R, y=CBS1_78_R))

cbs_02dec2021_racelike_gDNA_rep2_vs_rep3_gg<- 
  cbs_02dec2021_racelike_gDNA_rep2_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="gDNA rep 2", y="gDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_racelike_gDNA_rep2_vs_rep3_gg)

# cDNA
cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt[
  !(is.na(CBS1_79_R)|is.na(CBS1_80_R)), cor(CBS1_79_R, CBS1_80_R)]

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt[
  !(is.na(CBS1_79_R)|is.na(CBS1_81_R)), cor(CBS1_79_R, CBS1_81_R)]

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt[
  !(is.na(CBS1_80_R)|is.na(CBS1_81_R)), cor(CBS1_80_R, CBS1_81_R)]

# cDNA rep 1 vs rep 2
cbs_02dec2021_racelike_cDNA_rep1_vs_rep2_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt, 
  aes(x=CBS1_79_R, y=CBS1_80_R))

cbs_02dec2021_racelike_cDNA_rep1_vs_rep2_gg<- 
  cbs_02dec2021_racelike_cDNA_rep1_vs_rep2_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="cDNA rep 1", y="cDNA rep 2")+
  theme_cowplot()

print(cbs_02dec2021_racelike_cDNA_rep1_vs_rep2_gg)

# cDNA rep 1 vs rep 3
cbs_02dec2021_racelike_cDNA_rep1_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt, 
  aes(x=CBS1_79_R, y=CBS1_81_R))

cbs_02dec2021_racelike_cDNA_rep1_vs_rep3_gg<- 
  cbs_02dec2021_racelike_cDNA_rep1_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="cDNA rep 1", y="cDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_racelike_cDNA_rep1_vs_rep3_gg)

# cDNA rep 2 vs rep 3
cbs_02dec2021_racelike_cDNA_rep2_vs_rep3_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt, 
  aes(x=CBS1_80_R, y=CBS1_81_R))

cbs_02dec2021_racelike_cDNA_rep2_vs_rep3_gg<- 
  cbs_02dec2021_racelike_cDNA_rep2_vs_rep3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="lm")+
  coord_cartesian(xlim=c(-7.2,-2.8), ylim=c(-7.2,-2.8))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  labs(x="cDNA rep 2", y="cDNA rep 3")+
  theme_cowplot()

print(cbs_02dec2021_racelike_cDNA_rep2_vs_rep3_gg)

# Impute missing values and construct a heatmap
cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat<- as.matrix(
  cbs_02dec2021_racelike_postprocess_subtract_filt_cast_dt)

# Use KNN imputation to fill in CAF values for incomplete variants
# (i.e. those variants not found across all fractions)
# drop singleton variants in the process (variants found in only one fraction)

row_is_observed<- function(x){sum(!is.na(x))>=2}
obs_rows<- apply(cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat[,-1], 1, row_is_observed)

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_obs<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat[obs_rows,]

# Impute 
cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat_res<- knn.impute(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_obs[,-1],
  k=10, cat.var=NULL)

cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat_res_dt<- cbind(
  cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_obs[,"VAR_ID"],
  as.data.table(cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat_res))

names(cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat_res_dt)[1]<- "VAR_ID"

# Plot the impute frequencies in a heatmap
input_colors<- brewer.pal(n=3, "Set2")

heatmap_annotations<- data.frame(
    Input=as.factor(rep(c("cDNA", "gDNA", "Plasmid"), times=c(3,3,1))))

heatmap_colors<- list(
  Input=c("cDNA"=input_colors[1], "gDNA"=input_colors[2], "Plasmid"=input_colors[3]))

pheatmap(
  mat=cbs_02dec2021_racelike_postprocess_subtract_filt_cast_impute_mat_res,
  clustering_method="ward.D2",
  clustering_distance_rows="euclidean",
  clustering_distance_cols="euclidean")

# Examine error rate at the amino acid positions 278-280 compared to global error rates
#cbs_02dec2021_racelike_postprocess_dt[,Homopolymer:=NULL]
cbs_02dec2021_racelike_postprocess_dt[AA_POS%in%as.character(seq(278,280)), Homopolymer:="True"]
cbs_02dec2021_racelike_postprocess_dt[is.na(Homopolymer), Homopolymer:="False"]
cbs_02dec2021_racelike_postprocess_dt[,Homopolymer:=as.factor(Homopolymer)]

error_rate_278_280_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_dt[Input=="Negative_control"],
  aes(x=Homopolymer, y=log10_CAF, col=Homopolymer))

error_rate_278_280_gg+
  geom_boxplot(position="dodge")+
  scale_color_jco()+
  labs(y="log10 frequency")+
  theme_cowplot()

error_rate_278_280_wilcox_res<- wilcox.test(
  cbs_02dec2021_racelike_postprocess_dt[Input=="Negative_control" & Homopolymer=="True", log10_CAF],
  cbs_02dec2021_racelike_postprocess_dt[Input=="Negative_control" & Homopolymer=="False", log10_CAF])

```

```{r Global comparison between gDNA and cDNA frequencies}

# Amplicon, no singletons
cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
  Chemistry=="Amplicon" & !VAR_ID%in%amplicon_singletons,
  .(Median_log10_CAF=median(log10_CAF, na.rm=TRUE)), by=.(Input, VAR_ID)], 
  VAR_ID~Input, value.var="Median_log10_CAF")

cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_dt, 
  aes(x=gDNA, y=cDNA))

cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_gg<- 
  cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  coord_cartesian(xlim=c(-7.1,-2.9), ylim=c(-7.1,-2.9))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  geom_smooth(method="lm")+
  labs(x="gDNA", y="cDNA")+
  theme_cowplot()

print(cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_gg)

cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_dt[
  !(is.na(gDNA)|is.na(cDNA)),cor(gDNA, cDNA)]


# RACE-like, no singletons
cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    Chemistry=="RACE-like" & !VAR_ID%in%racelike_singletons,
  .(Median_log10_CAF=median(log10_CAF, na.rm=TRUE)), by=.(Input, VAR_ID)], 
  VAR_ID~Input, value.var="Median_log10_CAF")

cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_dt, 
  aes(x=gDNA, y=cDNA))

cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_gg<- 
  cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  scale_x_continuous(breaks=seq(-6,-3,1))+
  scale_y_continuous(breaks=seq(-6,-3,1))+
  coord_cartesian(xlim=c(-6,-2.7), ylim=c(-6,-2.7))+
  geom_smooth(method="lm")+
  labs(x="gDNA", y="cDNA")+
  theme_cowplot()

print(cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_gg)

cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_dt[
  !(is.na(gDNA)|is.na(cDNA)),cor(gDNA, cDNA)]


# Amplicon, complete obs.
cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_cor_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    Chemistry=="Amplicon" & VAR_ID%in%amplicon_complete,
  .(Median_log10_CAF=median(log10_CAF, na.rm=TRUE)), by=.(Input, VAR_ID)], 
  VAR_ID~Input, value.var="Median_log10_CAF")

cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_cor_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_cor_dt, 
  aes(x=gDNA, y=cDNA))

cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_cor_gg<- 
  cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  coord_cartesian(xlim=c(-7.1,-2.9), ylim=c(-7.1,-2.9))+
  scale_x_continuous(breaks=seq(-7,-3))+
  scale_y_continuous(breaks=seq(-7,-3))+
  geom_smooth(method="lm")+
  labs(x="gDNA", y="cDNA")+
  theme_cowplot()

print(cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_cor_gg)

cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_cor_dt[
  !(is.na(gDNA)|is.na(cDNA)),cor(gDNA, cDNA)]


# RACE-like, complete obs.
cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_cor_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    Chemistry=="RACE-like" & VAR_ID%in%racelike_complete,
  .(Median_log10_CAF=median(log10_CAF, na.rm=TRUE)), by=.(Input, VAR_ID)], 
  VAR_ID~Input, value.var="Median_log10_CAF")

cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_cor_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_cor_dt, 
  aes(x=gDNA, y=cDNA))

cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  scale_x_continuous(breaks=seq(-6,-3,1))+
  scale_y_continuous(breaks=seq(-6,-3,1))+
  coord_cartesian(xlim=c(-6,-2.7), ylim=c(-6,-2.7))+
  geom_smooth(method="lm")+
  labs(x="gDNA", y="cDNA")+
  theme_cowplot()

cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_cor_dt[
  !(is.na(gDNA)|is.na(cDNA)),cor(gDNA, cDNA)]

# Arrange the relevant comparisons into a grid for amplicon chemistry
plot_grid(
  cbs_02dec2021_amplicon_gDNA_rep1_vs_plasmid_gg, 
  cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg, 
  cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg, 
  cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_cor_gg)


```

As expected, there was a strong positive correlation between variant abundance in gDNA and abundance in cDNA. Filtering out variants without observations in both inputs in all replicates slightly improved correlation.

Determine overlap of variants called by both chemistries in tile 2 and 4.

```{r Overlap in calls}

# I filtered the total codon permutations to those that are able to be captured by 
# both amplicon and RACE-like libraries
# grep -v ^# pDEST_HC_Rec_Bxb_v2_CBS_WT.codon.permuts.tile2_4.sort.vcf | wc -l

cbs_tile2_4_all_nnk_vars<- 2676

chem_compare_amplicon_varids<- cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
  Chemistry=="Amplicon" & POS%in%amplicon_tile2_4_target & 
    !VAR_ID%in%amplicon_singletons, unique(VAR_ID)]

chem_compare_racelike_varids<- cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
  Chemistry=="RACE-like" & POS%in%amplicon_tile2_4_target & 
    !VAR_ID%in%racelike_singletons, unique(VAR_ID)]

chem_compare_venn_diagram_file<- paste(
  cbs_02dec2021_dir, "amplicon_vs_racelike_no_singletons_venn_reseq.svg", sep="/")

chem_compare_venn_diagram<- venn.diagram(list(
  chem_compare_amplicon_varids, chem_compare_racelike_varids), imagetype="svg",
  filename=chem_compare_venn_diagram_file, category.names=c("Amplicon" , "RACE-like"),
  output=TRUE, fill=brewer.pal(n=3,"Pastel2")[c(2,3)], height=5, width=5,
  hyper.test=TRUE, total.population=cbs_tile2_4_all_nnk_vars, lower.tail=FALSE)

chem_compare_intersect_varids<- chem_compare_amplicon_varids[
  chem_compare_amplicon_varids%in%chem_compare_racelike_varids]

phyper(q=length(chem_compare_intersect_varids), 
       m=length(chem_compare_amplicon_varids),
       n=cbs_tile2_4_all_nnk_vars - length(chem_compare_amplicon_varids), 
       k=length(chem_compare_racelike_varids), lower.tail=FALSE)

```

```{r Chemistry comparison}

chem_compare_thresh<- (-5.2)

cbs_02dec2021_both_postprocess_subtract_rf_comp_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    (Chemistry=="Amplicon" & !VAR_ID%in%amplicon_singletons) | 
      (Chemistry=="RACE-like" & !VAR_ID%in%racelike_singletons),
  .(Median_log10_CAF=median(log10_CAF, na.rm=TRUE)),
  by=.(Chemistry, Input, VAR_ID, VAR_TYPE)]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_dt, 
  VAR_ID~Chemistry+Input, value.var="Median_log10_CAF")

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_dt[
    Amplicon_gDNA>=chem_compare_thresh]

# gDNA
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt, 
  aes(x=Amplicon_gDNA, y=`RACE-like_gDNA`))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg<- 
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.7)+
  scale_x_continuous(breaks=seq(-6,-3,1))+
  scale_y_continuous(breaks=seq(-6,-3,1))+
  coord_cartesian(xlim=c(-6,-2.7), ylim=c(-6,-2.7))+
  geom_smooth(method="lm")+
  labs(x="Amplicon", y="RACE-like")+
  theme_cowplot()

print(cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg)

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt[
  !(is.na(Amplicon_gDNA)|is.na(`RACE-like_gDNA`)), 
  cor(Amplicon_gDNA, `RACE-like_gDNA`, method="pearson")]

# cDNA
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt, 
  aes(x=Amplicon_cDNA, y=`RACE-like_cDNA`))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg<- 
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.7)+
  scale_x_continuous(breaks=seq(-6,-3,1))+
  scale_y_continuous(breaks=seq(-6,-3,1))+
  coord_cartesian(xlim=c(-6,-2.7), ylim=c(-6,-2.7))+
  geom_smooth(method="lm")+
  labs(x="Amplicon",y="RACE-like")+
  theme_cowplot()

print(cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg)

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt[
  !(is.na(Amplicon_cDNA)|is.na(`RACE-like_cDNA`)), 
  cor(Amplicon_cDNA, `RACE-like_cDNA`, method="pearson")]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt[
  ,Amplicon_RACElike_cDNA_diff:=Amplicon_cDNA-`RACE-like_cDNA`]


# Plot as histogram
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_hist_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt, 
  aes(x=Amplicon_RACElike_cDNA_diff))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_hist_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_histogram(alpha=0.75, fill="steelblue")+
  labs(x="log10 frequency diff, amplicon - RACE-like")+
  theme_cowplot()


# Look at cDNA to gDNA diff between chemistries
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_dt, 
  VAR_ID+Chemistry~Input, value.var="Median_log10_CAF")

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_dt[,log10_diff:=cDNA-gDNA]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_dt<- 
  melt(data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_dt,
       id.vars=c("VAR_ID", "Chemistry"), value.name="log10_diff", variable.name="Input")

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_dt[
    Input=="log10_diff"], VAR_ID~Chemistry, value.var="log10_diff")
  
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_dt[
  !(is.na(Amplicon)|is.na(`RACE-like`)), cor(Amplicon, `RACE-like`, method="pearson")]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_dt, 
  aes(x=Amplicon, y=`RACE-like`))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.7)+
  geom_smooth(method="lm")+
  labs(x="log10 diff, amplicon", 
       y="log10 diff, RACE-like")+
  theme_cowplot()

```

```{r Chemistry comparison normalized}

chem_compare_thresh<- (-11.97344)

cbs_02dec2021_both_postprocess_subtract_rf_comp_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    (Chemistry=="Amplicon" & !VAR_ID%in%amplicon_singletons) | 
      (Chemistry=="RACE-like" & !VAR_ID%in%racelike_singletons),
  .(Median_log_CAF_norm=median(log_CAF_norm, na.rm=TRUE)),
  by=.(Chemistry, Input, VAR_ID, VAR_TYPE)]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_dt, 
  VAR_ID~Chemistry+Input, value.var="Median_log_CAF_norm")

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_dt[
    Amplicon_gDNA>=chem_compare_thresh]

# gDNA
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt, 
  aes(x=Amplicon_gDNA, y=`RACE-like_gDNA`))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg<- 
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.7)+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  geom_smooth(method="lm")+
  labs(x="Amplicon", y="RACE-like")+
  theme_cowplot()

print(cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg)

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt[
  !(is.na(Amplicon_gDNA)|is.na(`RACE-like_gDNA`)), 
  cor(Amplicon_gDNA, `RACE-like_gDNA`, method="pearson")]

# cDNA
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt, 
  aes(x=Amplicon_cDNA, y=`RACE-like_cDNA`))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg<- 
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.7)+
  coord_cartesian(xlim=c(-16,-6), ylim=c(-16,-6))+
  scale_x_continuous(breaks=seq(-16,-6,2))+
  scale_y_continuous(breaks=seq(-16,-6,2))+
  geom_smooth(method="lm")+
  labs(x="Amplicon",y="RACE-like")+
  theme_cowplot()

print(cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg)

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt[
  !(is.na(Amplicon_cDNA)|is.na(`RACE-like_cDNA`)), 
  cor(Amplicon_cDNA, `RACE-like_cDNA`, method="pearson")]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt[
  ,Amplicon_RACElike_cDNA_diff:=Amplicon_cDNA-`RACE-like_cDNA`]


# Plot as histogram
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_hist_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_filt_dt, 
  aes(x=Amplicon_RACElike_cDNA_diff))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_hist_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_histogram(alpha=0.75, fill="steelblue")+
  labs(x="log ratio diff, amplicon - RACE-like")+
  theme_cowplot()


# Look at cDNA to gDNA diff between chemistries
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_dt, 
  VAR_ID+Chemistry~Input, value.var="Median_log_CAF_norm")

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_dt[,log_diff:=cDNA-gDNA]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_dt<- 
  melt(data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_dt,
       id.vars=c("VAR_ID", "Chemistry"), value.name="log_diff", variable.name="Input")

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_dt<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_dt[
    Input=="log_diff"], VAR_ID~Chemistry, value.var="log_diff")
  
cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_dt[
  !(is.na(Amplicon)|is.na(`RACE-like`)), cor(Amplicon, `RACE-like`, method="pearson")]

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_dt, 
  aes(x=Amplicon, y=`RACE-like`))

cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_input_melt_reshape_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.7)+
  geom_smooth(method="lm")+
  labs(x="log ratio, amplicon", 
       y="log ratio, RACE-like")+
  theme_cowplot()

```

Show all correlations between replicates for amplicon and RACE-like chemistries. 

```{r Supplementary Figure 4}

plot_grid(
  cbs_02dec2021_amplicon_gDNA_rep1_vs_rep2_gg,
  cbs_02dec2021_amplicon_gDNA_rep1_vs_rep3_gg,
  cbs_02dec2021_amplicon_gDNA_rep2_vs_rep3_gg,
  cbs_02dec2021_amplicon_cDNA_rep1_vs_rep2_gg,
  cbs_02dec2021_amplicon_cDNA_rep1_vs_rep3_gg,
  cbs_02dec2021_amplicon_cDNA_rep2_vs_rep3_gg,
  cbs_02dec2021_racelike_gDNA_rep1_vs_rep2_gg,
  cbs_02dec2021_racelike_gDNA_rep1_vs_rep3_gg,
  cbs_02dec2021_racelike_gDNA_rep2_vs_rep3_gg,
  cbs_02dec2021_racelike_cDNA_rep1_vs_rep2_gg,
  cbs_02dec2021_racelike_cDNA_rep1_vs_rep3_gg,
  cbs_02dec2021_racelike_cDNA_rep2_vs_rep3_gg,
  ncol=3)

plot_grid(
  cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_cor_gg,
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_gdna_gg,
  cbs_02dec2021_both_postprocess_subtract_rf_comp_cast_cdna_gg, ncol=1)


```

```{r Mean variance relationship by chemistry and input type}

# Remove variants with only one replicate supporting the variant in each group
cbs_02dec2021_both_postprocess_subtract_counts_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[
    ,.N, by=.(Chemistry, VAR_ID)]

amplicon_singletons<- cbs_02dec2021_both_postprocess_subtract_counts_dt[
  Chemistry=="Amplicon" & N==1, unique(VAR_ID)]

racelike_singletons<- cbs_02dec2021_both_postprocess_subtract_counts_dt[
  Chemistry=="RACE-like" & N==1, unique(VAR_ID)]

amplicon_complete<- cbs_02dec2021_both_postprocess_subtract_counts_dt[
  Chemistry=="Amplicon" & N==6, unique(VAR_ID)]

racelike_complete<- cbs_02dec2021_both_postprocess_subtract_counts_dt[
  Chemistry=="RACE-like" & N==6, unique(VAR_ID)]

# Amplicon, no singletons
cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_dt[
  Input!="Input_library" & Chemistry=="Amplicon" & !VAR_ID%in%amplicon_singletons,
  .(Mean_log10_CAF=mean(log10_CAF, na.rm=TRUE), 
    Var_log10_CAF=var(log10_CAF)),
  by=.(Input, VAR_ID)]

cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_dt, 
  aes(x=Mean_log10_CAF, y=Var_log10_CAF))

cbs_02dec2021_both_postprocess_subtract_summ_nosing_amplicon_gg+
  facet_wrap(~Input)+
  geom_hline(yintercept=c(0.5), col="grey", linetype=2)+
  geom_point()+
  geom_smooth(se=FALSE)+
  labs(x="log10 frequency mean", 
       y="log10 frequency variance",
       title="Amplicon, no singletons")+
  theme_cowplot()

# RACE-like, no singletons
cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_dt[
  Input!="Input_library" & Chemistry=="RACE-like" & !VAR_ID%in%racelike_singletons,
  .(Mean_log10_CAF=mean(log10_CAF, na.rm=TRUE), 
    Var_log10_CAF=var(log10_CAF)),
  by=.(Input, VAR_ID)]

cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_dt, 
  aes(x=Mean_log10_CAF, y=Var_log10_CAF))

cbs_02dec2021_both_postprocess_subtract_summ_nosing_racelike_gg+
  facet_wrap(~Input)+
  geom_hline(yintercept=c(0.5), col="grey", linetype=2)+
  geom_point()+
  geom_smooth(se=FALSE)+
  labs(x="log10 frequency mean", 
       y="log10 frequency variance",
       title="RACE-like, no singletons")+
  theme_cowplot()

# Amplicon, complete obs
cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_dt[
  Input!="Input_library" & Chemistry=="Amplicon" & VAR_ID%in%amplicon_complete,
  .(Mean_log10_CAF=mean(log10_CAF, na.rm=TRUE), 
    Var_log10_CAF=var(log10_CAF)),
  by=.(Input, VAR_ID)]

cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_dt, 
  aes(x=Mean_log10_CAF, y=Var_log10_CAF))

cbs_02dec2021_both_postprocess_subtract_summ_comp_amplicon_gg+
  facet_wrap(~Input)+
  geom_hline(yintercept=c(0.5), col="grey", linetype=2)+
  geom_point()+
  geom_smooth(se=FALSE)+
  labs(x="log10 frequency mean", 
       y="log10 frequency variance",
       title="Amplicon, complete obs.")+
  theme_cowplot()

# RACE-like, complete obs
cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_dt[
  Input!="Input_library" & Chemistry=="RACE-like" & VAR_ID%in%racelike_complete,
  .(Mean_log10_CAF=mean(log10_CAF, na.rm=TRUE), 
    Var_log10_CAF=var(log10_CAF)),
  by=.(Input, VAR_ID)]

cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_dt, 
  aes(x=Mean_log10_CAF, y=Var_log10_CAF))

cbs_02dec2021_both_postprocess_subtract_summ_comp_racelike_gg+
  facet_wrap(~Input)+
  geom_hline(yintercept=c(0.5), col="grey", linetype=2)+
  geom_point()+
  geom_smooth(se=FALSE)+
  labs(x="log10 frequency mean", 
       y="log10 frequency variance",
       title="RACE-like, complete obs.")+
  theme_cowplot()

```

Mean variance relationships differed for the chemistries most likely due to differences in sequencing depth (RACE-like AMP libraries do not have the same level of coverage that amplicon libraries have).

One amplicon gDNA library appeared to be an outlier from clustering analysis. This is also observed in higher variance in gDNA compared to cDNA for amplicon libraries. Remove this gDNA library and substitute the plasmid library as it is very similar to the other two gDNA replicates.

## Differential abundance, amplicon method

```{r Plot density of variants by frequency}

freq_distr_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt, 
  aes(x=log_CAF_norm, col=Input))

freq_distr_gg+
  facet_wrap(~Chemistry)+
  geom_density()+
  theme_cowplot()+
  labs(x="Normalized log frequency")

```

```{r limma- amplicon chemistry}

amplicon_fdr_thresh<- 0.1
amplicon_log_caf_thresh<- -15

cbs_02dec2021_amplicon_postprocess_subtract_filt_counts_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="Amplicon" & log_CAF >= amplicon_log_caf_thresh, .N, by=.(VAR_ID)]

limma_amplicon_complete<- cbs_02dec2021_amplicon_postprocess_subtract_filt_counts_dt[
  N==6, unique(VAR_ID)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="Amplicon" & VAR_ID%in%limma_amplicon_complete]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_dt<- dcast(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt, 
  formula=VAR_ID~Sample, value.var="log_CAF")

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_mat<- as.matrix(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_dt[,-c("VAR_ID")])

rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_mat)<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_dt[,VAR_ID]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset<- 
  Biobase::ExpressionSet(assayData=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_mat)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
    ,as.factor(unique(Input)), by=.(Sample)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_mat<- 
  as.matrix(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_dt[,-c("Sample")])

rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_mat)<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[,unique(Sample)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_design<- 
  model.matrix(~V1, data=as.data.frame(
    cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_mat), 
    list(V1=contr.treatment(n=2, base=2)))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit<- 
  lmFit(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset,
        cbs_02dec2021_amplicon_postprocess_subtract_filt_design)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb<-
  eBayes(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top<- topTable(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb, coef=ncol(cbs_02dec2021_amplicon_postprocess_subtract_filt_design), number=50)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_dt<- as.data.table( topTable(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb, coef=ncol(cbs_02dec2021_amplicon_postprocess_subtract_filt_design), number=nrow(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb), sort.by="none"))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig<- subset(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top, 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top$adj.P.Val < amplicon_fdr_thresh)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids<- rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids<- rownames(subset(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig, 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig$logFC>0))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids<- rownames(subset(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig, 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig$logFC<0))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
  VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids,
  Direction:="RNA_up"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
  VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids,
  Direction:="RNA_down"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[,Direction:=as.factor(Direction)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
  ,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

amplicon_gDNA_med_log_caf<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
  Input=="gDNA", median(log_CAF)]

amplicon_cDNA_med_log_caf<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
  Input=="cDNA", median(log_CAF)]

aaas_cols<- pal_aaas()(2)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
    VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_dt[,AA_POS_int:=as.integer(AA_POS)]
  
# Visualize the frequencies
cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_eset_fit_eb_top_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_dt,
  aes(x=log_CAF, y=reorder(ID_SHORT, AA_POS_int), col=Input))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_eset_fit_eb_top_gg+
  geom_vline(xintercept=amplicon_cDNA_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_color_aaas()+
  theme_cowplot()

```

```{r limma- amplicon chemistry with WT normalization}

amplicon_fdr_thresh<- 0.1
amplicon_log_caf_thresh<- -15

cbs_02dec2021_amplicon_postprocess_subtract_filt_counts_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="Amplicon" & log_CAF_norm >= amplicon_log_caf_thresh, .N, by=.(VAR_ID)]

limma_amplicon_complete<- cbs_02dec2021_amplicon_postprocess_subtract_filt_counts_dt[
  N==6, unique(VAR_ID)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="Amplicon" & VAR_ID%in%limma_amplicon_complete]

caf_norm_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt,
  aes(x=log_CAF, y=log_CAF_norm))

caf_norm_gg+
  facet_wrap(~Input)+
  geom_point()+
  theme_cowplot()+
  labs(x="log CAF", y="Normalized log CAF")

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_norm_dt<- dcast(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt, 
  formula=VAR_ID~Sample, value.var="log_CAF_norm")

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_mat_norm<- as.matrix(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_norm_dt[,-c("VAR_ID")])

rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_mat_norm)<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_norm_dt[,VAR_ID]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_norm<- 
  Biobase::ExpressionSet(assayData=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_mat_norm)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_norm_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
    ,as.factor(unique(Input)), by=.(Sample)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_mat_norm<- 
  as.matrix(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_norm_dt[,-c("Sample")])

rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_mat_norm)<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[,unique(Sample)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_design_norm<- 
  model.matrix(~V1, data=as.data.frame(
    cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_design_mat_norm), 
    list(V1=contr.treatment(n=2, base=2)))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_norm<- 
  lmFit(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_norm,
        cbs_02dec2021_amplicon_postprocess_subtract_filt_design_norm)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm<-
  eBayes(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_norm)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt<- as.data.table( topTable(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm, coef=ncol(cbs_02dec2021_amplicon_postprocess_subtract_filt_design_norm), number=nrow(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm), sort.by="none"))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt[
  ,VAR_ID:=rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_norm<- topTable(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm, coef=ncol(cbs_02dec2021_amplicon_postprocess_subtract_filt_design_norm), number=50)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig_norm<- subset(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_norm, 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_norm$adj.P.Val < amplicon_fdr_thresh)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids_norm<- rownames(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig_norm)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids_norm<- rownames(subset(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig_norm, 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig_norm$logFC>0))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids_norm<- rownames(subset(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig_norm, 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_sig_norm$logFC<0))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
  VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids_norm,
  Direction:="RNA_up"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
  VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids_norm,
  Direction:="RNA_down"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[,Direction:=as.factor(Direction)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
  ,ID_SHORT:=paste(substr(AA_CHANGE, 3, 100), paste(REF, ALT, sep=">"), sep=", ")]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
  ,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

amplicon_gDNA_med_log_caf_norm<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
  Input=="gDNA", median(log_CAF_norm)]

amplicon_cDNA_med_log_caf_norm<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
  Input=="cDNA", median(log_CAF_norm)]

aaas_cols<- pal_aaas()(2)

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_norm_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_norm_dt[
    VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids_norm]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_norm_dt[,AA_POS_int:=as.integer(AA_POS)]
  
# Visualize the frequencies
cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_eset_fit_eb_top_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_norm_dt,
  aes(x=log_CAF_norm, y=reorder(ID_SHORT, AA_POS_int), col=Input))

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_eset_fit_eb_top_gg+
  geom_vline(xintercept=amplicon_cDNA_med_log_caf_norm, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf_norm, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  scale_x_continuous(breaks=seq(-14,-10, 2))+
  geom_point(position=position_jitterdodge(jitter.width=0.5))+
  labs(y="Variant ID", x="log frequency")+
  scale_color_aaas()+
  theme_cowplot()

```

```{r limma- amplicon chemistry normalized vs non-normalized}

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_compare_dt<- 
  merge(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_dt[
    ,.(VAR_ID, logFC, adj.P.Val, AveExpr)],
        cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt[
          ,.(VAR_ID, logFC, adj.P.Val, AveExpr)], by="VAR_ID")

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_compare_dt[
  ,AveExpr:=(AveExpr.x+AveExpr.y)/2]

amplicon_limma_compare_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_compare_dt, 
  aes(x=logFC.x, y=logFC.y, col=AveExpr))

amplicon_limma_compare_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point()+
  coord_cartesian(xlim=c(-2.5,2.5), ylim=c(-2.5,2.5))+
  scale_color_viridis(option="C")+
  theme_cowplot()+
  labs(x="Non-normalized", y="WT-normalized", 
       col="Mean log
frequency")

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_compare_dt[
  ,cor.test(x=logFC.x, y=logFC.y)]

```

## Differential abundance, RACE-like method

```{r limma- RACE-like chemistry}

racelike_fdr_thresh<- 0.1
racelike_log_caf_thresh<- -12

cbs_02dec2021_racelike_postprocess_subtract_filt_counts_dt<- 
    cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="RACE-like" & log_CAF >= racelike_log_caf_thresh, .N, by=.(VAR_ID)]

limma_racelike_complete<- cbs_02dec2021_racelike_postprocess_subtract_filt_counts_dt[
  N==6, unique(VAR_ID)]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="RACE-like"& VAR_ID%in%limma_racelike_complete]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_dt<- dcast(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt, 
  formula=VAR_ID~Sample, value.var="log_CAF")

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_mat<- as.matrix(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_dt[,-c("VAR_ID")])

rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_mat)<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_dt[,VAR_ID]

# Make a design matrix
cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_design_dt<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
    ,as.factor(unique(Input)), by=.(Sample)]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design_mat<- 
  as.matrix(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_design_dt[,-c("Sample")])

rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design_mat)<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_design_dt[,Sample]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design<- 
  model.matrix(~V1, data=as.data.frame(
    cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design_mat), 
    list(V1=contr.treatment(n=2, base=2)))

# Make ExpressionSet
cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset<- 
  Biobase::ExpressionSet(assayData=cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_mat)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit<- 
  lmFit(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset,
        cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb<-
  eBayes(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_dt<- as.data.table(
  topTable(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb, 
           coef=ncol(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design), 
           number=nrow(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb), 
           sort.by="none"))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb)]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top<- topTable(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb, coef=ncol(
      cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design), number=50)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_sig<- subset(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top$adj.P.Val < racelike_fdr_thresh)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids<- rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_sig)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids<- rownames(subset(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top$logFC>0))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids<- rownames(subset(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top$logFC<0))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  ,ID_LONG:=paste(AA_CHANGE, POS, REF, ALT, sep=":")]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  VAR_ID%in%cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids,
  Direction:="RNA_up"]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  VAR_ID%in%cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids,
  Direction:="RNA_down"]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[,Direction:=as.factor(Direction)]

racelike_gDNA_med_log10_caf<- cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  Input=="gDNA", median(log10_CAF)]

racelike_cDNA_med_log10_caf<- cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  Input=="cDNA", median(log10_CAF)]

aaas_cols<- pal_aaas()(2)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_dt<-
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
    VAR_ID%in%cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids]

# Visualize the frequencies
cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_top_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_dt,
  aes(x=log10_CAF, y=reorder(ID_SHORT, AA_POS2), col=Input))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_top_gg+
  geom_vline(xintercept=racelike_cDNA_med_log10_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=racelike_gDNA_med_log10_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  scale_x_continuous(breaks=seq(-6,-3,1))+
  coord_cartesian(xlim=c(-6,-3))+
  labs(y="Variant ID", x="log10 frequency")+
  scale_color_aaas()+
  theme_cowplot()

# c("pDEST_HC_Rec_Bxb_v2_CBS_WT:5324:GTG:TAT", "pDEST_HC_Rec_Bxb_v2_CBS_WT:5904:TG:GT")

```

```{r limma- RACE-like chemistry with WT normalization}

racelike_fdr_thresh<- 0.1
racelike_log_caf_thresh<- -12

cbs_02dec2021_racelike_postprocess_subtract_filt_counts_dt<- 
    cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="RACE-like" & log_CAF_norm >= racelike_log_caf_thresh, .N, by=.(VAR_ID)]

limma_racelike_complete<- cbs_02dec2021_racelike_postprocess_subtract_filt_counts_dt[
  N==6, unique(VAR_ID)]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    Chemistry=="RACE-like" & VAR_ID%in%limma_racelike_complete]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_dt<- dcast(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt, 
  formula=VAR_ID~Sample, value.var="log_CAF_norm")

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_mat<- as.matrix(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_dt[,-c("VAR_ID")])

rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_mat)<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_dt[,VAR_ID]

# Make a design matrix
cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_design_dt<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
    ,as.factor(unique(Input)), by=.(Sample)]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design_mat<- 
  as.matrix(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_design_dt[,-c("Sample")])

rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design_mat)<- 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_design_dt[,Sample]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design<- 
  model.matrix(~V1, data=as.data.frame(
    cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design_mat), 
    list(V1=contr.treatment(n=2, base=2)))

# Make ExpressionSet
cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset<- 
  Biobase::ExpressionSet(assayData=cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_mat)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit<- 
  lmFit(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset,
        cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb<-
  eBayes(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt<- as.data.table(
  topTable(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb, 
           coef=ncol(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design), 
           number=nrow(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb), 
           sort.by="none"))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt[
  ,VAR_ID:=rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb)]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top<- topTable(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb, coef=ncol(
      cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_design), number=50)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_sig<- subset(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top$adj.P.Val < racelike_fdr_thresh)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids<- rownames(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_sig)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids<- rownames(subset(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top$logFC>0))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids<- rownames(subset(
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top, 
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top$logFC<0))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  ,ID_SHORT:=paste(substr(AA_CHANGE, 3, 100), paste(REF, ALT, sep=">"), sep=", ")]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  ,ID_LONG:=paste(AA_CHANGE, POS, REF, ALT, sep=":")]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  VAR_ID%in%cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_up_varids,
  Direction:="RNA_up"]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  VAR_ID%in%cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_down_varids,
  Direction:="RNA_down"]

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[,Direction:=as.factor(Direction)]

racelike_gDNA_med_log10_caf<- cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  Input=="gDNA", median(log_CAF_norm)]

racelike_cDNA_med_log10_caf<- cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
  Input=="cDNA", median(log_CAF_norm)]

aaas_cols<- pal_aaas()(2)

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_norm_dt<-
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_dt[
    VAR_ID%in%cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids]

# Visualize the frequencies
cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_top_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_norm_dt,
  aes(x=log_CAF_norm, y=reorder(ID_SHORT, AA_POS2), col=Input))

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_top_gg+
  geom_vline(xintercept=racelike_cDNA_med_log10_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=racelike_gDNA_med_log10_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="Enrichment score")+
  scale_color_aaas()+
  theme_cowplot()

```

```{r limma- RACE-like chemistry normalized vs non-normalized}

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_compare_dt<- 
  merge(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_dt[
    ,.(VAR_ID, logFC, adj.P.Val, AveExpr)],
        cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt[
          ,.(VAR_ID, logFC, adj.P.Val, AveExpr)], by="VAR_ID")

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_compare_dt[
  ,AveExpr:=(AveExpr.x+AveExpr.y)/2]

racelike_limma_compare_gg<- ggplot(
  data=cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_compare_dt, 
  aes(x=logFC.x, y=logFC.y, col=AveExpr))

racelike_limma_compare_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point()+
  coord_cartesian(xlim=c(-2.5,2.5), ylim=c(-2.5,2.5))+
  scale_color_viridis(option="C")+
  theme_cowplot()+
  labs(x="Non-normalized", y="WT-normalized", 
       col="Mean log
frequency")

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_sig_compare_dt[
  ,cor.test(x=logFC.x, y=logFC.y)]

```

```{r limma- amplicon in RACE-like}

amplicon_in_racelike_mat<- dcast(
  data=cbs_02dec2021_both_postprocess_subtract_rf_drop_dt[Chemistry=="RACE-like"], 
  formula=VAR_ID~Sample, value.var="log_CAF_norm")

amplicon_in_racelike_mat_varids<- amplicon_in_racelike_mat[,1]

amplicon_in_racelike_mat<- amplicon_in_racelike_mat[,-1]
amplicon_in_racelike_mat<- apply(amplicon_in_racelike_mat, 2, as.numeric)

# Set nonobserved variants to the approximate limit of detection for RACE-like libraries
# take the min frequency observed among replicates, then find the max of these and use as LOD
racelike_lod<- max(apply(amplicon_in_racelike_mat, 2, min, na.rm=T))

amplicon_in_racelike_mat_nas<- apply(
  amplicon_in_racelike_mat, 1, function(x) sum(is.na(x)))

amplicon_in_racelike_mat[is.na(amplicon_in_racelike_mat)]<- racelike_lod
amplicon_in_racelike_dt<- as.data.table(amplicon_in_racelike_mat)
amplicon_in_racelike_dt[,VAR_ID:=amplicon_in_racelike_mat_varids]
amplicon_in_racelike_dt[,N_NAs:=amplicon_in_racelike_mat_nas]
rm(amplicon_in_racelike_mat)

amplicon_in_racelike_sig_dt<- amplicon_in_racelike_dt[
    VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids_norm]

amplicon_in_racelike_sig_dt[,Chemistry:=factor("RACE-like", levels=c("Amplicon", "RACE-like"))]
amplicon_in_racelike_sig_dt[,N_NAs:=as.factor(N_NAs)]

amplicon_in_racelike_sig_melt_dt<- melt(
  data=amplicon_in_racelike_sig_dt, id.vars=c("Chemistry", "VAR_ID", "N_NAs"), 
  variable.name="Sample", value.name="log_CAF_norm")

# Merge with the amplicon data, then summarize and cast
amplicon_in_racelike_sig_merge_dt<- rbind(
  amplicon_in_racelike_sig_melt_dt[,-c("N_NAs")], 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_dt[
    VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_top_varids_norm,
    .(Chemistry, VAR_ID, Sample, log_CAF_norm)], fill=TRUE)

amplicon_in_racelike_sig_merge_dt[Sample%in%c(gdna_samples_replace), Input:="gDNA"]
amplicon_in_racelike_sig_merge_dt[is.na(Input), Input:="cDNA"]
amplicon_in_racelike_sig_merge_dt[,Input:=as.factor(Input)]

amplicon_in_racelike_sig_merge_cast_dt<- dcast(
  amplicon_in_racelike_sig_merge_dt[
    ,.(Median_log_CAF_norm=median(log_CAF_norm)),
    by=.(Chemistry, Input, VAR_ID)],
  VAR_ID~Chemistry+Input, value.var="Median_log_CAF_norm")

sig_annots_dt<- unique(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_sig_norm_dt[
  ,.(VAR_ID, ID_LONG, ID_SHORT)])

amplicon_in_racelike_sig_merge_cast_annot_dt<- merge(
  amplicon_in_racelike_sig_merge_cast_dt, sig_annots_dt, by="VAR_ID")

amplicon_in_racelike_sig_merge_cast_annot_dt[,Amplicon_diff:=Amplicon_cDNA-Amplicon_gDNA]
amplicon_in_racelike_sig_merge_cast_annot_dt[,RACElike_diff:=`RACE-like_cDNA`-`RACE-like_gDNA`]
amplicon_in_racelike_sig_merge_cast_annot_dt[,delta_delta:=RACElike_diff-Amplicon_diff]

amplicon_in_racelike_sig_merge_cast_annot_dt[,N_NAs:=amplicon_in_racelike_sig_dt[
  match(amplicon_in_racelike_sig_merge_cast_annot_dt[,VAR_ID], VAR_ID), N_NAs]]

amplicon_in_racelike_sig_merge_cast_annot_dt[,N_NAs:=as.factor(N_NAs)]

amplicon_in_racelike_gg<- ggplot(
  data=amplicon_in_racelike_sig_merge_cast_annot_dt[
    !is.na(delta_delta) & N_NAs=="0"],
  aes(x=Amplicon_diff, y=RACElike_diff, label=ID_SHORT))

# col="Number NAs (of 6)"
amplicon_in_racelike_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-1.5,1.5), ylim=c(-1.5,1.5))+
  geom_point(size=4, col="steelblue")+
  geom_text_repel(box.padding=1)+
  labs(x="Amplicon log ratio", 
       y="RACE-like log ratio")+
  theme_cowplot()

amplicon_in_racelike_sig_merge_cast_annot_dt[
  !(is.na(Amplicon_diff)|is.na(RACElike_diff)),
  cor.test(Amplicon_diff, RACElike_diff, method="pearson")]

amplicon_in_racelike_sig_merge_cast_annot_dt[
  !(is.na(Amplicon_diff)|is.na(RACElike_diff)),
  cor.test(Amplicon_diff, RACElike_diff, method="spearman")]

# Correlation is quite good for variants observed in all replicates
# for each library preparation method
amplicon_in_racelike_sig_merge_cast_annot_dt[
  N_NAs=="0" & !(is.na(Amplicon_diff)|is.na(RACElike_diff)), 
  cor.test(Amplicon_diff, RACElike_diff, method="pearson")]


```
 
## Amplicon variant effect map

```{r Amplicon VE heatmap}

cbs_02dec2021_both_postprocess_subtract_counts_by_input_dt<- 
  cbs_02dec2021_both_postprocess_subtract_rf_dt[
    ,.N, by=.(Chemistry, Input, VAR_ID)]

amplicon_ve_var_ids<- cbs_02dec2021_both_postprocess_subtract_counts_by_input_dt[
  Chemistry=="Amplicon" & N>=2, unique(VAR_ID)]

amplicon_summ_cast_dt<- dcast(
  data=amplicon_dt[
    VAR_ID%in%amplicon_complete,
    .(Mean_log_CAF_norm=mean(log_CAF_norm, na.rm=TRUE)), 
    by=.(Input, AA_POS, REF_AA, ALT_AA)], 
  AA_POS+REF_AA+ALT_AA~Input, value.var="Mean_log_CAF_norm")

amplicon_summ_cast_dt[,cDNA_gDNA_diff:=cDNA-gDNA]
amplicon_summ_cast_dt[,AA_POS:=as.integer(AA_POS)]

amplicon_summ_cast_dt[AA_POS<80, Tile:="Tile_2"]
amplicon_summ_cast_dt[AA_POS>100, Tile:="Tile_4"]

# Reorder ALT AA by amino acid chemistry
amplicon_summ_cast_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
amplicon_summ_cast_dt[,ALT_AA_level:=as.integer(ALT_AA)]

amplicon_ve_heatmap_gg<- ggplot(
  data=amplicon_summ_cast_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=cDNA_gDNA_diff))

amplicon_ve_heatmap_gg+
  scale_x_continuous(breaks=seq(35, 150, 5))+
  facet_wrap(~Tile, scales="free_x")+
  geom_tile()+
  scale_fill_gradient2()+
  labs(x="Amino acid position", y="Alternate", 
       fill="cDNA-gDNA
log ratio")+
  theme_cowplot()

```

## Compare heme and catalytic domains in amplicon method

Determine if there areposition-specific effects.

```{r Min diff by domain}

amplicon_summ_cast_by_pos_dt<- amplicon_summ_cast_dt[
  ,.(Min_diff=min(cDNA_gDNA_diff)), by=.(AA_POS, Tile)]

amplicon_ve_by_pos_gg<- ggplot(
  data=amplicon_summ_cast_by_pos_dt, 
  aes(x=Tile, y=Min_diff, col=Tile))

amplicon_ve_by_pos_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5)+
  scale_color_jco()+
  geom_jitter(width=0.2)+
  labs(x="Amino acid position",
       y="Min cDNA-gDNA log ratio per AA")+
  theme_cowplot()

levene.test(y=amplicon_summ_cast_by_pos_dt[,Min_diff], 
            group=amplicon_summ_cast_by_pos_dt[,Tile], kruskal.test=TRUE)

amplicon_summ_cast_by_pos_dt[,wilcox.test(formula=Min_diff~Tile, data=.SD)]
    
```

## Investigate heme containing proteins

Determine if there are shared properties in sequence for heme binding regions.

```{r Heme proteins}

roth_fc_dir<- paste(base_dir, "CBS_02DEC2021", "Genome_Medicine_2020_suppl", sep="/")
roth_yeast_fc_st6<- "13073_2020_711_MOESM8_ESM.xlsx"
roth_yeast_fc_st6_uniprot_fa<- paste(roth_fc_dir, "Heme_proteins.fa", sep="/")
roth_yeast_fc_st6_uniprot_fa_clean<- paste(roth_fc_dir, "Heme_proteins_shortname.fa", sep="/")
roth_yeast_fc_st6_uniprot_fa_clean_center<- paste(roth_fc_dir, "Heme_proteins_Fe_features_slop.fa", sep="/")
roth_yeast_fc_st6_uniprot_fa_clean_ds<- paste(roth_fc_dir, "Heme_proteins_Fe_features_downstream.fa", sep="/")

roth_yeast_fc_st6_dt<- as.data.table(read_xlsx(paste(roth_fc_dir, roth_yeast_fc_st6, sep="/"), skip=1, na="NA"))
names(roth_yeast_fc_st6_dt)<- gsub(pattern="...", replacement="_", x=names(roth_yeast_fc_st6_dt), fixed=TRUE)
names(roth_yeast_fc_st6_dt)<- gsub(pattern=" ", replacement="_", x=names(roth_yeast_fc_st6_dt), fixed=TRUE)

roth_yeast_fc_st6_heme_dt<- roth_yeast_fc_st6_dt[grepl("Heme", Cofactor)]
roth_yeast_fc_st6_heme_uniprot_ids<- roth_yeast_fc_st6_heme_dt[,unique(UniprotID)]

roth_yeast_fc_st6_plp_dt<- roth_yeast_fc_st6_dt[grepl("Pyridoxal", Cofactor)]
roth_yeast_fc_st6_plp_uniprot_ids<-  roth_yeast_fc_st6_plp_dt[,unique(UniprotID)]

write.table(roth_yeast_fc_st6_heme_uniprot_ids, paste(
  roth_fc_dir, "13073_2020_711_MOESM8_ESM_heme_Uniprot_ids.txt", sep="/"), 
  sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)

write.table(roth_yeast_fc_st6_heme_uniprot_ids, paste(
  roth_fc_dir, "13073_2020_711_MOESM8_ESM_PLP_Uniprot_ids.txt", sep="/"), 
  sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)


# Read in the protein sequences and reformat the header for intersection
roth_yeast_fc_st6_uniprot_fa_ss<- readAAStringSet(
  roth_yeast_fc_st6_uniprot_fa, format="fasta", use.names=TRUE)

roth_yeast_fc_st6_uniprot_fa_ss_shortname<- sapply(strsplit(names(
  roth_yeast_fc_st6_uniprot_fa_ss), "|", fixed=T), "[", 2)

roth_yeast_fc_st6_uniprot_fa_ss_clean<- copy(roth_yeast_fc_st6_uniprot_fa_ss)
names(roth_yeast_fc_st6_uniprot_fa_ss_clean)<- roth_yeast_fc_st6_uniprot_fa_ss_shortname

writeXStringSet(roth_yeast_fc_st6_uniprot_fa_ss_clean, 
                roth_yeast_fc_st6_uniprot_fa_clean, format="fasta")


# Get the GFF and FASTA info from Uniprot
# After extracting the regions spanning and downstream of the coordinating residue, 
# read in results and determine differences in amino acid composition
roth_yeast_fc_st6_uniprot_fa_ss_clean_center<- readAAStringSet(
  roth_yeast_fc_st6_uniprot_fa_clean_center, format="fasta", use.names=TRUE)

roth_yeast_fc_st6_uniprot_fa_ss_clean_ds<- readAAStringSet(
  roth_yeast_fc_st6_uniprot_fa_clean_ds, format="fasta", use.names=TRUE)

# Get length of each sequence
roth_yeast_fc_st6_uniprot_fa_ss_clean_dt<- data.table(
  seq=as.character(roth_yeast_fc_st6_uniprot_fa_ss_clean))

# Generate 100 random subsequences from each protein for control sequences
gen_random_seqs<- function(seq, n=100, window=50){
  max_index<- nchar(seq) - window
  indices<- sample(seq(1, max_index), size=n, replace=FALSE)
  subsequences<- NULL
  for(i in indices){
    seq_split<- unlist(strsplit(seq, ""))
    subsequence<- paste0(seq_split[seq(i, i+window)], collapse="")
    subsequences<- c(subsequences, subsequence)
  }
  return(subsequences)
}

roth_yeast_fc_st6_uniprot_fa_ss_clean_control_dt<- 
  roth_yeast_fc_st6_uniprot_fa_ss_clean_dt[
  ,.(seq=gen_random_seqs(seq)), by=seq_len(
    nrow(roth_yeast_fc_st6_uniprot_fa_ss_clean_dt))]

roth_yeast_fc_st6_uniprot_fa_ss_clean_control<- AAStringSet(
  roth_yeast_fc_st6_uniprot_fa_ss_clean_control_dt[,seq])

# Generate sequence logos of the centered and downstream amino acids
# near heme binding residues
roth_yeast_fc_st6_uniprot_fa_ss_clean_center_seqlogo<- ggseqlogo(
  data=as.character(roth_yeast_fc_st6_uniprot_fa_ss_clean_center))

roth_yeast_fc_st6_uniprot_fa_ss_clean_ds_seqlogo<- ggseqlogo(
  data=as.character(roth_yeast_fc_st6_uniprot_fa_ss_clean_ds))

roth_yeast_fc_st6_uniprot_fa_ss_clean_control_seqlogo<- ggseqlogo(
  data=as.character(roth_yeast_fc_st6_uniprot_fa_ss_clean_control))

roth_yeast_fc_st6_uniprot_fa_ss_clean_center_seqlogo+
  theme(axis.text.x=element_text(angle=90))

roth_yeast_fc_st6_uniprot_fa_ss_clean_ds_seqlogo+
  theme(axis.text.x=element_text(angle=90))

#base_priors<- colMeans(alphabetFrequency(query_seqs, as.prob=TRUE, baseOnly=TRUE))
#base_priors<- base_priors[-length(base_priors)]

# log2probratio
#pwm<- PWM(query_seqs, type=c("prob"), prior.params=base_priors)
#pwm_melt<- as.data.table(reshape2::melt(pwm))
#names(pwm)<- c("Base", "Position", "Posterior_prob")


```

## tRNA-seq correlation with limma differential variants

Determine if tRNAs for any of the alternate codons are more abundant than the reference codon and correlate this difference with logFCs for variants.

Use data from:
High-resolution quantitative profiling of tRNA abundance and modification status in eukaryotes by mim-tRNAseq
https://pubmed.ncbi.nlm.nih.gov/33581077/

Characterizing Expression and Processing of Precursor and Mature Human tRNAs by Hydro-tRNAseq and PAR-CLIP (Hydro-tRNAseq)
https://pubmed.ncbi.nlm.nih.gov/28793268/

## Codon stability and tRNA abundance comparison

Determine if variants that increase/decrease abundance have altered codon optimality/stability and/or tRNA abundance.

```{r Codon stability coefficient comparison}

codon_opti_fdr_thresh<- 0.15

codon_optimality_dir<- paste(base_dir, "CBS_02DEC2021", "codon_optimality", sep="/")
codon_optimality_wu_2019<- paste(codon_optimality_dir, "elife-45396-fig1-data2-v2.csv", sep="/")
codon_optimality_wu_2019_dt<- fread(codon_optimality_wu_2019, sep=",", header=TRUE)
codon_optimality_wu_2019_dt[,ALT_CODON:=codon]

cbs_02dec2021_amplicon_postprocess_subtract_filt_uniq_dt<- unique(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_dt[
    ,.(VAR_ID, REF, ALT, REF_CODON, ALT_CODON, AA_CHANGE, AA_POS)])

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt<- 
  merge(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt,
        cbs_02dec2021_amplicon_postprocess_subtract_filt_uniq_dt, by="VAR_ID")

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,REF_CODON_OPTI:=codon_optimality_wu_2019_dt[
    match(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[,REF_CODON], codon), `293T_ORFome`]]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,ALT_CODON_OPTI:=codon_optimality_wu_2019_dt[
    match(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], codon), `293T_ORFome`]]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,CODON_OPTI_diff:=ALT_CODON_OPTI-REF_CODON_OPTI]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_fdr_varids<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_dt[
  adj.P.Val < codon_opti_fdr_thresh, unique(VAR_ID)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_fdr_varids,
  Significant:="True"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  is.na(Significant), Significant:="False"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  logFC>0 & Significant=="True", Direction:="RNA up"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  logFC<0 & Significant=="True", Direction:="RNA down"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  Significant=="False", Direction:="N.S."]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,Direction:=factor(Direction, levels=c("N.S.", "RNA down", "RNA up"))]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  Direction!="N.S." & (CODON_OPTI_diff>0.05 | CODON_OPTI_diff<(-.1)), Label:=gsub("p.", "", AA_CHANGE)]

codon_optimality_merge_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt,
  aes(x=Direction, y=CODON_OPTI_diff))

jitter_pos<- position_jitter(width=0.25, seed=9)

codon_optimality_merge_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=CODON_OPTI_diff, col=adj.P.Val), position=jitter_pos, inherit.aes=FALSE)+
  geom_text_repel(data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=CODON_OPTI_diff, label=Label), box.padding=0.7, position=jitter_pos, inherit.aes=FALSE)+
  labs(x="", y="Alt. - Ref. CSC", col="FDR")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

# RNA down should come before RNA up for alternative=less
cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  !is.na(CODON_OPTI_diff) & Significant=="True", wilcox.test(
    formula=CODON_OPTI_diff~Direction, data=.SD, alternative="less", exact=FALSE)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  !is.na(CODON_OPTI_diff) & Direction!="RNA up", wilcox.test(
    formula=CODON_OPTI_diff~Direction, data=.SD, alternative="greater")]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  !is.na(CODON_OPTI_diff) & Direction!="RNA down", wilcox.test(
    formula=CODON_OPTI_diff~Direction, data=.SD, alternative="less")]


cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_fdr_aa_changes<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  VAR_ID%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_fdr_varids,
  unique(AA_CHANGE)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
    AA_CHANGE%in%cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_fdr_aa_changes]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_counts_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_dt[
    ,.N,by=.(AA_CHANGE)]

# Remove variants that do not have 1 N.S. and 1 RNA up/down for comparison
codon_opti_comp_aa_changes<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_counts_dt[
    N!=1, unique(AA_CHANGE)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt<- 
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_dt[
    AA_CHANGE%in%codon_opti_comp_aa_changes]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  ,Group:=as.integer(as.factor(AA_CHANGE))]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  ,Direction:=factor(Direction, levels=c("RNA down", "RNA up", "N.S."))]

# Further remove variants with more than two variants for each AA
codon_opti_comp_aa_changes_drop<- c("p.A38C")
codon_opti_inconsistent_aa_changes<- c("p.H67T", "p.A130S", "p.P41T", "p.C109G", "p.E144L")
codon_opti_inconclusive_aa_changes<- c("p.P59L", "p.H66R")

# cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[,AA_CHANGE_class:=NULL]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  , AA_CHANGE_class:="Agree"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  AA_CHANGE%in%codon_opti_inconsistent_aa_changes, AA_CHANGE_class:="Disagree"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  AA_CHANGE%in%codon_opti_inconclusive_aa_changes, AA_CHANGE_class:="Inconclusive"]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  ,AA_CHANGE_class:=factor(AA_CHANGE_class)]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
  ,length(unique(AA_CHANGE)), by=.(AA_CHANGE_class)]

# Compare only amino acids with two variants
# pbinom(13,18,0.5,lower.tail=FALSE)
# 0.01544189

# Compare amino acids with more than two variants that are inconclusive
# pbinom(13,20,0.5,lower.tail=FALSE)
# 0.05765915

codon_opti_cols<- c("darkorange", "dodgerblue", "black")

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_final_dt<-
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_dt[
    !AA_CHANGE%in%codon_opti_comp_aa_changes_drop]

# Determine the null distribution of CSC diff scores
csc_diff_dist<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,as.numeric(dist(CODON_OPTI_diff, method="manhattan"))]

csc_diff_dist_nonas<- csc_diff_dist[!is.na(csc_diff_dist)]

csc_diff_dist_1st_quartile<- quantile(csc_diff_dist_nonas, probs=0.25)
csc_diff_dist_3rd_quartile<- quantile(csc_diff_dist_nonas, probs=0.75)

csc_diff_cast_dt<- dcast(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_final_dt, 
  AA_CHANGE~VAR_ID, value.var="CODON_OPTI_diff")

get_diff<- function(x){
  x_vec<- as.vector(x)
  x_no_na<- sort(x_vec[!is.na(x_vec)])
  x_diff<- abs(x_no_na[length(x_no_na)] - x_no_na[1])
  return(x_diff)
}

csc_diff_cast_dt$csc_diff<- apply(csc_diff_cast_dt[,-c("AA_CHANGE")], 1, get_diff)
csc_diff_sig_aa_changes<- csc_diff_cast_dt[csc_diff>=csc_diff_dist_3rd_quartile, unique(AA_CHANGE)]
csc_diff_nonsig_aa_changes<- csc_diff_cast_dt[csc_diff<csc_diff_dist_1st_quartile, unique(AA_CHANGE)]

# Plot only those variants in the expected direction with a difference in the upper quartile
codon_opti_comp_aa_changes_gg<- ggplot(
  data=cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_opti_comp_final_dt[
    !AA_CHANGE%in%csc_diff_nonsig_aa_changes & AA_CHANGE%in%csc_diff_sig_aa_changes & AA_CHANGE_class=="Agree"],
  aes(y=reorder(AA_CHANGE, CODON_OPTI_diff, sum), x=CODON_OPTI_diff))

codon_opti_comp_aa_changes_gg+
  facet_wrap(~AA_CHANGE_class, scales="free_y")+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  scale_color_manual(values=codon_opti_cols)+
  geom_line(aes(group=Group))+
  geom_point(aes(col=Direction), size=5)+
  coord_cartesian(xlim=c(-0.2, 0.2))+
  labs(y="Amino acid", x="Alt. - Ref. CSC", col="Direction")+
  theme_cowplot()

#pbinom(9,14,0.5,lower.tail=FALSE)

```

```{r tRNA abundance amplicon}

trna_fdr_thresh<- 0.15

tRNA_dir<- paste(base_dir, "CBS_02DEC2021", "tRNA", sep="/")

hydro_tRNAseq_raw_dt<- fread(paste(tRNA_dir, "hydrotrnaseq_raw.txt", sep="/"), sep="\t", header=T)

mim_tRNAseq_raw_dt<- as.data.table(
  read_xlsx(paste(tRNA_dir, "1-s2.0-S1097276521000484-mmc2.xlsx", sep="/"), skip=3, na="NA"))

# Hydro-tRNAseq summarization
hydro_tRNAseq_raw_agg_dt<- hydro_tRNAseq_raw_dt[
  ,.(Sum_count=sum(`Read count`), log_sum_count=log(sum(`Read count`))), 
  by=.(Isotype, Anticodon, Codon)]

#' Generates expression entries for wobble codons for Hydro-tRNAseq dataset
#'
#' @param hydrotrnaseq_summarized_dt data.table containing summarized expression values for tRNAs
#' @return data.table original data.table with entries for wobble codons
expand_wobble_pairs<- function(hydrotrnaseq_summarized_dt){
  
  wobble_anticodons<- hydrotrnaseq_summarized_dt[substr(Anticodon, 1, 1)%in%c("G", "T"), unique(Anticodon)]
  
  # For each anticodon with a G or T at the 5' end, there can be one other codon that can pair with it
  n_wobble_anticodons<- length(wobble_anticodons)

  # Initialize the data.table
  wobble_codon_dt<- data.table(Isotype=vector(mode="character", length=n_wobble_anticodons),
                               Anticodon=vector(mode="character", length=n_wobble_anticodons),
                               Codon=vector(mode="character", length=n_wobble_anticodons), 
                               Sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               log_sum_count=vector(mode="numeric", length=n_wobble_anticodons))
  
  # Iterate over the anticodons that can form wobble pairing, and create entries for wobble codons
  for(i in 1:length(wobble_anticodons)){
    anticodon<- wobble_anticodons[i]
    wobble_bp<- substr(anticodon, 1, 1)
    
    # These wobble pairing rules are the original described by Crick, which are actually conservative
    if(wobble_bp=="G"){
      wobble_anticodon<- paste0("A", substr(anticodon, 2, 3))
    } else if(wobble_bp=="T"){
      wobble_anticodon<- paste0("C", substr(anticodon, 2, 3))
    }
    
    wobble_codon<- revcomp(wobble_anticodon)
    wobble_aa<- names(AA_CODONS_LIST)[sapply(AA_CODONS_LIST, is_in, wobble_codon)]
    
    # Use the tRNA expression values of the canonical anticodon for the wobble codon
    wobble_codon_dt[i,]<- data.table(Isotype=wobble_aa, Anticodon=anticodon, Codon=wobble_codon, 
                                     Sum_count=hydrotrnaseq_summarized_dt[Anticodon==anticodon, Sum_count],
                                     log_sum_count=hydrotrnaseq_summarized_dt[Anticodon==anticodon, log_sum_count])
  }
  
  complete_dt<- rbind(hydrotrnaseq_summarized_dt, wobble_codon_dt, fill=T)
  setkey(complete_dt, Isotype, Anticodon, Codon)
  return(complete_dt)
}

# We accrue 23 codons by taking into account wobble pairs
hydro_tRNAseq_raw_agg_wobble_dt<- expand_wobble_pairs(hydro_tRNAseq_raw_agg_dt)

# Summarize at the level of codon after adding wobble pairs
hydro_tRNAseq_raw_agg_wobble_summ_dt<- hydro_tRNAseq_raw_agg_wobble_dt[
  ,.(Sum_count=sum(Sum_count), log_sum_count=log(sum(Sum_count))), 
  by=.(Isotype, Codon)]

# mim-tRNAseq summarization
mim_tRNAseq_raw_dt[,Anticodon:=sapply(strsplit(`Unique tRNA`, "-"), "[", 2)]
mim_tRNAseq_raw_dt[,Isotype:=sapply(strsplit(`Unique tRNA`, "-"), "[", 1)]
mim_tRNAseq_raw_dt[,HEK293T_mean:=(`HEK293T rep 1` + `HEK293T rep 2`)/2]
mim_tRNAseq_raw_dt[,Codon:=revcomp(Anticodon), by=seq_len(nrow(mim_tRNAseq_raw_dt))]
mim_tRNAseq_raw_dt[,Isotype:=names(aa_map)[sapply(aa_map, is_in, Isotype)], 
                   by=seq_len(nrow(mim_tRNAseq_raw_dt))]

mim_tRNAseq_raw_filt_dt<- mim_tRNAseq_raw_dt[!grepl("iMet|eColiLys|SeC|Sup", `Unique tRNA`)]

mim_tRNAseq_raw_filt_agg_dt<- mim_tRNAseq_raw_filt_dt[
  ,.(Sum_count=sum(HEK293T_mean), 
     log_sum_count=log(sum(HEK293T_mean)+1), 
     Sum_copy_number=sum(`Gene copy number`)),
  by=.(Isotype, Codon, Anticodon)]


#' Generates expression entries for wobble codons for mim-tRNAseq dataset
#'
#' @param mimtrnaseq_summarized_dt data.table containing summarized expression values for tRNAs
#' @return data.table original data.table with entries for wobble codons
expand_wobble_pairs<- function(mimtrnaseq_summarized_dt){
  
  wobble_anticodons<- mimtrnaseq_summarized_dt[substr(Anticodon, 1, 1)%in%c("G", "T"), unique(Anticodon)]
  
  # For each anticodon with a G or T at the 5' end, there can be one other codon that can pair with it
  n_wobble_anticodons<- length(wobble_anticodons)

  # Initialize the data.table
  wobble_codon_dt<- data.table(Isotype=vector(mode="character", length=n_wobble_anticodons),
                               Anticodon=vector(mode="character", length=n_wobble_anticodons),
                               Codon=vector(mode="character", length=n_wobble_anticodons), 
                               Sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               log_sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               Sum_copy_number=vector(mode="integer", length=n_wobble_anticodons))
  
  # Iterate over the anticodons that can form wobble pairing, and create entries for wobble codons
  for(i in 1:length(wobble_anticodons)){
    anticodon<- wobble_anticodons[i]
    wobble_bp<- substr(anticodon, 1, 1)
    
    # These wobble pairing rules are the original described by Crick, which are actually conservative
    if(wobble_bp=="G"){
      wobble_anticodon<- paste0("A", substr(anticodon, 2, 3))
    } else if(wobble_bp=="T"){
      wobble_anticodon<- paste0("C", substr(anticodon, 2, 3))
    }
    
    wobble_codon<- revcomp(wobble_anticodon)
    wobble_aa<- names(AA_CODONS_LIST)[sapply(AA_CODONS_LIST, is_in, wobble_codon)]
    
    # Use the tRNA expression values of the canonical anticodon for the wobble codon
    wobble_codon_dt[i,]<- data.table(Isotype=wobble_aa, Anticodon=anticodon, Codon=wobble_codon, 
                                     Sum_count=mimtrnaseq_summarized_dt[Anticodon==anticodon, Sum_count],
                                     log_sum_count=mimtrnaseq_summarized_dt[Anticodon==anticodon, log_sum_count],
                                     Sum_copy_number=mimtrnaseq_summarized_dt[Anticodon==anticodon, Sum_copy_number])
  }
  
  complete_dt<- rbind(mimtrnaseq_summarized_dt, wobble_codon_dt, fill=T)
  setkey(complete_dt, Isotype, Anticodon, Codon)
  return(complete_dt)
}

mim_tRNAseq_raw_filt_agg_wobble_dt<- expand_wobble_pairs(mim_tRNAseq_raw_filt_agg_dt)

mim_tRNAseq_raw_filt_agg_wobble_nostop_dt<- mim_tRNAseq_raw_filt_agg_wobble_dt[
  !Codon%in%c("TAG", "TAA", "TGA")]

# Finally aggregate at the level of codon 
mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt<-
  mim_tRNAseq_raw_filt_agg_wobble_nostop_dt[
    ,.(Sum_count=sum(Sum_count), log_sum_count=log(sum(Sum_count)+1), 
       Sum_copy_number=sum(Sum_copy_number)), 
    by=.(Isotype, Codon)]


# tRNA abundance with amplicon loosened FDR
cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,REF_CODON_hydro_log_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), log_sum_count]]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,ALT_CODON_hydro_log_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), log_sum_count]]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,hydro_log_sum_diff:=ALT_CODON_hydro_log_sum-REF_CODON_hydro_log_sum]


cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,REF_CODON_mim_log_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), log_sum_count]]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,ALT_CODON_mim_log_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), log_sum_count]]

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  ,mim_log_sum_diff:=ALT_CODON_mim_log_sum-REF_CODON_mim_log_sum]

mim_trna_dt<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
    !is.na(mim_log_sum_diff) & adj.P.Val < trna_fdr_thresh]

if("Label"%in%names(mim_trna_dt)){
  mim_trna_dt[,Label:=NULL]
}

mim_trna_dt[logFC<(-1.5), Label:=gsub("p.", "", AA_CHANGE)]

mim_trnaseq_spline_fit<- lm(data=mim_trna_dt[logFC<0], formula=logFC~splines::bs(mim_log_sum_diff, knots=0, degree=2))
mim_trnaseq_spline_pred<- predict(mim_trnaseq_spline_fit, newdata=mim_trna_dt[logFC<0], se=T)

mim_trnaseq_spline_dt<- data.table(
  x_val=mim_trna_dt[logFC<0, mim_log_sum_diff], 
  y_val=mim_trnaseq_spline_pred[["fit"]], 
  SE=mim_trnaseq_spline_pred[["se.fit"]])

mim_trnaseq_cor_gg<- ggplot(
  data=mim_trna_dt[logFC<0], 
  aes(x=mim_log_sum_diff, y=logFC, col=adj.P.Val, label=Label))

mim_trnaseq_cor_gg+
  geom_text_repel(box.padding=1, col="black")+
  geom_point(size=3)+
  geom_line(data=mim_trnaseq_spline_dt, 
            aes(x=x_val, y=y_val), inherit.aes=FALSE, col="blue")+
  coord_cartesian(ylim=c(-3, -0.5))+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="mim-tRNAseq log(sum),
ALT to REF codon", 
       y="log FC, cDNA to gDNA", col="FDR")+
  theme_cowplot()

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  !is.na(mim_log_sum_diff) & adj.P.Val < trna_fdr_thresh & logFC < 0, 
  cor.test(mim_log_sum_diff, logFC, method="spearman", exact=FALSE)]

hydroseq_dt<- cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
    !is.na(hydro_log_sum_diff) & adj.P.Val < trna_fdr_thresh]

if("Label"%in%names(hydroseq_dt)){
  hydroseq_dt[,Label:=NULL]
}

hydroseq_dt[logFC<(-1.5), Label:=gsub("p.", "", AA_CHANGE)]

hydroseq_spline_fit<- lm(data=hydroseq_dt[logFC<0], formula=logFC~splines::bs(mim_log_sum_diff, knots=0, degree=2))
hydroseq_spline_pred<- predict(hydroseq_spline_fit, newdata=hydroseq_dt[logFC<0], se=T)

hydroseq_spline_dt<- data.table(
  x_val=hydroseq_dt[logFC<0, mim_log_sum_diff], 
  y_val=hydroseq_spline_pred[["fit"]], 
  SE=hydroseq_spline_pred[["se.fit"]])

hydro_trnaseq_cor_gg<- ggplot(
  data=hydroseq_dt[logFC<0], 
  aes(x=hydro_log_sum_diff, y=logFC, col=adj.P.Val, label=Label))

hydro_trnaseq_cor_gg+
  geom_text_repel(box.padding=1, col="black")+
  geom_point(size=3)+
  geom_line(data=hydroseq_spline_dt, 
            aes(x=x_val, y=y_val), inherit.aes=FALSE, col="blue")+
  coord_cartesian(ylim=c(-3, -0.5))+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="Hydro-tRNAseq log(sum),
ALT to REF codon", 
       y="log FC, cDNA to gDNA", col="FDR")+
  theme_cowplot()

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_merge_dt[
  !is.na(hydro_log_sum_diff) & adj.P.Val < trna_fdr_thresh & logFC < 0, 
  cor.test(hydro_log_sum_diff, logFC, method="spearman")]

```

## Compare yeast functional complementation data with differential variants

Determine if variants that increase or decrease abundance impact fitness in yeast.

```{r Correlate logFC of cDNA to gDNA to fitness of variants in yeast functional complementation}

yfc_fdr_thresh<- 0.15

roth_fc_dir<- paste(base_dir, "CBS_02DEC2021", "Genome_Medicine_2020_suppl", sep="/")
roth_yeast_fc_st1<- "13073_2020_711_MOESM1_ESM.xlsx"
roth_yeast_fc_st4<- "13073_2020_711_MOESM6_ESM.xlsx"

roth_yeast_fc_st4_dt<- as.data.table(read_xlsx(paste(roth_fc_dir, roth_yeast_fc_st4, sep="/"), skip=2, na="NA"))
names(roth_yeast_fc_st4_dt)<- gsub(pattern="...", replacement="_", x=names(roth_yeast_fc_st4_dt), fixed=TRUE)

#llr_gg<- ggplot(data=roth_yeast_fc_st4_dt, aes(x=LLR, y=score_6))
#llr_gg+geom_point()+geom_smooth()+labs(x="log-likelihood ratio", y="High vitamin B6 score")+theme_cowplot()

# Relabel amino acids
roth_yeast_fc_st4_dt[,AA_CHANGE:=convert_aa_notation(HGVS), by=seq_len(nrow(roth_yeast_fc_st4_dt))]

# Annotate the amplicon and RACE-like datasets with all data

cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_uniq_dt<- unique(
  cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_dt[
    ,.(VAR_ID, REF, ALT, REF_CODON, ALT_CODON, AA_CHANGE, AA_POS)])

cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_merge_dt<- 
  merge(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt,
        cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_uniq_dt, by="VAR_ID")

cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_merge_dt<- 
  merge(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_dt,
        cbs_02dec2021_both_postprocess_subtract_rf_drop_filt_uniq_dt, by="VAR_ID")

amplicon_racelike_merge_cols<- c("VAR_ID", "REF", "ALT", "REF_CODON", "ALT_CODON", "AA_CHANGE", "AA_POS", "logFC", "AveExpr", "adj.P.Val")

# First merge VAR_ID annotations with the RACE-like data
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt<- rbind(
  cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_merge_dt[
    ,..amplicon_racelike_merge_cols],
  cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_merge_dt[
    ,..amplicon_racelike_merge_cols])

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,Chemistry:=as.factor(rep(
    c("Amplicon", "RACE-like"),
    times=c(nrow(cbs_02dec2021_amplicon_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_merge_dt),
            nrow(cbs_02dec2021_racelike_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_all_norm_merge_dt))))]

# Annotate Hydro-tRNAseq counts
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,REF_CODON_hydro_log_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[,REF_CODON], Codon), log_sum_count]]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,ALT_CODON_hydro_log_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[,ALT_CODON], Codon), log_sum_count]]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,hydro_log_sum_diff:=ALT_CODON_hydro_log_sum-REF_CODON_hydro_log_sum]

# Annotate mim-tRNAseq counts
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,REF_CODON_mim_log_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[,REF_CODON], Codon), log_sum_count]]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,ALT_CODON_mim_log_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[,ALT_CODON], Codon), log_sum_count]]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,mim_log_sum_diff:=ALT_CODON_mim_log_sum-REF_CODON_mim_log_sum]

# Annotate CSC score diff
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,REF_CODON_OPTI:=codon_optimality_wu_2019_dt[
    match(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[,REF_CODON], codon), `293T_ORFome`]]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,ALT_CODON_OPTI:=codon_optimality_wu_2019_dt[
    match(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[,ALT_CODON], codon), `293T_ORFome`]]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
  ,CODON_OPTI_diff:=ALT_CODON_OPTI-REF_CODON_OPTI]
  
# Now merge with variants using the AA_CHANGE as key
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt<- merge(
  cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_norm_merge_dt[
    adj.P.Val<yfc_fdr_thresh], 
  roth_yeast_fc_st4_dt, by="AA_CHANGE", all.x=TRUE)

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_missense_down_dt<- 
  cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
    REF_CODON!=ALT_CODON & !grepl("*", AA_CHANGE, fixed=T) & logFC<0,]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_missense_down_dt[
  AA_POS%in%cbs_uniprot_mut_summ_dt[,Positions], .N] / nrow(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_missense_down_dt)


cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  ,Max_score:=max(c(score_6, score_8)), by=seq_len(nrow(
    cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt))]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  !is.na(Max_score), cor(logFC, Max_score, method="spearman")]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  ,logFC_direction:=ifelse(logFC<0, "Down", "Up")]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  ,Score_direction:=ifelse(Max_score<0.5, "Down", "Up")]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  ,LLR_direction:=ifelse(LLR>0, "Down", "Up")]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  ,table(logFC_direction, Score_direction)]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  ,table(logFC_direction, LLR_direction)]

cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt, 
  aes(x=logFC, y=LLR, col=adj.P.Val))

cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="log FC", y="log-likehood ratio
(Sun and Weile et. al., 2020)", col="FDR")+
  theme_cowplot()

# Determine if this is consistent when looking at scores
# High vitamin B6 score
cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt, 
  aes(x=logFC, y=score_6, col=(adj.P.Val)))

cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg+
  geom_hline(yintercept=0.5, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="log FC", y="Score, high vitamin B6, 
(Sun and Weile et. al., 2020)", col="FDR")+
  theme_cowplot()

# Low vitamin B6 score
cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt, 
  aes(x=logFC, y=score_8, col=adj.P.Val))

cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg+
  geom_hline(yintercept=0.5, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="log FC", y="Score, low vitamin B6, 
(Sun and Weile et. al., 2020)", col="FDR")+
  theme_cowplot()

# Max vitamin B6 score
cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt, 
  aes(x=logFC, y=Max_score, col=adj.P.Val))

cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_gg+
  geom_hline(yintercept=0.5, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="log FC", y="Max score, low or high vitamin B6, 
(Sun and Weile et. al., 2020)", col="FDR")+
  theme_cowplot()

# No strong correlations are observed between logFC effect size and YFC score
# Nonetheless, pick out a few variants that show expected/unexpected differences

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_dt<-
  cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
    , .(Chemistry, VAR_ID, REF, ALT, REF_CODON, ALT_CODON, AA_POS, AA_CHANGE, 
        logFC, logFC_direction, AveExpr, adj.P.Val, 
        Max_score, Score_direction, LLR, flags, score_6, stderr_7, score_8, stderr_9, delta, stderr_13,
        hydro_log_sum_diff, mim_log_sum_diff, CODON_OPTI_diff)]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

setkey(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_dt, Chemistry, logFC, Max_score)

write.table(
  cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_dt,
  paste(cbs_02dec2021_dir, "cbs_02dec2021_both_postprocess_subtract_rf_yfc.txt", 
        sep="/"), sep="\t", quote=FALSE, row.names=FALSE)

# pbinom(89,191,0.5,lower.tail=FALSE)

# How many missense variants are at positions implicated in deficiency?
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_dt[
  AA_POS%in%uniprot_variants & REF_CODON!=ALT_CODON & !grepl("*", AA_CHANGE, fixed=T), length(unique(VAR_ID))]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_dt[
    REF_CODON!=ALT_CODON & !grepl("*", AA_CHANGE, fixed=T), length(unique(VAR_ID))]

yfc_final_varids<- c("pDEST_HC_Rec_Bxb_v2_CBS_WT:4673:GC:TG", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4673:G:A", 
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4886:T:G", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4886:TGT:ACG", 
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4991:GA:CT", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4754:CAT:AGG",
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4916:AAG:TTT", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4925:ATC:TTT",
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4695:G:C", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4964:ACG:TGT",
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4943:GAG:ATT", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4757:CAC:TTT",
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4676:AAG:GTT", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4973:CCC:TGT", 
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4733:CGG:ATT", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4715:TGC:AGG",
                     "pDEST_HC_Rec_Bxb_v2_CBS_WT:4749:C:T", "pDEST_HC_Rec_Bxb_v2_CBS_WT:4933:G:T")

# Variants consistent with Weile and Sun 2020 validation dataset
yfc_final_varids_val<- c(
  "pDEST_HC_Rec_Bxb_v2_CBS_WT:4754:CAT:AGG",
  "pDEST_HC_Rec_Bxb_v2_CBS_WT:4923:GC:TT",
  "pDEST_HC_Rec_Bxb_v2_CBS_WT:4736:CCT:TTG",
  "pDEST_HC_Rec_Bxb_v2_CBS_WT:4935:G:A",
  "pDEST_HC_Rec_Bxb_v2_CBS_WT:5631:CC:GT",
  "pDEST_HC_Rec_Bxb_v2_CBS_WT:5972:G:C")

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt<- 
  cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_dt[
    VAR_ID%in%yfc_final_varids & Chemistry=="Amplicon", .(AA_POS, AA_CHANGE, REF_CODON, ALT_CODON, logFC, Max_score)]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt[
  ,logFC:=round(logFC, 3)]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt[
  ,Max_score:=round(Max_score, 3)]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt[
  ,AA_CHANGE:=sub("p.", "", AA_CHANGE, fixed=T)]

cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt[,AA_POS:=as.integer(AA_POS)]
setkey(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt, AA_POS)
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt[,AA_POS:=NULL]

names(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt)<-
  c("Change", "Ref. codon", "Alt. codon", "log FC", "Fitness score [0,1]")

write.table(
  cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_examp_final_dt,
  paste(cbs_02dec2021_dir, "cbs_02dec2021_both_postprocess_subtract_rf_yfc_examp.txt", 
        sep="/"), sep="\t", quote=FALSE, row.names=FALSE)


# Use the validation dataset and compare neutral and deleterious mutations
roth_yeast_fc_st1_dt<- as.data.table(read_xlsx(paste(
  roth_fc_dir, roth_yeast_fc_st1, sep="/"), skip=1, na="NA"))

roth_yeast_fc_st1_dt[,AA_CHANGE:=paste0("p.", `WT AA`, `AA Position`, `Mutant AA`)]

# Merging on exact amino acid change leads to few points for comparison; thus merge on just AA position

# Merge with the validation variant at level of position
roth_yeast_fc_st1_dt[,AA_POS:=`AA Position`]
cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt[
  ,AA_POS:=as.integer(AA_POS)]

cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_val_dt<-
  merge(cbs_02dec2021_both_postprocess_subtract_filt_limma_freq_cast_eset_fit_eb_merge_yfc_dt,
        roth_yeast_fc_st1_dt, by="AA_POS", all.x=TRUE)

cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_val_final_dt<-
  cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_val_dt[
  ,.(Chemistry, VAR_ID, REF_CODON, ALT_CODON, 
     logFC, adj.P.Val, mim_log_sum_diff, hydro_log_sum_diff, 
     AA_POS, AA_CHANGE.x, AA_CHANGE.y, Label, Domain, flags,
     PROVEAN, `PolyPhen-2`, CADD, MAF, LLR, 
     `VE map fitness\n(low vitamin B6)`, 
     `VE map fitness\n(high vitamin B6)`)]

setkey(cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_val_final_dt, Chemistry, logFC)

write.table(
  cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_val_final_dt,
  paste(cbs_02dec2021_dir, "cbs_02dec2021_both_postprocess_subtract_filt_yfc_merge_final.txt", 
        sep="/"), sep="\t", quote=FALSE, row.names=FALSE)

neutral_del_gg<- ggplot(
  data=cbs_02dec2021_both_postprocess_subtract_rf_yfc_filt_cast_merge_val_final_dt,
  aes(x=Label, y=logFC, col=LLR))

neutral_del_gg+
  geom_boxplot(width=0.5, col="black")+
  geom_jitter(width=0.25)+
  scale_color_viridis(option="C", begin=0.1, end=0.9)+
  labs(x="Class", y="log FC", 
       col="log-likelihood ratio")+
  theme_cowplot()

```

## Analysis of Weile and Sun SUMO1 dataset

Benchmark satmut_utils and DiMSum on SUMO1 dataset with one tile and two replicates.

```{r SUMO1 dataset prep for DiMSum STEAM}

sumo1_base_dir<- paste(base_dir, "SUMO1", "results", "postprocess", sep="/")
sumo1_dimsum_postprocess_dt<- merge_dimsum_tables(indir=sumo1_base_dir)

# Clean up column names as DiMSum does not except sample names with non-alphanumeric letters
sumo1_dimsum_postprocess_names<- names(sumo1_dimsum_postprocess_dt)
sumo1_srr_col_idx<- grepl("SRR", sumo1_dimsum_postprocess_names)
sumo1_srr_col_keys<- sumo1_dimsum_postprocess_names[!sumo1_srr_col_idx]
sumo1_srr_col_samples<- sumo1_dimsum_postprocess_names[sumo1_srr_col_idx]
sumo1_srr_col_samples_clean<- sapply(strsplit(sumo1_srr_col_samples, "_"), "[", 1)
names(sumo1_dimsum_postprocess_dt)<- c(sumo1_srr_col_keys, sumo1_srr_col_samples_clean)

# Define controls for background subtract
sumo1_nc_sample_reps<- c("SRR5680641", "SRR5680642")

# Aggregate the NC samples then remove the replicates
sumo1_dimsum_postprocess_dt[,SRR5680641_2:=as.integer((SRR5680641 + SRR5680642)/2)]
sumo1_dimsum_postprocess_dt[,`:=`(SRR5680641=NULL, SRR5680642=NULL)]

# Verify that the WT samples have much lower counts than plasmid library or virus samples
sumo1_dimsum_postprocess_melt_dt<- melt(
  data=sumo1_dimsum_postprocess_dt[,-c("nt_seq")], id.vars=c("VAR_ID"), 
  variable.name="Sample", value.name="Counts")

sumo1_count_distr_gg<- ggplot(
  data=sumo1_dimsum_postprocess_melt_dt, 
  aes(x=log10(Counts+1), col=Sample))

sumo1_count_distr_gg+
  geom_density()+
  theme_cowplot()

# Aggregate duplicates before subtraction
sumo1_nc_samples<- list("NC"="SRR5680641_2")

sumo1_test_samples<- list("Test"=c("SRR5680629", "SRR5680633", "SRR5680634", "SRR5680643"))

# Subtract counts from the WT libraries
sumo1_dimsum_postprocess_subtract_dt<- subtract_neg_control_cpms(
  dt=sumo1_dimsum_postprocess_dt, test_samples=sumo1_test_samples, nc_samples=sumo1_nc_samples)

has_negative_counts<- function(x){
  res<- any(x<0)
  return(res)
}

sumo1_dimsum_postprocess_subtract_drop<- apply(
  sumo1_dimsum_postprocess_subtract_dt[,-c("nt_seq", "VAR_ID")], 1, has_negative_counts)

sumo1_drop_fields<- c(unlist(sumo1_nc_samples), "VAR_ID")

sumo1_dimsum_postprocess_subtract_drop_dt<- sumo1_dimsum_postprocess_subtract_dt[
  !sumo1_dimsum_postprocess_subtract_drop, -c("VAR_ID"), with=FALSE]

sumo1_dimsum_postprocess_filt_dt<- sumo1_dimsum_postprocess_dt[,-sumo1_drop_fields,with=FALSE]

write.table(sumo1_dimsum_postprocess_filt_dt, 
            paste(sumo1_base_dir, "SUMO1_DiMSum.raw.counts.txt", sep="/"), 
            row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t")

write.table(sumo1_dimsum_postprocess_subtract_drop_dt, 
            paste(sumo1_base_dir, "SUMO1_DiMSum.subtract.counts.txt", sep="/"), 
            row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t")

```


```{r SUMO1 satmut_utils vs DiMSum call}

sumo1_analysis_dir<- paste(base_dir, "SUMO1", "results", "dimsum_analysis", sep="/")
sumo1_analysis_dimsum_dir<- paste(sumo1_analysis_dir, "dimsum_results", sep="/")
sumo1_analysis_satmut_utils_dir<- paste(sumo1_analysis_dir, "dimsum_results_satmut_utils_subtract", sep="/")

fitness_files<- paste(c("fitness_wildtype.txt", "fitness_synonymous.txt", 
                        "fitness_singles.txt", "fitness_doubles.txt"), collapse="|")

sumo1_dimsum_list<- lapply(dir(sumo1_analysis_dimsum_dir, pattern=fitness_files, full.names=T), fread)
sumo1_satmut_utils_list<- lapply(dir(sumo1_analysis_satmut_utils_dir, pattern=fitness_files, full.names=T), fread)

sumo1_dimsum_list_nrows<- sapply(sumo1_dimsum_list, nrow)
sumo1_satmut_utils_list_nrows<- sapply(sumo1_satmut_utils_list, nrow)
  
sumo1_dimsum_all_dt<- rbindlist(sumo1_dimsum_list, fill=T)
sumo1_satmut_utils_all_dt<- rbindlist(sumo1_satmut_utils_list, fill=T)

# Annotate fitness source
sumo1_dimsum_all_dt[,Type:=as.factor(rep(c("Double", "Missense", "Silent", "Wildtype"), times=sumo1_dimsum_list_nrows))]
sumo1_satmut_utils_all_dt[,Type:=as.factor(rep(c("Double", "Missense", "Silent", "Wildtype"), times=sumo1_satmut_utils_list_nrows))]

sumo1_both_dt<- rbind(sumo1_dimsum_all_dt, sumo1_satmut_utils_all_dt)

sumo1_both_dt[,Software:=as.factor(rep(c("DiMSum", "satmut_utils"), times=c(
  nrow(sumo1_dimsum_all_dt), nrow(sumo1_satmut_utils_all_dt))))]

rm(list=c("sumo1_dimsum_list", "sumo1_satmut_utils_list", "sumo1_dimsum_all_dt", "sumo1_satmut_utils_all_dt"))

# sumo1_both_dt[is.na(fitness), .N, by=.(Type, Software)]
# Doubles have no fitness scores

# sumo1_both_dt[is.na(nt_seq), .N, by=.(Type, Software)]
# nt_seq is not available for doubles and singles?

# sumo1_both_dt[,length(unique(aa_seq)), by=.(Type, Software)]
# AA makes sense

# the amplicon amino acids are annotated for DiMSum but
# the full protein sequences is provided by satmut_utils postprocessing script
# Slice the full protein sequence for merge with DiMSum
sumo1_both_dt[Software=="satmut_utils", nt_seq:=substr(nt_seq, 82, 231)]
sumo1_both_dt[Software=="satmut_utils", aa_seq:=substr(aa_seq, 28, 77)]

# Annotate nonsense variants
sumo1_both_dt[STOP==TRUE, Type:="Nonsense"]
sumo1_both_dt[,Type:=factor(Type, levels=c("Silent", "Missense", "Nonsense", "Wildtype"))]

# Cast the dataset so fitness estimates can be compared by scatterplot
# Need both nt_seq and aa_seq as key to compare all variants
sumo1_both_cast_dt<- dcast(
  data=sumo1_both_dt[Type!="Double"], formula=nt_seq+aa_seq+Type~Software, value.var="fitness")

# Compare fitness estimates between software
sumo1_fitness_cor_gg<- ggplot(
  data=sumo1_both_cast_dt[Type!="Wildtype"], aes(x=DiMSum, y=satmut_utils))

sumo1_fitness_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point()+
  geom_density2d()+
  theme_cowplot()+
  expand_limits(y=c(-6,3), x=c(-6,3))+
  labs(x="Fitness (DiMSum)", y="Fitness (satmut_utils)")

# sumo1_both_cast_dt[!is.na(satmut_utils) & !is.na(DiMSum), cor(x=DiMSum, y=satmut_utils)]
# 0.84 Pearson correlation

# Plot distribution of fitness estimates for synonymous and missense variants
sumo1_fitness_distr_gg<- ggplot(
  data=sumo1_both_dt[Type!="Wildtype"],
  aes(x=Type, y=fitness, col=Software))

sumo1_fitness_distr_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(position="dodge")+
  scale_color_manual(values=pal_npg()(5)[c(1,4)])+
  coord_cartesian(ylim=c(-6, 3))+
  theme_cowplot()+
  labs(x="", y="Fitness")+
  theme(legend.position="bottom")

# Run tests
sumo1_silent_wilcox<- sumo1_both_dt[Type=="Silent", wilcox.test(formula=fitness~Software, data=.SD)]
sumo1_missense_wilcox<- sumo1_both_dt[Type=="Missense", wilcox.test(formula=fitness~Software, data=.SD)]
sumo1_nonsense_wilcox<- sumo1_both_dt[Type=="Nonsense", wilcox.test(formula=fitness~Software, data=.SD)]

# Plot number and overlap of calls
sumo1_both_dt[,Key:=paste(aa_seq, nt_seq, sep=":")]

sumo1_both_list<- split(sumo1_both_dt[Type!="Double", Key], sumo1_both_dt[Type!="Double", Software])

sumo1_venn_diagram_file<- paste(sumo1_analysis_dir, "sumo1_venn_diagram.svg", sep="/")

sumo1_venn_diagram<- venn.diagram(sumo1_both_list, imagetype="svg",
  filename=sumo1_venn_diagram_file, output=TRUE, fill=pal_npg()(5)[c(1,4)], 
  height=5, width=5)

```


## Analysis of Doud and Bloom WSN HA dataset

Benchmark satmut_utils and DiMSum on HA dataset with six tiles and three replicates.
 
```{r WSN HA dataset prep for DiMSum STEAM}

wsn_ha_base_dir<- paste(base_dir, "HA", "results", "postprocess", sep="/")
dimsum_postprocess_dt<- merge_dimsum_tables(indir=wsn_ha_base_dir)

# Clean up column names as DiMSum does not except smaple names with non-alphanumeric letters
dimsum_postprocess_names<- names(dimsum_postprocess_dt)
srr_col_idx<- grepl("SRR", dimsum_postprocess_names)
srr_col_keys<- dimsum_postprocess_names[!srr_col_idx]
srr_col_samples<- dimsum_postprocess_names[srr_col_idx]
srr_col_samples_clean<- sapply(strsplit(srr_col_samples, "_"), "[", 1)
names(dimsum_postprocess_dt)<- c(srr_col_keys, srr_col_samples_clean)

# Verify that the WT samples have much lower counts than plasmid library or virus samples
dimsum_postprocess_melt_dt<- melt(
  data=dimsum_postprocess_dt[,-c("nt_seq")], id.vars=c("VAR_ID"), variable.name="Sample", value.name="Counts")

wsn_ha_count_distr_gg<- ggplot(
  data=dimsum_postprocess_melt_dt, 
  aes(x=log10(Counts+1), col=Sample))

wsn_ha_count_distr_gg+
  geom_density()+
  theme_cowplot()

# 655 and 659 are the WT controls
# Indeed they are left-shifted compared to other samples
# In addition the virus samples (660-662) are also left-shifted compared to 
# plasmid library (655-657), presumably because many mutations are not fit
wsn_ha_wt_samples<- c("SRR3113655", "SRR3113659")

nc_samples<- list("NC1"="SRR3113655", "NC2"="SRR3113659")

test_samples<- list("Plasmid"=c("SRR3113656", "SRR3113657", "SRR3113658"), 
                    "Virus"=c("SRR3113660", "SRR3113661", "SRR3113662"))

# Subtract counts from the WT libraries
dimsum_postprocess_subtract_dt<- subtract_neg_control_cpms(
  dt=dimsum_postprocess_dt, test_samples=test_samples, nc_samples=nc_samples)

has_negative_counts<- function(x){
  res<- any(x<0)
  return(res)
}

dimsum_postprocess_subtract_drop<- apply(
  dimsum_postprocess_subtract_dt[,-c("nt_seq", "VAR_ID")], 1, has_negative_counts)

wsn_ha_drop_fields<- c(wsn_ha_wt_samples, "VAR_ID")

dimsum_postprocess_subtract_drop_dt<- dimsum_postprocess_subtract_dt[
  !dimsum_postprocess_subtract_drop, -c("VAR_ID"), with=FALSE]

dimsum_postprocess_filt_dt<- dimsum_postprocess_dt[,-wsn_ha_drop_fields,with=FALSE]

dimsum_postprocess_var_id_map_dt<- dimsum_postprocess_dt[,.(VAR_ID, nt_seq)]
dimsum_postprocess_var_id_map_dt[,nt_seq:=tolower(nt_seq)]

write.table(dimsum_postprocess_var_id_map_dt, 
            paste(wsn_ha_base_dir, "WSN_HA_id_map.txt", sep="/"), 
            row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t")

write.table(dimsum_postprocess_filt_dt, 
            paste(wsn_ha_base_dir, "WSN_HA_DiMSum.raw.counts.txt", sep="/"), 
            row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t")

write.table(dimsum_postprocess_subtract_drop_dt, 
            paste(wsn_ha_base_dir, "WSN_HA_DiMSum.subtract.counts.txt", sep="/"), 
            row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t")

```

```{r WSN HA satmut_utils vs DiMSum call}

# Do not read in doubles as some amplicons have no doubles
fitness_files<- paste(c("fitness_wildtype.txt", "fitness_synonymous.txt", "fitness_singles.txt"), collapse="|")
fitness_file_ids<- c("Missense", "Silent", "Wildtype")

ha_analysis_dir<- paste(base_dir, "HA", sep="/")
ha_analysis_dimsum_1_dir<- paste(ha_analysis_dir, "dimsum_results_amp1", sep="/")
ha_analysis_dimsum_2_dir<- paste(ha_analysis_dir, "dimsum_results_amp2", sep="/")
ha_analysis_dimsum_3_dir<- paste(ha_analysis_dir, "dimsum_results_amp3", sep="/")
ha_analysis_dimsum_4_dir<- paste(ha_analysis_dir, "dimsum_results_amp4", sep="/")
ha_analysis_dimsum_5_dir<- paste(ha_analysis_dir, "dimsum_results_amp5", sep="/")
ha_analysis_dimsum_6_dir<- paste(ha_analysis_dir, "dimsum_results_amp6", sep="/")
ha_analysis_satmut_utils_dir<- paste(ha_analysis_dir, "dimsum_results_satmut_utils_subtract", sep="/")

# Read in all fitness files from each dir and track records for annotation
ha_dimsum_1_list<- lapply(dir(ha_analysis_dimsum_1_dir, pattern=fitness_files, full.names=T), fread)
ha_dimsum_2_list<- lapply(dir(ha_analysis_dimsum_2_dir, pattern=fitness_files, full.names=T), fread)
ha_dimsum_3_list<- lapply(dir(ha_analysis_dimsum_3_dir, pattern=fitness_files, full.names=T), fread)
ha_dimsum_4_list<- lapply(dir(ha_analysis_dimsum_4_dir, pattern=fitness_files, full.names=T), fread)
ha_dimsum_5_list<- lapply(dir(ha_analysis_dimsum_5_dir, pattern=fitness_files, full.names=T), fread)
ha_dimsum_6_list<- lapply(dir(ha_analysis_dimsum_6_dir, pattern=fitness_files, full.names=T), fread)
ha_satmut_utils_list<- lapply(dir(ha_analysis_satmut_utils_dir, pattern=fitness_files, full.names=T), fread)

ha_dimsum_1_list_nrows<- sapply(ha_dimsum_1_list, nrow)
ha_dimsum_2_list_nrows<- sapply(ha_dimsum_2_list, nrow)
ha_dimsum_3_list_nrows<- sapply(ha_dimsum_3_list, nrow)
ha_dimsum_4_list_nrows<- sapply(ha_dimsum_4_list, nrow)
ha_dimsum_5_list_nrows<- sapply(ha_dimsum_5_list, nrow)
ha_dimsum_6_list_nrows<- sapply(ha_dimsum_6_list, nrow)
ha_satmut_utils_list_nrows<- sapply(ha_satmut_utils_list, nrow)
  
ha_dimsum_1_dt<- rbindlist(ha_dimsum_1_list, fill=T)
ha_dimsum_2_dt<- rbindlist(ha_dimsum_2_list, fill=T)
ha_dimsum_3_dt<- rbindlist(ha_dimsum_3_list, fill=T)
ha_dimsum_4_dt<- rbindlist(ha_dimsum_4_list, fill=T)
ha_dimsum_5_dt<- rbindlist(ha_dimsum_5_list, fill=T)
ha_dimsum_6_dt<- rbindlist(ha_dimsum_6_list, fill=T)
ha_satmut_utils_all_dt<- rbindlist(ha_satmut_utils_list, fill=T)

# Annotate fitness source
# Use missense for "single" and silent for "synonymous" files, annotate nonsense after using STOP field
ha_dimsum_1_dt[,Type:=as.factor(rep(fitness_file_ids, times=ha_dimsum_1_list_nrows))]
ha_dimsum_2_dt[,Type:=as.factor(rep(fitness_file_ids, times=ha_dimsum_2_list_nrows))]
ha_dimsum_3_dt[,Type:=as.factor(rep(fitness_file_ids, times=ha_dimsum_3_list_nrows))]
ha_dimsum_4_dt[,Type:=as.factor(rep(fitness_file_ids, times=ha_dimsum_4_list_nrows))]
ha_dimsum_5_dt[,Type:=as.factor(rep(fitness_file_ids, times=ha_dimsum_5_list_nrows))]
ha_dimsum_6_dt[,Type:=as.factor(rep(fitness_file_ids, times=ha_dimsum_6_list_nrows))]
ha_satmut_utils_all_dt[,Type:=as.factor(rep(fitness_file_ids, times=ha_satmut_utils_list_nrows))]

ha_dimsum_all_dt<- rbind(ha_dimsum_1_dt, ha_dimsum_2_dt, ha_dimsum_3_dt, ha_dimsum_4_dt, ha_dimsum_5_dt, ha_dimsum_6_dt)

ha_dimsum_all_dt[,Amplicon:=rep(paste("amp", seq(1,6), sep=""), times=c(
  nrow(ha_dimsum_1_dt), nrow(ha_dimsum_2_dt), nrow(ha_dimsum_3_dt), 
  nrow(ha_dimsum_4_dt), nrow(ha_dimsum_5_dt), nrow(ha_dimsum_6_dt)))]

ha_both_dt<- rbind(ha_dimsum_all_dt, ha_satmut_utils_all_dt, fill=TRUE)

ha_both_dt[,Software:=as.factor(rep(c("DiMSum", "satmut_utils"), times=c(
  nrow(ha_dimsum_all_dt), nrow(ha_satmut_utils_all_dt))))]

ha_both_dt[STOP==TRUE, Type:="Nonsense"]
ha_both_dt[,Type:=factor(Type, levels=c("Silent", "Missense", "Nonsense", "Wildtype"))]

rm(list=c("ha_dimsum_1_list", "ha_dimsum_2_list", "ha_dimsum_3_list", "ha_dimsum_4_list", "ha_dimsum_5_list", "ha_dimsum_6_list",
          "ha_dimsum_1_dt", "ha_dimsum_2_dt", "ha_dimsum_3_dt", "ha_dimsum_4_dt", "ha_dimsum_5_dt", "ha_dimsum_6_dt",
          "ha_satmut_utils_list", "ha_dimsum_all_dt", "ha_satmut_utils_all_dt"))

# ha_both_dt[is.na(nt_seq), .N, by=.(Type, Software)]
# nt_seq is not available for singles?

# ha_both_dt[,length(unique(aa_seq)), by=.(Type, Software)]
# AA makes sense

# The amplicon amino acids are annotated for DiMSum but
# the full protein sequences is provided by satmut_utils postprocessing script
# Slice the full protein sequence for merge with DiMSum
ha_amplicon_nt_ends<- list("amp1"=c(28,312), "amp2"=c(313,597), "amp3"=c(598,882), 
                           "amp4"=c(883, 1167), "amp5"=c(1168, 1452), "amp6"=c(1453, 1725))

get_ha_amplicon_aa_lengths<- function(x){
  amp_range<- x[length(x)] - x[1] + 1
  num_aas<- amp_range/3
  return(num_aas)
}

subtract_cds_start<- function(x, cds_start_idx=27){
  res<- x - cds_start_idx
  return(res)
}

# Need to convert to reference-based to CDS based coordinates for slicing
ha_amplicon_nt_cds_ends<- lapply(ha_amplicon_nt_ends, subtract_cds_start)

# sapply(ha_amplicon_nt_cds_ends, get_ha_amplicon_aa_lengths)
# Each amplicon is 95 AAs except for amp6 (91 AAs)

ha_amplicon_aa_ends<- list("amp1"=c(1, 95), "amp2"=c(96, 190), "amp3"=c(191, 285), 
                           "amp4"=c(286, 380), "amp5"=c(381, 475), "amp6"=c(476, 566))

ha_amplicon_nt_seqs<- list(
  "amp1"="ATGAAGGCAAAACTACTGGTCCTGTTATATGCATTTGTAGCTACAGATGCAGACACAATATGTATAGGCTACCATGCGAACAACTCAACCGACACTGTTGACACAATACTCGAGAAGAATGTGGCAGTGACACATTCTGTTAACCTGCTCGAAGACAGCCACAACGGGAAACTATGTAAATTAAAAGGAATAGCCCCACTACAATTGGGGAAATGTAACATCACCGGATGGCTCTTGGGAAATCCAGAATGCGACTCACTGCTTCCAGCGAGATCATGGTCCTAC",
  "amp2"="ATTGTAGAAACACCAAACTCTGAGAATGGAGCATGTTATCCAGGAGATCTCATCGACTATGAGGAACTGAGGGAGCAATTGAGCTCAGTATCATCATTAGAAAGATTCGAAATATTTCCCAAGGAAAGTTCATGGCCCAACCACACATTCAACGGAGTAACAGTATCATGCTCCCATAGGGGAAAAAGCAGTTTTTACAGAAATTTGCTATGGCTGACGAAGAAGGGGGATTCATACCCAAAGCTGACCAATTCCTATGTGAACAATAAAGGGAAAGAAGTCCTT",
  "amp3"="GTACTATGGGGTGTTCATCACCCGTCTAGCAGTGATGAGCAACAGAGTCTCTATAGTAATGGAAATGCTTATGTCTCTGTAGCGTCTTCAAATTATAACAGGAGATTCACCCCGGAAATAGCTGCAAGGCCCAAAGTAAGAGATCAACATGGGAGGATGAACTATTACTGGACCTTGCTAGAACCCGGAGACACAATAATATTTGAGGCAACTGGTAATCTAATAGCACCATGGTATGCTTTCGCACTGAGTAGAGGGTTTGAGTCCGGCATCATCACCTCAAAC", 
  "amp4"="GCGTCAATGCATGAGTGTAACACGAAGTGTCAAACACCCCAGGGAGCTATAAACAGCAATCTCCCTTTCCAGAATATACACCCAGTCACAATAGGAGAGTGCCCAAAATATGTCAGGAGTACCAAATTGAGGATGGTTACAGGACTAAGAAACATCCCATCCATTCAATACAGAGGTCTATTTGGAGCCATTGCTGGTTTTATTGAGGGGGGATGGACTGGAATGATAGATGGATGGTATGGTTATCATCATCAGAATGAACAGGGATCAGGCTATGCAGCGGAT",
  "amp5"="CAAAAAAGCACACAAAATGCCATTAACGGGATTACAAACAAGGTGAACTCTGTTATCGAGAAAATGAACACTCAATTCACAGCTGTGGGTAAAGAATTCAACAACTTAGAAAAAAGGATGGAAAATTTAAATAAAAAAGTTGATGATGGGTTTCTGGACATTTGGACATATAATGCAGAATTGTTAGTTCTACTGGAAAATGAAAGGACTTTGGATTTCCATGACTTAAATGTGAAGAATCTGTACGAGAAAGTAAAAAGCCAATTAAAGAATAATGCCAAAGAA",
  "amp6"="ATCGGAAATGGGTGTTTTGAGTTCTACCACAAGTGTGACAATGAATGCATGGAAAGTGTAAGAAATGGGACTTATGATTATCCAAAATATTCAGAAGAATCAAAGTTGAACAGGGAAAAGATAGATGGAGTGAAATTGGAATCAATGGGGGTGTATCAGATTCTGGCGATCTACTCAACTGTCGCCAGTTCACTGGTGCTTTTGGTCTCCCTGGGGGCAATCAGTTTCTGGATGTGTTCTAATGGGTCTTTGCAGTGCAGAATATGCATC")

ha_amplicon_aa_dss<- DNAStringSet(unlist(ha_amplicon_nt_seqs))
ha_amplicon_aa_seqs<- as.list(as.character(translate(ha_amplicon_aa_dss)))

# Merge with the VAR_ID map to determine what amplicon variants are from
# in satmut_utils calls
ha_both_merge_dt<- merge(ha_both_dt, dimsum_postprocess_var_id_map_dt, by="nt_seq", all.x=TRUE)

# Use the VAR_ID POS to determine the originating amplicon for silent variants
ha_both_merge_dt[,VAR_POS:=as.integer(sapply(strsplit(VAR_ID, ":"), "[", 2))]

ha_both_merge_dt[Software=="satmut_utils" & VAR_POS%in%seq(
  ha_amplicon_nt_ends[[1]][1], ha_amplicon_nt_ends[[1]][2]), Amplicon:="amp1"]

ha_both_merge_dt[Software=="satmut_utils" & VAR_POS%in%seq(
  ha_amplicon_nt_ends[[2]][1], ha_amplicon_nt_ends[[2]][2]), Amplicon:="amp2"]

ha_both_merge_dt[Software=="satmut_utils" & VAR_POS%in%seq(
  ha_amplicon_nt_ends[[3]][1], ha_amplicon_nt_ends[[3]][2]), Amplicon:="amp3"]

ha_both_merge_dt[Software=="satmut_utils" & VAR_POS%in%seq(
  ha_amplicon_nt_ends[[4]][1], ha_amplicon_nt_ends[[4]][2]), Amplicon:="amp4"]

ha_both_merge_dt[Software=="satmut_utils" & VAR_POS%in%seq(
  ha_amplicon_nt_ends[[5]][1], ha_amplicon_nt_ends[[5]][2]), Amplicon:="amp5"]

ha_both_merge_dt[Software=="satmut_utils" & VAR_POS%in%seq(
  ha_amplicon_nt_ends[[6]][1], ha_amplicon_nt_ends[[6]][2]), Amplicon:="amp6"]

ha_both_merge_dt[Software=="satmut_utils" & is.na(Amplicon), Amplicon:="WT"]

# Now use amplicon amino acid sequences to annotate missense variants, 
# as these variants are not annotated with a nt_seq and are aggregated
# at protein-level across genetic variants 
ha_both_merge_dt[Software=="satmut_utils" & Type%in%c("Missense","Nonsense") & 
                   !grepl(ha_amplicon_aa_seqs[[1]], aa_seq, fixed=T), Amplicon:="amp1"]

ha_both_merge_dt[Software=="satmut_utils" & Type%in%c("Missense","Nonsense") & 
                   !grepl(ha_amplicon_aa_seqs[[2]], aa_seq, fixed=T), Amplicon:="amp2"]

ha_both_merge_dt[Software=="satmut_utils" & Type%in%c("Missense","Nonsense") & 
                   !grepl(ha_amplicon_aa_seqs[[3]], aa_seq, fixed=T), Amplicon:="amp3"]

ha_both_merge_dt[Software=="satmut_utils" & Type%in%c("Missense","Nonsense") & 
                   !grepl(ha_amplicon_aa_seqs[[4]], aa_seq, fixed=T), Amplicon:="amp4"]

ha_both_merge_dt[Software=="satmut_utils" & Type%in%c("Missense","Nonsense") & 
                   !grepl(ha_amplicon_aa_seqs[[5]], aa_seq, fixed=T), Amplicon:="amp5"]

ha_both_merge_dt[Software=="satmut_utils" & Type%in%c("Missense","Nonsense") & 
                   !grepl(ha_amplicon_aa_seqs[[6]], aa_seq, fixed=T), Amplicon:="amp6"]

# Update nt seqs
# First add the stop codon back as this was not included in the nt_seq for satmut_utils'
# postprocessed DiMSum count files; only needed for silent calls with annotated nt_seq
# from ha_amplicon_nt_cds_ends
ha_both_merge_dt[Software=="satmut_utils" & Type=="Silent", nt_seq:=paste(nt_seq, "tga", sep="")]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp1", nt_seq:=substr(nt_seq, 1, 285)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp2", nt_seq:=substr(nt_seq, 286, 570)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp3", nt_seq:=substr(nt_seq, 571, 855)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp4", nt_seq:=substr(nt_seq, 856, 1140)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp5", nt_seq:=substr(nt_seq, 1141, 1425)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp6", nt_seq:=substr(nt_seq, 1426, 1698)]

# Update AA seqs
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp1", aa_seq:=substr(aa_seq, 1, 95)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp2", aa_seq:=substr(aa_seq, 96, 190)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp3", aa_seq:=substr(aa_seq, 191, 285)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp4", aa_seq:=substr(aa_seq, 286, 380)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp5", aa_seq:=substr(aa_seq, 381, 475)]
ha_both_merge_dt[Software=="satmut_utils" & Amplicon=="amp6", aa_seq:=substr(aa_seq, 476, 566)]

# Cast the dataset so fitness estimates can be compared by scatterplot
# Need both nt_seq and aa_seq as key to compare all variants
# Aggregate wildtype-estimates from each amplicon for DiMSum
# satmut_utils only has one WT estimate as all amplicons analyzed together

# First add labels to a few variants to highlight
ha_fitness_thresh<- 1.75
ha_examp_aa_dt<- ha_both_merge_dt[Software=="satmut_utils" & fitness > ha_fitness_thresh, 
                 .(aa_seq, AA_CHANGE=paste(WT_AA, Pos, Mut, sep=""))]

ha_both_cast_dt<- dcast(
  data=ha_both_merge_dt, 
  formula=nt_seq+aa_seq+Type~Software, value.var="fitness", fun.aggregate=mean)

ha_both_cast_dt[,Label:=ha_examp_aa_dt[match(ha_both_cast_dt[,aa_seq], aa_seq), AA_CHANGE]]

# Compare fitness estimates between software
ha_fitness_cor_gg<- ggplot(
  data=ha_both_cast_dt[Type!="Wildtype"], aes(x=DiMSum, y=satmut_utils, label=Label))

ha_fitness_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point()+
  geom_density2d()+
  geom_text_repel()+
  theme_cowplot()+
  expand_limits(y=c(-6,3), x=c(-6,3))+
  labs(x="Fitness (DiMSum)", y="Fitness (satmut_utils)")

# ha_both_cast_dt[!is.na(satmut_utils) & !is.na(DiMSum), cor(x=DiMSum, y=satmut_utils)]
# 0.78 Pearson correlation

# Plot distribution of fitness estimates for synonymous and missense variants
ha_fitness_distr_gg<- ggplot(
  data=ha_both_dt[Type!="Wildtype"],
  aes(x=Type, y=fitness, col=Software))

ha_fitness_distr_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(position="dodge")+
  scale_color_manual(values=pal_npg()(5)[c(1,4)])+
  coord_cartesian(ylim=c(-6, 3))+
  theme_cowplot()+
  labs(x="", y="Fitness")+
  theme(legend.position="bottom")

# Run tests
ha_silent_wilcox<- ha_both_dt[Type=="Silent", wilcox.test(formula=fitness~Software, data=.SD)]
ha_missense_wilcox<- ha_both_dt[Type=="Missense", wilcox.test(formula=fitness~Software, data=.SD)]
ha_nonsense_wilcox<- ha_both_dt[Type=="Nonsense", wilcox.test(formula=fitness~Software, data=.SD)]

# Plot number and overlap of calls
ha_both_merge_dt[,Key:=paste(aa_seq, nt_seq, sep=":")]

ha_both_list<- split(ha_both_merge_dt[,Key], ha_both_merge_dt[,Software])

ha_venn_diagram_file<- paste(ha_analysis_dir, "ha_venn_diagram.svg", sep="/")

ha_venn_diagram<- venn.diagram(ha_both_list, imagetype="svg",
  filename=ha_venn_diagram_file, output=TRUE, fill=pal_npg()(5)[c(1,4)], 
  height=5, width=5)

```


## SUMO1 and HA dataset variant counts

```{r SUMO1 and HA variant counts}

sumo1_ha_both_dt<- rbind(sumo1_both_dt, ha_both_dt, fill=TRUE)

sumo1_ha_both_dt[,Dataset:=as.factor(rep(
  c("SUMO1", "HA"), times=c(nrow(sumo1_both_dt), nrow(ha_both_dt))))]

sumo1_ha_both_dt[is.na(Type), Type:="Double"]
sumo1_ha_both_dt[,Type:=factor(Type, levels=c(
  "Silent", "Missense", "Nonsense", "Wildtype", "Double"))]

sumo1_ha_both_counts_dt<- sumo1_ha_both_dt[,.N,by=.(Dataset, Software, Type)]

# For simplicity, fitness_doubles.txt were not read for HA analysis, as only two double variants
# were found in amplicons 1 and 2. Add these back in:

sumo1_ha_both_counts_update_dt<- rbind(
  sumo1_ha_both_counts_dt, 
  data.table(Dataset=rep("HA", 2), Software=c("DiMSum", "satmut_utils"), 
             Type=rep("Double", 2), N=c(2,0)))

sumo1_ha_both_counts_update_dt[,Dataset:=factor(Dataset,levels=c("SUMO1", "HA"))]

# Do not plot wildtype or double variants for comparison
sumo1_ha_both_counts_gg<- ggplot(
  data=sumo1_ha_both_counts_update_dt[!Type%in%c("Wildtype", "Double")], 
  aes(x=Type, y=log10(N+1), col=Software))

sumo1_ha_both_counts_gg+
  facet_wrap(~Dataset)+
  geom_point(position=position_jitterdodge(dodge.width=0.5), size=4)+
  labs(x="", y="log10(variant counts)")+
  scale_color_manual(values=pal_npg()(5)[c(1,4)])+
  coord_cartesian(ylim=c(1,4))+
  theme_cowplot()+
  theme(legend.position="bottom")

# Determine differences in fitness estimates
sumo1_ha_both_fitness_diff_dt<- sumo1_ha_both_dt[
  ,.(Median_fitness=median(fitness, na.rm=T)), by=.(Software, Dataset, Type)]

sumo1_ha_both_fitness_diff_cast_dt<- dcast(
  data=sumo1_ha_both_fitness_diff_dt, formula=Dataset+Software~Type, value.var="Median_fitness")

sumo1_ha_both_fitness_diff_cast_dt[,`:=`(Silent_missense_diff=Silent-Missense, Silent_nonsense_diff=Silent-Nonsense)]

# sumo1_ha_both_fitness_diff_cast_dt[,.(Software, Dataset, Silent_missense_diff, Silent_nonsense_diff)]
# For both datasets, satmut_utils increases the distance between 
# silent and missense, and silent and nonsense variants


```


## SUMO1 and HA dataset biological replicate correlation

```{r SUMO1 and HA replicate correlation}

sumo1_ha_cor_dir<- paste(base_dir, "SUMO1_HA_correlation", sep="/")
sumo1_ha_cor_file<- paste(sumo1_ha_cor_dir, "SUMO1_HA_rep_correlations.txt", sep="/")
sumo1_ha_cor_dt<- fread(sumo1_ha_cor_file, sep="\t")

sumo1_ha_cor_summ_dt<- sumo1_ha_cor_dt[
  ,.(Mean_cor=mean(Correlation), SD_cor=sd(Correlation)), 
  by=.(Software, Dataset, Source)]

sumo1_ha_cor_summ_dt[is.na(SD_cor), SD_cor:=0.0]
sumo1_ha_cor_summ_dt[,Dataset:=factor(Dataset, levels=c("SUMO1", "HA"))]

sumo1_ha_cor_gg<- ggplot(
  data=sumo1_ha_cor_summ_dt, 
  aes(x=Source, fill=Software, y=Mean_cor, 
      ymax=Mean_cor+SD_cor, ymin=Mean_cor-SD_cor))

sumo1_ha_cor_gg+
  facet_wrap(~Dataset)+
  geom_bar(stat="identity", position="dodge", alpha=0.9)+
  geom_errorbar(position=position_dodge(0.9), col="grey", width=0.3)+
  scale_fill_manual(values=pal_npg()(5)[c(1,4)])+
  theme_cowplot()+
  theme(legend.position="bottom")
  

```

